---
title: "CDK7 vs GR metrics"
author: "Clemens Hug"
date: "2023-07-28"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(data.table)
library(qs)
library(powerjoin)
library(synExtra)
library(IHW)
library(conflicted)
library(furrr)
library(openxlsx)
library(seriation)
library(withr)

conflicts_prefer(
  dplyr::filter,
  dplyr::select,
  base::setdiff
)

synapser::synLogin()
syn <- synDownloader("~/data", .cache = TRUE)

syn_analysis <- synMkdir("syn52226362", "analysis")
```


```{r}
ensembl_gtf_combined <- syn("syn71972264") %>%
  read_csv()

gene_id_to_name_map <- ensembl_gtf_combined %>%
  split(.$assay) %>%
  map(\(x) with(x, set_names(gene_name, ensembl_gene_id)))

deep_meta <- syn("syn60529546") %>%
  readxl::read_excel() %>%
  mutate(
    sample_id = paste0("S", Sample)
  ) %>%
  dplyr::rename(cell_line = `Cell line`) %>%
  mutate(
    cell_line = str_replace_all(cell_line, fixed("-"), "")
  )

dge_meta <- syn("syn71968688") %>%
  read_csv()

dge_counts_combined <- syn("syn71969557") %>%
  read_csv()

deseq_res_combined <- synPluck(
  syn_analysis, "cdk7_deseq_res_combined_long.csv"
) %>%
  syn() %>%
  read_csv()

deseq_res_gr <- synPluck(
  syn_analysis, "cdk7_deseq_gr_and_normal_res_long.csv.gz"
) %>%
  syn() %>%
  read_csv()

deseq_res_baseline_dge <- synPluck(
  syn_analysis, "cdk7_deseq_baseline_res.csv"
) %>%
  syn() %>%
  read_csv()

deseq_res_counts_gr_corrected <- synPluck(
  syn_analysis, "cdk7_deseq_res_counts_gr_corrected_long.csv"
) %>%
  syn() %>%
  read_csv()

deseq_res_acquired <- syn("syn69692661") %>%
  read_csv()


deseq_res_acquired_ykl <- syn("syn69696874") %>%
  read_csv()

counts_baseline <- syn("syn60530854") %>%
  read_csv() %>%
  mutate(
    cell_line = str_replace_all(cell_line, fixed("-"), "")
  )

deep_baseline_varstab_zscores <- syn("syn68500190") %>%
  read_csv()

deep_baseline_varstab_differences <- syn("syn68509714") %>%
  read_csv()

gr_raw <- syn("syn68499802") %>%
  read_csv() %>%
  mutate(neglog_GR50 = -log10(GR50)) %>%
  select(
    cell_line, agent, GR50, GRmax, GR_AOC, neglog_GR50
  )

```

Comparing gene expression of each cell line compared to the grand
mean of all cell lines at baseline (DMSO)

Because there are big batch effects between cell lines evidenced
by the PCA plots, only use the 2019_08 experiment here. That way
we also only have 24h time point instead of 24h and 48h.

## Correlating baseline expression and GR scores

Correlate GR50 vs log2FC of each cell line compared to the grand
mean. Hopefully we're able to identify some genes that are correlated

First have to compute the mean vst count for DGE data, where we have more than
one replicate per cell line.

```{r}
dge_counts_baseline_mean <- dge_counts_combined %>%
  filter(
    type == "varstab"
  ) %>%
  pivot_longer(
    cols = -c(ensembl_gene_id, type, hgnc_symbol, gene_name),
    names_to = "sample_id",
    values_to = "count"
  ) %>%
  power_inner_join(
    dge_meta %>%
      filter(
        experiment == "cdk_467_2019_08",
        agent == "control",
      ) %>%
      select(cell_line, sample_id),
    by = "sample_id",
    check = check_specs(
      duplicate_keys_right = "warn"
    )
  ) %>%
  group_by(cell_line, ensembl_gene_id, hgnc_symbol, gene_name) %>%
  summarize(
    count_sd = sd(count),
    count = mean(count),
    .groups = "drop"
  )

```

and figure out which genes are eligible

```{r}
cor_eligible_genes <- bind_rows(
  deep = counts_baseline %>%
    filter(type == "raw"),
  dge = dge_counts_combined %>%
    filter(type == "raw") %>%
    select(
      ensembl_gene_id,
      any_of(
        dge_meta %>%
          filter(
            experiment == "cdk_467_2019_08",
            agent == "control",
          ) %>%
          pull(sample_id)
      )
    ) %>%
    pivot_longer(
      -c(ensembl_gene_id),
      names_to = "sample_id",
      values_to = "count"
    ),
  .id = "assay"
) %>%
  group_by(assay, ensembl_gene_id) %>%
  summarize(
    n_over_10 = sum(count > 10),
    .groups = "drop"
  ) %>%
  filter(n_over_10 > 3)


gr_long <- gr_raw %>%
  pivot_longer(
    cols = -c(cell_line, agent),
    names_to = "gr_metric",
    values_to = "gr_value"
  )
```

```{r}
library(IHW)

compute_correlation <- function(gr_mat, lfc_mat, ...) {
  mean_expr <- rowMeans(lfc_mat, na.rm = TRUE) %>%
    enframe(name = "ensembl_gene_id", value = "mean")
  raw_cor <- psych::corr.test(
    t(lfc_mat),
    t(gr_mat),
    adjust = "fdr",
    ci = FALSE,
    ...
  )
  cor_df <- raw_cor[c("r", "p", "p.adj")] %>%
    imap(
      ~as_tibble(.x, rownames = "ensembl_gene_id") %>%
        pivot_longer(
          cols = -ensembl_gene_id,
          names_to = "gr_metric",
          values_to = .y
        )
    ) %>%
    purrr::reduce(
      left_join,
      by = c("ensembl_gene_id", "gr_metric")
    ) %>%
    power_inner_join(
      mean_expr,
      by = "ensembl_gene_id",
      check = check_specs(
        unmatched_keys_right = "warn",
        unmatched_keys_left = "warn",
        duplicate_keys_right = "warn"
      )
    ) %>%
    # left_join(
    #   ensembl_gtf %>%
    #     select(ensembl_gene_id, hgnc_symbol, gene_name),
    #   by = "ensembl_gene_id"
    # ) %>%
    group_by(gr_metric) %>%
    mutate(
      p.adj = p.adjust(
        p, method = "BH"
      ),
      p.adj_ihw = ihw(
        p,
        mean,
        alpha = 0.05
      ) %>%
        adj_pvalues()
    ) %>%
    ungroup()
  cor_df
}

baseline_gr_cor_inputs <- tribble(
  ~assay, ~expression_metric, ~expression_data,
  "dge", "log2FoldChange", deseq_res_baseline_dge %>%
    transmute(
      cell_line,
      ensembl_gene_id,
      expression_value = log2FoldChange
    ),
  "dge", "signed_p", deseq_res_baseline_dge %>%
    transmute(
      cell_line,
      ensembl_gene_id,
      expression_value = -log10(
        if_else(
          pvalue == 0,
          .Machine$double.xmin,
          replace_na(pvalue, 1)
        )
      ) * sign(log2FoldChange)
    ),
  "dge", "vst", dge_counts_baseline_mean %>%
    select(cell_line, ensembl_gene_id, expression_value = count),
  "deep", "vst", counts_baseline %>%
    filter(type == "varstab") %>%
    select(
      cell_line,
      ensembl_gene_id,
      expression_value = count
    )
) %>%
  crossing(
    cor_metric = c("spearman")
  ) %>%
  inner_join(
    cor_eligible_genes %>%
      group_nest(assay, .key = "eligible_genes"),
    by = "assay"
  )

qsave(
  baseline_gr_cor_inputs,
  file.path("data", "baseline_gr_cor_inputs.qs")
)

gr_mats <- gr_raw %>%
  filter(agent != "Staurosporine") %>%
  select(agent, cell_line, GR50, GRmax, GR_AOC, neglog_GR50) %>%
  split(.$agent) %>%
  map(
    \(x) select(x, -agent) %>%
      column_to_rownames("cell_line") %>%
      as.matrix() %>%
      t()
  )

baseline_gr_cor_res_raw <- baseline_gr_cor_inputs %>%
  # filter(expression_metric == "vst", assay == "deep") %>%
  mutate(
    cor_res = pmap(
      pick(everything()),
      \(cor_metric, expression_data, eligible_genes, expression_metric, ...) {
        message(expression_metric, "...")
        expr_mat <- expression_data %>%
          pivot_wider(
            names_from = cell_line,
            values_from = expression_value
          ) %>%
          column_to_rownames("ensembl_gene_id") %>%
          as.matrix()
        # browser()
        map(
          gr_mats,
          \(x) {
            common_cell_lines <- intersect(
              colnames(x),
              colnames(expr_mat)
            )
            compute_correlation(
              x[, common_cell_lines],
              expr_mat[eligible_genes$ensembl_gene_id, common_cell_lines],
              method = cor_metric
            )
          }
        )
      }
    )
  )

baseline_gr_cor_res <- baseline_gr_cor_res_raw %>%
  transmute(
    assay, expression_metric,
    cor_metric,
    cor_res = map(
      cor_res,
      \(x) bind_rows(x, .id = "agent")
    )
  ) %>%
  unnest(cor_res)

write_csv(
  baseline_gr_cor_res,
  file.path("data", "baseline_gr_cor_res.csv.gz")
)

```

## Baseline vs GR plots

```{r}
compound_meta <- tribble(
  ~agent, ~target, ~cdk_target, ~inhibition_mode_full,
  "Adavosertib", "Wee1", "CDK1", "ATP-competitive",
  "AZD5363", "ATK1/2/3", "CDK4/6", "ATP-competitive",
  "BAY1895344", "ATR", "CDK1/2", "ATP-competitive",
  "CT7001", "CDK7", "CDK7", "ATP-competitive",
  "JQ1", "BRD4", "none", "Acetyl-lysine competitive",
  "JWZ", "CDK7", "CDK7", "PROTAC degrader",
  "LDC4297", "CDK7", "CDK7", "ATP-competitive",
  "LY3405105", "CDK7", "CDK7", "Covalent",
  "Olaparib", "PARP", "none", "Competitive",
  "Prexasertib", "CHK1", "CDK1", "ATP-competitive",
  "RP-6306", "Wee1", "CDK1", "ATP-competitive",
  "SY-5609", "CDK7", "CDK7", "ATP-competitive",
  "YKL-5-124", "CDK7", "CDK7", "Covalent",
) %>%
  mutate(
    inhibition_mode = case_when(
      grepl("competitive", inhibition_mode_full, ignore.case = TRUE) ~ "Competitive",
      grepl("Covalent", inhibition_mode_full) ~ "Covalent",
      grepl("degrader", inhibition_mode_full) ~ "Degrader",
      TRUE ~ "Other"
    )
  )

```

```{r}
baseline_gr_cor_df_top_10_data <- baseline_gr_cor_res %>%
  arrange(p) %>%
  group_by(agent, gr_metric, cor_metric, expression_metric, assay) %>%
  slice_head(n = 12) %>%
  ungroup() %>%
  left_join(
    baseline_gr_cor_inputs %>%
      select(-eligible_genes) %>%
      unnest(expression_data),
    by = c("assay", "expression_metric", "cor_metric", "ensembl_gene_id"),
    relationship = "many-to-many"
  ) %>%
  powerjoin::power_inner_join(
    gr_long,
    by = c("agent", "cell_line", "gr_metric"),
    check = powerjoin::check_specs(
      duplicate_keys_right = "warn"
    )
  )

baseline_gr_cor_res %>% filter(agent == "AZD5363", cell_line == "OCE2")

ps <- baseline_deep_varstab_cor_df_top_10_data %>%
  group_nest(agent, gr_metric) %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
        data %>%
          arrange(p) %>%
          mutate(across(gene_name, fct_inorder)),
        aes(count_rank, gr_value, color = cell_line)
      ) +
        geom_point() +
        geom_smooth(color = "black", method = "lm") +
        geom_text(
          aes(
            label = paste0("r = ", signif(r, 2), " p = ", signif(p, 2))
          ),
          data = \(x) distinct(x, gene_name, r, p),
          x = -Inf, y = Inf, hjust = 0, vjust = 1,
          inherit.aes = FALSE
        ) +
        facet_wrap(~gene_name, scales = "free") +
        labs(
          y = gr_metric,
          x = "Variance stabilized count rank",
          title = gr_metric
        ) +
        theme_bw() +
        theme(strip.background = element_blank())
    )
  )

pwalk(
  ps,
  \(gr_metric, p, agent, ...) {
    ggsave(
      file.path("plots_cdk7", paste0("deep_baseline_vs_gr_top10_spearman_correlated_ranks_", agent, "_", gr_metric, ".pdf")),
      p, width = 12, height = 8
    )
  }
)


ps <- baseline_deep_varstab_cor_df_top_10_data %>%
  group_nest(gr_metric) %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
        data,
        aes(count, gr_value, color = cell_line)
      ) +
        geom_point() +
        geom_smooth(color = "black", method = "lm") +
        geom_text(
          aes(
            label = paste0("r = ", signif(r, 2), " p = ", signif(p, 2))
          ),
          data = \(x) distinct(x, gene_name, r, p),
          x = -Inf, y = Inf, hjust = 0, vjust = 1,
          inherit.aes = FALSE
        ) +
        facet_wrap(~gene_name, scales = "free") +
        labs(
          y = gr_metric,
          x = "Variance stabilized count",
          title = gr_metric
        ) +
        theme_bw() +
        theme(strip.background = element_blank())
    )
  )

pwalk(
  ps,
  \(gr_metric, p, ...) {
    ggsave(
      file.path("plots_cdk7", paste0("deep_baseline_vs_gr_top10_spearman_correlated_", gr_metric, ".pdf")),
      p, width = 12, height = 8
    )
  }
)
```


### GR correlation heatmaps

```{r}
library(ComplexHeatmap)
library(seriation)

createAnnotation <- function(df, colors = list(), which = "column") {
  # Load necessary libraries
  color_maps <- list()

  # Iterate over each column in the dataframe
  for (col_name in names(df)) {
    # Check if the column is numeric
    if (is.numeric(df[[col_name]])) {
      # Create a color mapping function for numeric columns
      if (is.null(colors[[col_name]]))
        colors[[col_name]] <- c("white", "red")
      if (is.function(colors[[col_name]])) {
        col_fun <- colors[[col_name]]
      } else if (is.character(colors[[col_name]])) {
        n <- length(colors[[col_name]])
        if (n == 1) {
          col_fun <- circlize::colorRamp2(
            seq(from = min(df[[col_name]]), to = max(df[[col_name]]), length.out = 100),
            viridis(100, option = colors[[col_name]])
          )
        } else {
          col_fun <- circlize::colorRamp2(
            seq(from = min(df[[col_name]]), to = max(df[[col_name]]), length.out = n),
            colors[[col_name]]
          )
        }
      } else
        stop("Dont know how to handle colors for column ", col_name)
      color_maps[[col_name]] <- col_fun
    } else {
      if (is.character(colors[[col_name]])) {
        color_maps[[col_name]] <- colors[[col_name]]
        next
      }
      col_fun <- colors[[col_name]] %||% rainbow
      # Create a named vector of colors for categorical columns
      values <- df[[col_name]]
      unique_values <- if (is.factor(values)) levels(values) else unique(df[[col_name]])
      named_colors <- setNames(col_fun(length(unique_values)), unique_values)
      color_maps[[col_name]] <- named_colors
    }
  }

  # Combine all annotations
  combined_annotations <- HeatmapAnnotation(df = df, col = color_maps, which = which)

  return(combined_annotations)
}

cluster_fun_eucl <- function(mat, features_in_rows = TRUE) {
  if (!features_in_rows) {
    mat <- t(mat)
  }
  # mat_imp <- impute.knn(
  #   mat, rng.seed = 42, colmax = 90
  # )[["data"]]
  # if (!features_in_rows) {
  #   mat_imp <- t(mat_imp)
  #   mat <- t(mat)
  # }
  # browser()
  dist_mat <- dist(mat)
  # dist_mat <- as.dist(mat)
  clust <- hclust(dist_mat, method = "complete")
  reorder(clust, dist_mat, method = "OLO")
}

baseline_cor_df_hm_data <- baseline_gr_cor_res %>%
  filter(gr_metric != "neglog_GR50") %>%
  semi_join(
    group_by(., assay, expression_metric, gr_metric, cor_metric, agent) %>%
      drop_na(r) %>%
      arrange(abs(r)) %>%
      slice_tail(n = 30),
    by = c("assay", "expression_metric", "cor_metric", "gr_metric", "ensembl_gene_id")
  )

paletteer_shim <- function(palette_name) {
  \(n) {
    paletteer::paletteer_d(palette_name)[seq_len(n)]
  }
}

baseline_cor_df_hms_all_agents <- baseline_cor_df_hm_data %>%
  group_nest(assay, expression_metric, gr_metric, cor_metric) %>%
  mutate(
    mat = map(
      data,
      \(x) select(
        x, agent, ensembl_gene_id, r
      ) %>%
        pivot_wider(values_from = r, names_from = agent) %>%
        column_to_rownames("ensembl_gene_id") %>%
        as.matrix()
    ),
    hm = pmap(
      list(mat, gr_metric),
      \(x, y) {
        col_fun = circlize::colorRamp2(
          {
            qs <- quantile(x, c(.05, .95), na.rm = TRUE)
            absmax <- max(abs(qs))
            seq(-absmax, absmax, length.out = 11)
          },
          rev(RColorBrewer::brewer.pal(11, "RdBu"))
        )
        # browser()
        col_meta <- compound_meta %>%
          select(agent, target, cdk_target, inhibition_mode) %>%
          slice(match(colnames(x), agent)) %>%
          column_to_rownames("agent")
        Heatmap(
          x,
          name = paste("Correlation", y),
          col = col_fun,
          cluster_rows = cluster_fun_eucl,
          cluster_columns = cluster_fun_eucl,
          show_row_names = FALSE,
          bottom_annotation = createAnnotation(
            col_meta,
            colors = list(
              cdk_target = paletteer_shim("rockthemes::californication"),
              inhibition_mode = \(n) ggokabeito::palette_okabe_ito(seq_len(n))
            )
          )
        )
      }
    )
  )

pwalk(
  baseline_cor_df_hms_all_agents,
  \(gr_metric, hm, data, assay, expression_metric, cor_metric, ...) {
    withr::with_pdf(
      file.path("plots_cdk7", paste0("baseline_cor_hm_", assay, "_", expression_metric, "_", cor_metric, "_", gr_metric, ".pdf")),
      {
        draw(hm)
      },
    width = 7, height = 8
    )
  }
)
```

## GSEA on baseline GR correlation

```{r}
library(fgsea)
library(msigdbr)

database_abbreviations <- c(
  "REACTOME" = "R",
  "HALLMARK" = "H",
  "PID" = "P",
  "KEGG_MEDICUS" = "K",
  "GOBP" = "BP",
  "GOMF" = "MF",
  "GOCC" = "CC"
)

abbreviate_prefix <- function(strings, abbr_map = database_abbreviations) {
  # Create regex pattern from abbreviation names
  pattern <- paste0("^(", paste(names(abbr_map), collapse = "|"), ")_")

  # Replace matching prefixes with abbreviations
  str_replace(strings, pattern, function(x) {
    prefix <- str_remove(x, "_$")
    paste0(abbr_map[prefix], "_")
  })
}

all_gene_sets <- msigdbr(species = "Homo sapiens")
selected_gene_sets <- all_gene_sets %>%
  filter(
    gs_collection %in% c("H") |
      gs_subcollection %in% c(
        "CP:REACTOME",
        "CP:KEGG_MEDICUS"
      ),
    !str_detect(gs_name, coll("MEDICUS_PATHOGEN")),
    !str_detect(gs_name, coll("MEDICUS_VARIANT"))
  ) %>%
  mutate(
    database = str_extract(gs_name, paste0("^(", paste(names(database_abbreviations), collapse = "|"), ")")),
    pathway = abbreviate_prefix(gs_name)
  )

selected_gene_sets_list <- selected_gene_sets %>%
  group_nest(gs_collection, gs_subcollection, gs_name, database, pathway) %>%
  mutate(
    gs = map(
      data,
      \(x) unique(x$ensembl_gene)
    ) %>%
      set_names(pathway)
  )

baseline_gr_cor_fgsea_input <- baseline_gr_cor_res %>%
  drop_na(r) %>%
  group_nest(agent, assay, expression_metric, cor_metric, gr_metric) %>%
  mutate(
    de = map(
      data,
      \(x) set_names(x$r, x$ensembl_gene_id)
    )
  )

baseline_gr_cor_fgsea_res_raw <- baseline_gr_cor_fgsea_input %>%
  mutate(
    res = map(
      de,
      \(x) withr::with_seed(42, quietly(fgseaMultilevel)(
        selected_gene_sets_list$gs,
        x
      ))
    )
  )


baseline_gr_cor_fgsea_res_long <- baseline_gr_cor_fgsea_res_raw %>%
  select(agent, assay, expression_metric, cor_metric, gr_metric, res) %>%
  unnest_wider(res, transform = list(warnings = \(x) paste(x, collapse = ""))) %>%
  unnest(result) %>%
  mutate(
    signed_p = -sign(NES) * log10(padj),
    p_cut = cut(padj, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("**", "*", ""))
  )

write_csv(
  baseline_gr_cor_fgsea_res_long,
  file.path("data", "baseline_gr_cor_fgsea_res.csv.gz")
)
```


### Barplots

```{r}
fgsea_res_long_selected <- fgsea_res_long %>%
  filter(
    pathway %in% {
      group_by(., pathway) %>%
        summarize(n_sig = sum(padj < 0.01)) %>%
        filter(n_sig > 0) %>%
        pull(pathway)
    }
  ) %>%
  mutate(
    size_leading_edge = map_int(
      leadingEdge, length
    )
  ) %>%
  select(
    gr_metric, pathway,
    signed_p, pval, padj, p_cut, ES, NES, size_leading_edge, size,
    leadingEdge
  )


better_contrast <- function(
  colors1, colors2, threshold = 0
) {
  # browser()
  farver::compare_colour(
    farver::decode_colour(colors1),
    farver::decode_colour(colors2),
    from_space = "rgb",
    method = "cie2000"
  ) %>% {
      .[, 1] < (1 + threshold) * .[, 2]
    } %>%
    if_else(colors2[2], colors2[1])
}

ps <- fgsea_res_long_selected %>%
  filter(padj < .05) %>%
  group_nest(
    gr_metric
  ) %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
        data %>%
          arrange(NES) %>%
          mutate(
            pathway = fct_inorder(str_trunc(pathway, 45))
          ) %>%
          mutate(
            direction = if_else(NES > 0, "positive", "negative")
          ),
        aes(NES, pathway, fill = p_cut)
      ) +
        geom_col() +
        facet_wrap(~direction, scales = "free") +
        # geom_text(
        #   aes(
        #     x = x_pos,
        #     y = pathway,
        #     label = pathway,
        #     hjust = hjust
        #   ),
        #   data = \(x) mutate(
        #     x,
        #     x_pos = if_else(NES > 0, -.1, .1),
        #     hjust = if_else(NES > 0, 1, 0)
        #   ),
        #   inherit.aes = FALSE
        # ) +
        scale_fill_viridis_d(direction = -1) +
        labs(
          x = "Normalized Enrichment Score",
          y = NULL,
          title = gr_metric,
          fill = "P-value"
        )
    )
  )

pwalk(
  ps,
  \(gr_metric, p, data, ...) {
    ggsave(
      file.path("plots_cdk7", paste0("fgsea_bars_baseline_deep_", gr_metric, ".pdf")),
      p, width = 16, height = 10
    )
  }
)
```

### Heatmaps

```{r}
fgsea_res_long_selected_hm_data <- fgsea_res_long %>%
  semi_join(
    group_by(., gr_metric, agent) %>%
      arrange(padj) %>%
      slice_head(n = 20),
    by = c("gr_metric", "pathway")
  ) %>%
  mutate(
    size_leading_edge = map_int(
      leadingEdge, length
    )
  ) %>%
  transmute(
    agent, gr_metric, pathway,
    pathway_short = str_trunc(
      pathway, {
        # Find minimum length that preserves uniqueness
        min_length <- 30
        max_length <- 80
        unique_pathways <- unique(pathway)
        for (len in min_length:max_length) {
          truncated <- str_trunc(unique_pathways, len)
          if (length(unique(truncated)) == length(unique(unique_pathways))) {
            break
          }
        }
        len
      }
    ),
    signed_p, pval, padj, p_cut, ES, NES, size_leading_edge, size
  )

fgsea_res_long_selected_hms <- fgsea_res_long_selected_hm_data %>%
  group_nest(gr_metric) %>%
  mutate(
    mat = map(
      data,
      \(x) select(
        x, agent, pathway_short, signed_p
      ) %>%
        pivot_wider(values_from = signed_p, names_from = agent) %>%
        column_to_rownames("pathway_short") %>%
        as.matrix()
    ),
    hm = pmap(
      list(mat, gr_metric),
      \(x, y) {
        qs <- quantile(x, c(.05, .95), na.rm = TRUE)
        absmax <- max(abs(qs))
        col_fun <- circlize::colorRamp2(
          seq(-absmax, absmax, length.out = 11),
          rev(RColorBrewer::brewer.pal(11, "RdBu"))
        )
        col_meta <- compound_meta %>%
          select(agent, target, cdk_target, inhibition_mode) %>%
          slice(match(colnames(x), agent)) %>%
          column_to_rownames("agent")
        Heatmap(
          x,
          name = paste("NES", y),
          col = col_fun,
          cluster_rows = cluster_fun_eucl,
          cluster_columns = cluster_fun_eucl,
          show_row_names = TRUE,
          show_column_names = TRUE,
          row_names_max_width = max_text_width(
            rownames(x),
            gp = gpar(fontsize = 10)
          ),
          bottom_annotation = createAnnotation(
            col_meta,
            colors = list(
              cdk_target = paletteer_shim("rockthemes::californication"),
              inhibition_mode = \(n) ggokabeito::palette_okabe_ito(seq_len(n))
            )
          )
        )
      }
    )
  )

pwalk(
  fgsea_res_long_selected_hms,
  \(gr_metric, hm, data, ...) {
    withr::with_pdf(
      file.path("plots_cdk7", paste0("baseline_deep_varstab_rank_cor_fgsea_hm_", gr_metric, ".pdf")),
      draw(hm),
      width = 10, height = 18
    )
  }
)
```


Lots and lots of ties (>90% often), doing overrep analysis on top n genes instead

## Using fora instead

```{r}
baseline_gr_cor_fora_input <- baseline_gr_cor_res %>%
  drop_na(r) %>%
  mutate(
    direction = if_else(r > 0, "positive", "negative")
  ) %>%
  arrange(p) %>%
  group_by(
    agent, assay, expression_metric, cor_metric, gr_metric, direction
  ) %>%
  summarize(
    de = list(head(ensembl_gene_id[p < .05], n = 300)),
    universe = list(ensembl_gene_id),
    .groups = "drop"
  )


plan(multicore, workers = 6)
baseline_gr_cor_fora_res_raw <- baseline_gr_cor_fora_input %>%
  mutate(
    res = map2(
      de, universe,
      \(x, y) withr::with_seed(42, fora(
        selected_gene_sets_list$gs,
        x,
        universe = y
      ))
    ),
    res_reduced = future_pmap(
      list(res, de, universe),
      \(res, de, universe) {
        collapsePathwaysORA(
          filter(res, padj < .05),
          selected_gene_sets_list$gs,
          de,
          universe
        )
      },
      .progress = TRUE
    )
  )

baseline_gr_cor_fora_res_long <- baseline_gr_cor_fora_res_raw %>%
  transmute(
    agent, assay, expression_metric, cor_metric, gr_metric, direction,
    res = map2(
      res, res_reduced,
      \(full, reduced) {
        mutate(
          full,
          is_main = pathway %in% reduced$mainPathways,
          parent_pathway = reduced$parentPathways[pathway]
        )
      }
    )
  ) %>%
  unnest(res) %>%
  mutate(
    signed_p = if_else(direction == "up", 1, -1) * log10(padj),
    overlapNames = map2(
      overlapGenes, assay,
      \(x, y) gene_id_to_name_map[[y]][x]
    )
  )

write_csv(
  baseline_gr_cor_fora_res_long,
  file.path("data", "baseline_gr_cor_fora_res.csv.gz")
)
```

### Excel sheet

```{r}
significant_style <- createStyle(fontColour = "red")

baseline_gr_cor_fora_excel <- createWorkbook()

baseline_gr_cor_fora_res_long %>%
  select(-c(overlapGenes)) %>%
  filter(
    gr_metric != "neglog_GR50",
    assay == "deep",
    expression_metric == "vst",
    cor_metric == "spearman"
  ) %>%
  arrange(padj, pval) %>%
  group_by(agent) %>%
  group_walk(
    \(df, g) {
      addWorksheet(
        baseline_gr_cor_fora_excel,
        g$agent
      )
      writeDataTable(
        baseline_gr_cor_fora_excel,
        sheet = g$agent,
        x = df,
        tableStyle = "TableStyleLight8"
      )
      setColWidths(
        baseline_gr_cor_fora_excel, g$agent,
        cols = seq_len(ncol(df)),
        widths = "auto"
      )
      walk(
        which(names(df) == "padj"),
        \(i) {
          data_rows <- nrow(df) + 1L  # +1 for header
          de_col_data <- df[[i]]
          significant_rows <- which(!is.na(de_col_data) & de_col_data < 0.05) + 1L  # +1 for header
          if (length(significant_rows) > 0) {
            addStyle(
              wb = baseline_gr_cor_fora_excel,
              sheet = g$agent,
              style = significant_style,
              rows = significant_rows,
              cols = i,
              gridExpand = FALSE
            )
          }
        }
      )
    }
  )

saveWorkbook(
  baseline_gr_cor_fora_excel,
  file.path("plots_cdk7", "baseline_gr_cor_fora_overrepresentation.xlsx"),
  overwrite = TRUE
)
```


### Dot plots

```{r}
baseline_gr_cor_fora_dotplots <- baseline_gr_cor_fora_res_long %>%
  filter(gr_metric != "neglog_GR50") %>%
  mutate(
    log2_foldEnrichment = log2(if_else(foldEnrichment == 0, .5, foldEnrichment)),
    directed_enrichment = pmax(
      log2_foldEnrichment,
      0
    ) * if_else(direction == "positive", 1, -1),
    p_cut = cut(
      padj,
      breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
      labels = c("<0.001", "<0.01", "<0.05", "ns")
    )
  ) %>%
  group_nest(
    agent, assay, expression_metric, cor_metric
  ) %>%
  mutate(
    res = map(
      data,
      \(x) {
        # browser()
        # d <- semi_join(
        #   x,
        #   group_by(x, contrast) %>%
        #     filter(padj < y),
        #   by = "pathway"
        # )
        d <- semi_join(
          x,
          group_by(x, gr_metric, direction) %>%
            filter(padj < .05) %>%
            arrange(pval) %>%
            slice_head(n = 20),
          by = "pathway"
        )
        if (nrow(d) <= 6) {
          return(NULL)
        }
        # d <- semi_join(
        #   x,
        #   group_by(x, pathway) %>%
        #     filter(padj < .05, is_main) %>%
        #     summarize(groups = paste(sort(unique(contrast)), collapse = "_"), p_mean = 1 / sum(1 / padj, na.rm = TRUE)) %>%
        #     group_by(groups) %>%
        #     arrange(p_mean) %>%
        #     slice_head(n = 5),
        #     # bind_rows(tibble(pathway = "P_P38_MK2_PATHWAY")),
        #   by = "pathway"
        # )
        dwide <- d %>%
          arrange(pathway) %>%
          pivot_wider(
            id_cols = pathway,
            names_from = c(gr_metric, direction),
            values_from = directed_enrichment
          )
        dmat <- dwide %>%
          column_to_rownames("pathway") %>%
          as.matrix() %>%
          dist()
        # browser()
        clust <- hclust(dmat, method = "average") %>%
          reorder(dmat, method = "OLO")
        row_labels <- clust$labels[clust$order] %>% {
          set_names(
            str_replace_all(., coll("_"), " ") %>%
              str_sub(start = 2),
            .
          )
        }
        # sim_hm <- msigdbr_of_interest_sim_trans[
        #   rev(clust$labels[clust$order]), rev(clust$labels[clust$order])
        # ] %>%
        #   pheatmap(
        #     cluster_rows = FALSE,
        #     cluster_cols = FALSE
        #   )
        est_perc <- quantile(d$directed_enrichment, c(.05, .95))
        abs_max_est_perc <- max(abs(est_perc))
        # browser()
        p <- ggplot(
          d %>%
            mutate(
              pathway = factor(pathway, levels = clust$labels[clust$order])
            ),
          aes(
            x = gr_metric,
            y = fct_rev(pathway),
            color = directed_enrichment,
            size = p_cut,
            shape = direction
          )
        ) +
          geom_text(
            aes(
              label = if_else(direction == "positive", "◗", "◖"),
              hjust = if_else(direction == "positive", .5, .5)
            ),
            vjust = .5
          ) +
          guides(
            size = guide_legend(override.aes = list(shape = "◖"))
          ) +
          paletteer::scale_color_paletteer_c(
            palette = "ggthemes::Orange-Blue-White Diverging",
            limits = c(-abs_max_est_perc, abs_max_est_perc),
            oob = scales::squish,
            direction = -1
          ) +
          scale_size_manual(
            values = rev(c("<0.001" = 8, "<0.01" = 6, "<0.05" = 4, "ns" = 2))
          ) +
          scale_x_discrete(position = "top") +
          scale_y_discrete(
            breaks = names(row_labels),
            labels = unname(row_labels)
          ) +
          theme(
            axis.text.x = element_text(angle = 45, hjust = 0),
            panel.grid.major.x = element_blank()
          ) +
          labs(
            x = NULL,
            y = NULL,
            size = "FDR",
            color = "log2 fold\nenrichment"
          )
        list(
          p = p,
          clust_data = d
          # sim_hm = sim_hm
        ) %>%
          map(list)
      }
    )
  ) %>%
  select(-data) %>%
  unnest_wider(
    res
  )

pwalk(
  baseline_gr_cor_fora_dotplots,
  \(agent, assay, expression_metric, cor_metric, p, ...) {
    if (is.null(p[[1]])) {
      return(NULL)
    }
    ggsave(
      file.path(
        "plots_cdk7", paste0("fora_baseline_gr_cor_dotplot_", agent, "_", assay, "_", expression_metric, "_", cor_metric, ".pdf")
      ), p[[1]], width = 10, height = 8, dev = cairo_pdf
    )
  }
)

```

Broken down by metric instead of agent


```{r}
baseline_gr_cor_fora_dotplots <- baseline_gr_cor_fora_res_long %>%
  filter(
    gr_metric != "neglog_GR50",
    assay == "deep", expression_metric == "vst", cor_metric == "spearman"
  ) %>%
  mutate(
    log2_foldEnrichment = log2(if_else(foldEnrichment == 0, .5, foldEnrichment)),
    directed_enrichment = pmax(
      log2_foldEnrichment,
      0
    ) * if_else(direction == "positive", 1, -1),
    p_cut = cut(
      padj,
      breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
      labels = c("<0.001", "<0.01", "<0.05", "ns")
    )
  ) %>%
  group_nest(
    gr_metric
  ) %>%
  mutate(
    res = map(
      data,
      \(x) {
        # browser()
        # d <- semi_join(
        #   x,
        #   group_by(x, contrast) %>%
        #     filter(padj < y),
        #   by = "pathway"
        # )
        # d <- semi_join(
        #   x,
        #   group_by(x, gr_metric, direction) %>%
        #     filter(padj < .05) %>%
        #     arrange(pval) %>%
        #     slice_head(n = 20),
        #   by = "pathway"
        # )
        d <- semi_join(
          x,
          group_by(x, pathway, direction) %>%
            filter(padj < .05) %>%
            summarize(
              groups = paste(sort(unique(paste(agent, direction, sep = "_"))), collapse = "_"),
              p_mean = 1 / sum(1 / padj, na.rm = TRUE),
              .groups = "drop"
            ) %>%
            group_by(groups) %>%
            arrange(p_mean) %>%
            slice_head(n = 10),
            # bind_rows(tibble(pathway = "P_P38_MK2_PATHWAY")),
          by = "pathway"
        )
        if (nrow(d) <= 6) {
          return(NULL)
        }
        dwide <- d %>%
          arrange(pathway) %>%
          pivot_wider(
            id_cols = pathway,
            names_from = c(agent, direction),
            values_from = directed_enrichment
          )
        dmat <- dwide %>%
          column_to_rownames("pathway") %>%
          as.matrix() %>%
          dist()
        # browser()
        clust <- hclust(dmat, method = "average") %>%
          reorder(dmat, method = "OLO")
        row_labels <- clust$labels[clust$order] %>% {
          set_names(
            str_replace_all(., coll("_"), " ") %>%
              str_sub(start = 2),
            .
          )
        }
        # sim_hm <- msigdbr_of_interest_sim_trans[
        #   rev(clust$labels[clust$order]), rev(clust$labels[clust$order])
        # ] %>%
        #   pheatmap(
        #     cluster_rows = FALSE,
        #     cluster_cols = FALSE
        #   )
        est_perc <- quantile(d$directed_enrichment, c(.05, .95))
        abs_max_est_perc <- max(abs(est_perc))
        # browser()
        p <- ggplot(
          d %>%
            mutate(
              pathway = factor(pathway, levels = clust$labels[clust$order])
            ),
          aes(
            x = agent,
            y = fct_rev(pathway),
            color = directed_enrichment,
            size = p_cut,
            shape = direction
          )
        ) +
          geom_text(
            aes(
              label = if_else(direction == "positive", "◗", "◖"),
              hjust = if_else(direction == "positive", .5, .5)
            ),
            vjust = .5
          ) +
          guides(
            size = guide_legend(override.aes = list(shape = "◖"))
          ) +
          paletteer::scale_color_paletteer_c(
            palette = "ggthemes::Orange-Blue-White Diverging",
            limits = c(-abs_max_est_perc, abs_max_est_perc),
            oob = scales::squish,
            direction = -1
          ) +
          scale_size_manual(
            values = rev(c("<0.001" = 8, "<0.01" = 6, "<0.05" = 4, "ns" = 2))
          ) +
          scale_x_discrete(position = "top") +
          scale_y_discrete(
            breaks = names(row_labels),
            labels = unname(row_labels)
          ) +
          theme(
            axis.text.x = element_text(angle = 45, hjust = 0),
            panel.grid.major.x = element_blank()
          ) +
          labs(
            x = NULL,
            y = NULL,
            size = "FDR",
            color = "log2 fold\nenrichment"
          )
        list(
          p = p,
          clust_data = d
          # sim_hm = sim_hm
        ) %>%
          map(list)
      }
    )
  ) %>%
  select(-data) %>%
  unnest_wider(
    res
  )

pwalk(
  baseline_gr_cor_fora_dotplots,
  \(gr_metric, p, ...) {
    if (is.null(p[[1]])) {
      return(NULL)
    }
    ggsave(
      file.path(
        "plots_cdk7", paste0("fora_baseline_gr_cor_dotplot_by_metric_", gr_metric, ".pdf")
      ), p[[1]], width = 10, height = 8, dev = cairo_pdf
    )
  }
)

```

## INDRA queries

```{r}
library(indra.gsea)

baseline_deep_varstab_cor_df_indra_input <- baseline_deep_varstab_cor_df %>%
  mutate(
    direction = if_else(r > 0, "positive", "negative")
  ) %>%
  arrange(p) %>%
  filter(p < 0.05) %>%
  group_by(
    gr_metric, direction
  ) %>%
  slice_head(n = 100) %>%
  ungroup()

baseline_deep_varstab_cor_df_indra_res <- baseline_deep_varstab_cor_df_indra_input %>%
  group_by(gr_metric, direction) %>%
  summarize(
    res = list(
      indra_discrete_gsea(
        hgnc_symbol
      )
    ),
    .groups = "drop"
  )
baseline_deep_varstab_cor_df_indra_res$res[[5]]$`indra-downstream` %>%
  select(name, mlq, p, q)
baseline_deep_varstab_cor_df_indra_res$res[[5]]$`indra-upstream` %>%
  select(name, mlq, p, q)

baseline_deep_varstab_cor_df_indra_res$res[[6]]$`indra-downstream` %>%
  select(name, mlq, p, q)
baseline_deep_varstab_cor_df_indra_res$res[[6]]$`indra-upstream` %>%
  select(name, mlq, p, q)

baseline_deep_varstab_cor_df_indra_res$res[[7]]$`indra-downstream` %>%
  select(name, mlq, p, q)
baseline_deep_varstab_cor_df_indra_res$res[[7]]$`indra-upstream` %>%
  select(name, mlq, p, q)

baseline_deep_varstab_cor_df_indra_res$res[[8]]$`indra-downstream` %>%
  select(name, mlq, p, q)
baseline_deep_varstab_cor_df_indra_res$res[[8]]$`indra-upstream` %>%
  select(name, mlq, p, q)
```



```{r}

library(seriation)
cluster_df <- function(df, row_var, col_var, value_var, values_fill = 0) {
  # browser()
  mat <- df %>%
    select({{row_var}}, {{col_var}}, {{value_var}}) %>%
    pivot_wider(names_from = {{col_var}}, values_from = {{value_var}}, values_fill = values_fill) %>%
    column_to_rownames(rlang::as_name(rlang::enquo(row_var)))
  # browser()
  if (rlang::is_bare_numeric(pull(df, {{value_var}}))) {
    dist_rows <- dist(mat, method = "euclidian")
    dist_cols <- dist(t(mat), method = "euclidian")
  } else {
    # browser()
    dist_rows <- cluster::daisy(mat, metric = "gower")
    dist_cols <- t(mat) %>%
      as.data.frame() %>%
      mutate(across(everything(), \(x) factor(x, levels = levels(pull(df, {{value_var}}))))) %>%
      cluster::daisy(metric = "gower")
  }
  clust_rows <- hclust(dist_rows, method = "average") %>%
      reorder(dist_rows, method = "olo")
  clust_cols <- hclust(dist_cols, method = "average") %>%
      reorder(dist_cols, method = "olo")
  df %>%
    mutate(
      "{{row_var}}" := factor({{row_var}}, levels = clust_rows$labels[clust_rows$order]),
      "{{col_var}}" := factor({{col_var}}, levels = clust_cols$labels[clust_cols$order])
    )
}

p <- fgsea_res_p_long %>%
  filter(
    gr_metric != "GR50",
    str_starts(pathway, "REACTOME")
  ) %>%
  filter(
    pathway %in% {
      group_by(., pathway) %>%
        summarize(n_sig = sum(padj < 0.01)) %>%
        filter(n_sig > 0) %>%
        pull(pathway)
    }
  ) %>%
  cluster_df(
    pathway, gr_metric, NES
  ) %>%
  ggplot(
    aes(gr_metric, pathway, fill = NES)
  ) +
    geom_raster() +
    geom_text(
      aes(
        label = p_cut,
        color = after_scale(better_contrast(fill, c("white", "black"), 0))
      ),
      hjust = .5, vjust = .5
    ) +
    scico::scale_fill_scico(
      palette = "vik", midpoint = 0
    ) +
    theme_minimal()

```


### Deep count z-scores vs GR

```{r}

deep_baseline_nonzero_genes <- counts_baseline %>%
  filter(type == "normalized", count >= 10) %>%
  count(ensembl_gene_id) %>%
  filter(n >= 3)


deep_baseline_zscores_vs_gr <- bind_rows(
  varstab_differences = deep_baseline_varstab_differences,
  zscores = deep_baseline_varstab_zscores,
  varstab = counts_baseline %>%
    filter(type == "varstab") %>%
    select(ensembl_gene_id, sample_id, count) %>%
    pivot_wider(
      names_from = sample_id,
      values_from = count,
      values_fill = 0
    ),
  .id = "metric"
) %>%
  semi_join(
    deep_baseline_nonzero_genes,
    by = "ensembl_gene_id"
  ) %>%
  pivot_longer(
    -c(metric, ensembl_gene_id),
    names_to = "sample_id",
    values_to = "z_score"
  ) %>%
  power_inner_join(
    deep_meta %>%
      select(sample_id, cell_line),
    by = "sample_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  power_inner_join(
    gr_long,
    by = c("cell_line"),
    check = check_specs(
      unmatched_keys_left = "info",
      unmatched_keys_right = "info"
    )
  )

deep_zscores_vs_gr_glmnet_raw <- deep_baseline_zscores_vs_gr %>%
  group_by(gr_metric, metric) %>%
  summarize(
    res = {
      mat <- pivot_wider(
        pick(cell_line, ensembl_gene_id, z_score, gr_value),
        names_from = ensembl_gene_id,
        values_from = z_score
      ) %>%
        column_to_rownames("cell_line") %>%
        as.matrix()
      # browser()
      glmnet::cv.glmnet(
        x = mat[, colnames(mat) != "gr_value"],
        y = mat[, "gr_value"],
        alpha = 0.5, # Elastic net mixing parameter
        nfolds = 10,
        standardize = TRUE,
        intercept = TRUE,
        family = "gaussian"
      ) %>%
        list()
    },
    .groups = "drop"
  )

deep_zscores_vs_gr_glmnet_coefs <- deep_zscores_vs_gr_glmnet_raw %>%
  mutate(
    res = map(
      res,
      \(x) {
        coef(x, s = "lambda.min") %>%
          as.matrix() %>%
          as_tibble(rownames = "ensembl_gene_id") %>%
          filter(ensembl_gene_id != "(Intercept)", s0 != 0)
      }
    )
  ) %>%
  unnest(res) %>%
  power_inner_join(
    ensembl_gtf %>%
      distinct(ensembl_gene_id, gene_name),
    by = "ensembl_gene_id",
    check = check_specs(
      duplicate_keys_right = "warn",
      unmatched_keys_left = "warn"
    )
  )

ps <- deep_baseline_zscores_vs_gr %>%
  power_inner_join(
    deep_zscores_vs_gr_glmnet_coefs,
    by = c("gr_metric", "metric", "ensembl_gene_id"),
    check = check_specs(
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  group_by(
    gr_metric, metric
  ) %>%
  summarize(
    p = list(
      ggplot(
        pick(everything()),
        aes(
          x = z_score, y = gr_value
        )
      ) +
      geom_point() +
      facet_wrap(vars(gene_name)) +
      theme_bw() +
      labs(
        title = paste(cur_group()$gr_metric, cur_group()$metric, sep = " - "),
      )
    ),
    .groups = "drop"
  )
ps$p

```

### Simple correlation

```{r}
deep_baseline_zscores_vs_gr_cor_raw <- deep_baseline_zscores_vs_gr %>%
  group_by(gr_metric, metric, ensembl_gene_id) %>%
  summarize(
    res = list(
      cor.test(
        deep_baseline_zscores_vs_gr$z_score,
        deep_baseline_zscores_vs_gr$gr_value
      )
    ),
    .groups = "drop"
  )

```


## Correlation YKL differential expression with GR metrics

Both GR corrected and uncorrected

```{r}
ykl_cell_lines_to_keep <- intersect(
  deseq_res_gr %>%
    filter(agent == "YKL-5-124", experiment == "cdk_467_2019_08") %>%
    pull(cell_line),
  gr_raw$cell_line
)

deseq_res_gr_long <- deseq_res_gr %>%
  filter(
    agent == "YKL-5-124",
    experiment == "cdk_467_2019_08",
    cell_line %in% ykl_cell_lines_to_keep
  ) %>%
  select(
    cell_line, ensembl_gene_id, hgnc_symbol,
    ends_with(c("_normal", "_gr"))
  ) %>%
  pivot_longer(
    cols = c(ends_with(c("_normal", "_gr"))),
    names_pattern = "(.*)_(normal|gr)",
    names_to = c("metric", "gr_correction")
  ) |>
  pivot_wider(
    names_from = metric, values_from = value
  ) %>%
  bind_rows(
    deseq_res_counts_gr_corrected %>%
      filter(
        agent == "YKL-5-124",
        experiment == "cdk_467_2019_08",
        cell_line %in% ykl_cell_lines_to_keep
      ) %>%
      mutate(
        gr_correction = "gr_counts"
      )
  )

ykl_lfc_mat <- deseq_res_gr_long %>%
  group_nest(gr_correction) %>%
  rowwise() %>%
  mutate(
    gene_data = data %>%
      filter(
        cell_line %in% ykl_cell_lines_to_keep
      ) %>%
      group_by(ensembl_gene_id) %>%
      filter(any(padj < 0.05)) %>%
      summarize(
        n_up = sum(log2FoldChange > 0 & padj < 0.05, na.rm = TRUE),
        n_down = sum(log2FoldChange < 0 & padj < 0.05, na.rm = TRUE),
        .groups = "drop"
      ) %>%
      list(),
    mat = data %>%
      filter(ensembl_gene_id %in% gene_data$ensembl_gene_id) %>%
      select(ensembl_gene_id, cell_line, log2FoldChange) %>%
      pivot_wider(names_from = cell_line, values_from = log2FoldChange, values_fill = 0) %>%
      column_to_rownames("ensembl_gene_id") %>%
      as.matrix() %>% {
        # Ensure columns are in the same order as ykl_cell_lines_to_keep
        .[, ykl_cell_lines_to_keep]
      } %>%
      list()
  ) %>%
  ungroup()


ykl_gr_mat <- gr_raw %>%
  filter(cell_line %in% ykl_cell_lines_to_keep, agent == "YKL-5-124") %>%
  select(cell_line, GR50, GRmax, GR_AOC, neglog_GR50) %>%
  column_to_rownames("cell_line") %>%
  as.matrix() %>%
  t() %>% {
    .[, ykl_cell_lines_to_keep]
  }

ykl_gr_cor_df <- ykl_lfc_mat %>%
  rowwise() %>%
  mutate(
    cor = compute_correlation(
      ykl_gr_mat,
      mat
    ) %>%
    list()
  ) %>%
  ungroup()

ykl_gr_cor_df_long <- ykl_gr_cor_df %>%
  select(gr_correction, cor) %>%
  unnest(cor)

p <- ykl_gr_cor_df_long %>%
  pivot_longer(
    cols = c(r, p, p.adj),
    names_to = "cor_metric",
    values_to = "value"
  ) %>%
  ggplot(aes(value, color = cor_metric)) +
    stat_ecdf(geom = "step") +
    facet_grid(gr_correction ~ gr_metric)

ggsave(
  file.path("plots_cdk7", "ykl_gr_cor_ecdf.pdf"),
  p,
  width = 10,
  height = 8
)

```

### Checking direction of effect

```{r}

```

### Plotting example correlations


```{r}
ykl_gr_cor_sig <- ykl_gr_cor_df_long %>%
  filter(gr_metric %in% c("GR50", "neglog_GR50")) %>%
  group_by(gr_correction) %>%
  filter(
    ensembl_gene_id %in% {
      group_by(cur_data(), gr_metric) %>%
        arrange(p) %>%
        slice_head(n = 10) %>%
        pull(ensembl_gene_id)
    }
  ) %>%
  ungroup()


ykl_gr_cor_sig_lfc <- ykl_gr_cor_sig %>%
  power_inner_join(
    ykl_lfc_mat %>%
      select(gr_correction, data) %>%
      unnest(data) %>%
      select(gr_correction, cell_line, ensembl_gene_id, hgnc_symbol, log2FoldChange),
    by = c("gr_correction", "ensembl_gene_id"),
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  ) %>%
  power_inner_join(
    gr_raw %>%
      select(cell_line, GR50, GRmax, GR_AOC, neglog_GR50) %>%
      pivot_longer(-cell_line, names_to = "gr_metric", values_to = "gr_value"),
    by = c("cell_line", "gr_metric")
  )

p <- ykl_gr_cor_sig_lfc %>%
  group_nest(gr_correction, gr_metric) %>%
  mutate(
    p = pmap(
      list(data, gr_metric, gr_correction),
      function(df, g, gc) {
        ggplot(
          df %>%
            arrange(p) %>%
            mutate(across(hgnc_symbol, fct_inorder)),
          aes(gr_value, log2FoldChange)
        ) +
          geom_point() +
          geom_smooth(method = "lm") +
          geom_text(
            aes(x = x, y = y, label = p),
            hjust = 0, vjust = 1,
            nudge_x = 0.1, nudge_y = -0.1,
            data = \(df) distinct(df, hgnc_symbol, p) %>%
              mutate(
                across(p, \(x) signif(x, digits = 2)),
                x = if (g == "GR50") Inf else -Inf, y = Inf
              )
          ) +
          facet_wrap(~hgnc_symbol, scales = "free") +
          scale_x_continuous(
            trans = switch(
              g,
              neglog_GR50 = "identity",
              GR50 = c("reverse"),
              "identity"
            )
          ) +
          labs(
            x = g,
            y = "log2FoldChange YKL-5-124 vs DMSO",
            title = paste("GR correction:", gc, "Metric:", g)
          )
      }
    )
  )

ps <- patchwork::wrap_plots(p$p, ncol = 2)
ps <- gridExtra::arrangeGrob(grobs = p$p)


ggsave(
  file.path("plots_cdk7", "ykl_gr_cor_most_sig.pdf"),
  ps,
  width = 16,
  height = 16
)

```


```{r}
ykl_gr_cor_fgsea_input <- ykl_gr_cor_df_long %>%
  drop_na(r) %>%
  group_nest(gr_metric, gr_correction) %>%
  mutate(
    de = map(
      data,
      \(x) set_names(x$r, x$ensembl_gene_id)
    )
  )

ykl_gr_cor_fgsea_res_raw <- ykl_gr_cor_fgsea_input %>%
  mutate(
    res = map(
      de,
      \(x) withr::with_seed(42, quietly(fgseaMultilevel)(
        selected_gene_sets_list$gs,
        x
      ))
    ),
    res_reduced = future_pmap(
      list(res, de),
      \(res, de) {
        collapsePathways(
          filter(res$result, padj < .05),
          selected_gene_sets_list$gs,
          de
        )
      },
      .progress = TRUE
    )
  )

ykl_gr_cor_fgsea_res <- ykl_gr_cor_fgsea_res_raw %>%
  unnest_wider(res, transform = list(warnings = \(x) paste(x, collapse = ""))) %>%
  transmute(
    gr_metric, gr_correction,
    res = map2(
      result, res_reduced,
      \(full, reduced) {
        mutate(
          full,
          is_main = pathway %in% reduced$mainPathways,
          parent_pathway = reduced$parentPathways[pathway]
        )
      }
    )
  ) %>%
  unnest(res) %>%
  mutate(
    signed_p = -sign(NES) * log10(padj),
    p_cut = cut(padj, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("**", "*", "")),
    leadingEdgeNames = map(
      leadingEdge,
      \(x) gene_id_to_name_map[["dge"]][x]
    )
  ) %>%
  arrange(padj, pval)
```

### Excel sheet

```{r}
ykl_gr_cor_fgsea_excel <- createWorkbook()

ykl_gr_cor_fgsea_res %>%
  select(-c(leadingEdge)) %>%
  filter(
    gr_metric != "neglog_GR50"
  ) %>%
  arrange(padj, pval) %>%
  group_by(gr_correction) %>%
  group_walk(
    \(df, g) {
      addWorksheet(
        ykl_gr_cor_fgsea_excel,
        g$gr_correction
      )
      writeDataTable(
        ykl_gr_cor_fgsea_excel,
        sheet = g$gr_correction,
        x = df,
        tableStyle = "TableStyleLight8"
      )
      # setColWidths(
      #   ykl_gr_cor_fgsea_excel, g$gr_correction,
      #   cols = seq_len(ncol(df)),
      #   widths = "auto"
      # )
      walk(
        which(names(df) == "padj"),
        \(i) {
          data_rows <- nrow(df) + 1L  # +1 for header
          de_col_data <- df[[i]]
          significant_rows <- which(!is.na(de_col_data) & de_col_data < 0.05) + 1L  # +1 for header
          if (length(significant_rows) > 0) {
            addStyle(
              wb = ykl_gr_cor_fgsea_excel,
              sheet = g$gr_correction,
              style = significant_style,
              rows = significant_rows,
              cols = i,
              gridExpand = FALSE
            )
          }
        }
      )
    }
  )

saveWorkbook(
  ykl_gr_cor_fgsea_excel,
  file.path("plots_cdk7", "ykl_gr_cor_fgsea.xlsx"),
  overwrite = TRUE
)


```

### Dot plots

```{r}
ykl_gr_cor_fgsea_hm <- ykl_gr_cor_fgsea_res %>%
  filter(
    gr_metric != "neglog_GR50"
  ) %>%
  mutate(
    p_cut = cut(
      padj,
      breaks = c(-Inf, 0.001, 0.01, 0.05, Inf),
      labels = c("<0.001", "<0.01", "<0.05", "ns")
    )
  ) %>%
  group_nest(
    gr_correction
  ) %>%
  mutate(
    res = map(
      data,
      \(x) {
        # browser()
        # d <- semi_join(
        #   x,
        #   group_by(x, contrast) %>%
        #     filter(padj < y),
        #   by = "pathway"
        # )
        d <- semi_join(
          x,
          group_by(x, gr_metric) %>%
            filter(padj < .05) %>%
            arrange(pval) %>%
            slice_head(n = 20),
          by = "pathway"
        )
        # d <- semi_join(
        #   x,
        #   group_by(x, pathway, gr_metric) %>%
        #     filter(padj < .05) %>%
        #     summarize(
        #       groups = paste(sort(unique(paste(gr_metric, sep = "_"))), collapse = "_"),
        #       p_mean = 1 / sum(1 / padj, na.rm = TRUE),
        #       .groups = "drop"
        #     ) %>%
        #     group_by(groups) %>%
        #     arrange(p_mean) %>%
        #     slice_head(n = 20),
        #     # bind_rows(tibble(pathway = "P_P38_MK2_PATHWAY")),
        #   by = "pathway"
        # )
        if (nrow(d) <= 6) {
          return(NULL)
        }
        dwide <- d %>%
          arrange(pathway) %>%
          pivot_wider(
            id_cols = pathway,
            names_from = gr_metric,
            values_from = NES
          )
        dmat <- dwide %>%
          column_to_rownames("pathway") %>%
          as.matrix() %>%
          dist()
        # browser()
        clust <- hclust(dmat, method = "average") %>%
          reorder(dmat, method = "OLO")
        row_labels <- clust$labels[clust$order] %>% {
          set_names(
            str_replace_all(., coll("_"), " ") %>%
              str_sub(start = 2),
            .
          )
        }
        # sim_hm <- msigdbr_of_interest_sim_trans[
        #   rev(clust$labels[clust$order]), rev(clust$labels[clust$order])
        # ] %>%
        #   pheatmap(
        #     cluster_rows = FALSE,
        #     cluster_cols = FALSE
        #   )
        est_perc <- quantile(d$NES, c(.05, .95))
        abs_max_est_perc <- max(abs(est_perc))
        # browser()
        p <- ggplot(
          d %>%
            mutate(
              pathway = factor(pathway, levels = clust$labels[clust$order])
            ),
          aes(
            x = gr_metric,
            y = fct_rev(pathway),
            color = NES,
            size = p_cut
          )
        ) +
          geom_point(
            shape = 16
          ) +
          paletteer::scale_color_paletteer_c(
            palette = "ggthemes::Orange-Blue-White Diverging",
            limits = c(-abs_max_est_perc, abs_max_est_perc),
            oob = scales::squish,
            direction = -1
          ) +
          scale_size_manual(
            values = rev(c("<0.001" = 8, "<0.01" = 6, "<0.05" = 4, "ns" = 2))
          ) +
          scale_x_discrete(position = "top") +
          scale_y_discrete(
            breaks = names(row_labels),
            labels = unname(row_labels)
          ) +
          theme(
            axis.text.x = element_text(angle = 45, hjust = 0),
            panel.grid.major.x = element_blank()
          ) +
          labs(
            x = NULL,
            y = NULL,
            size = "FDR",
            color = "NES"
          )
        list(
          p = p,
          clust_data = d
          # sim_hm = sim_hm
        ) %>%
          map(list)
      }
    )
  ) %>%
  select(-data) %>%
  unnest_wider(
    res
  )

pwalk(
  ykl_gr_cor_fgsea_hm,
  \(gr_correction, p, ...) {
    ggsave(
      file.path(
        "plots_cdk7", paste0("ykl_gr_cor_fgsea_dotplot_", gr_correction, ".pdf")
      ), p[[1]], width = 8, height = 10, dev = cairo_pdf
    )
  }
)


```


### Bar plots of top enriched pathways

```{r}
ps <- ykl_gr_cor_fgsea_res %>%
  filter(padj < .05) %>%
  group_nest(
    gr_metric, gr_correction
  ) %>%
  mutate(
    p = map2(
      data, gr_metric,
      \(df, gr_metric) {
        ggplot(
          df %>%
            arrange(pval) %>%
            slice_head(n = 30) %>%
            mutate(pathway = fct_reorder(pathway, NES)),
          aes(NES, pathway, fill = p_cut)
        ) +
          geom_col() +
          geom_text(
            aes(
              x = x_pos,
              y = pathway,
              label = pathway,
              hjust = hjust
            ),
            data = \(x) mutate(
              x,
              x_pos = if_else(NES > 0, -.1, .1),
              hjust = if_else(NES > 0, 1, 0)
            ) %>%
              distinct(pathway, x_pos, hjust),
            inherit.aes = FALSE
          ) +
          scale_fill_viridis_d(direction = -1) +
          labs(
            x = "Normalized Enrichment Score",
            y = NULL,
            title = gr_metric
          ) +
          theme(
            axis.text.y = element_blank()
          )
      }
    )
  )

pwalk(
  ps,
  \(gr_metric, p, data, ...) {
    h <- nrow(data) * .2 + 2
    ggsave(
      file.path("plots_cdk7", paste0("fgsea_bars_gr_ykl_cor_", gr_metric, ".pdf")),
      p, width = 8, height = h
    )
  }
)
```

1. Share table of enrichment results and go through plots, put them in Dropbox with README


```{r}

library(seriation)
cluster_df <- function(df, row_var, col_var, value_var, values_fill = 0) {
  # browser()
  mat <- df %>%
    select({{row_var}}, {{col_var}}, {{value_var}}) %>%
    pivot_wider(names_from = {{col_var}}, values_from = {{value_var}}, values_fill = values_fill) %>%
    column_to_rownames(rlang::as_name(rlang::enquo(row_var)))
  # browser()
  if (rlang::is_bare_numeric(pull(df, {{value_var}}))) {
    dist_rows <- dist(mat, method = "euclidian")
    dist_cols <- dist(t(mat), method = "euclidian")
  } else {
    # browser()
    dist_rows <- cluster::daisy(mat, metric = "gower")
    dist_cols <- t(mat) %>%
      as.data.frame() %>%
      mutate(across(everything(), \(x) factor(x, levels = levels(pull(df, {{value_var}}))))) %>%
      cluster::daisy(metric = "gower")
  }
  clust_rows <- hclust(dist_rows, method = "average") %>%
      reorder(dist_rows, method = "olo")
  clust_cols <- hclust(dist_cols, method = "average") %>%
      reorder(dist_cols, method = "olo")
  df %>%
    mutate(
      "{{row_var}}" := factor({{row_var}}, levels = clust_rows$labels[clust_rows$order]),
      "{{col_var}}" := factor({{col_var}}, levels = clust_cols$labels[clust_cols$order])
    )
}

p <- fgsea_res_p_long %>%
  filter(
    gr_metric != "GR50",
    str_starts(pathway, "REACTOME")
  ) %>%
  filter(
    pathway %in% {
      group_by(., pathway) %>%
        summarize(n_sig = sum(padj < 0.01)) %>%
        filter(n_sig > 0) %>%
        pull(pathway)
    }
  ) %>%
  cluster_df(
    pathway, gr_metric, NES
  ) %>%
  ggplot(
    aes(gr_metric, pathway, fill = NES)
  ) +
    geom_raster() +
    geom_text(
      aes(
        label = p_cut,
        color = after_scale(better_contrast(fill, c("white", "black"), 0))
      ),
      hjust = .5, vjust = .5
    ) +
    scico::scale_fill_scico(
      palette = "vik", midpoint = 0
    ) +
    theme_minimal()

```

## Differential expression Excel sheets

```{r}
ykl_de_excel_wbs <- list()

deseq_res_gr_long %>%
  select(-c(experiment, agent, time)) %>%
  arrange(padj, pvalue) %>%
  group_by(gr_correction, cell_line) %>%
  group_walk(
    \(df, g) {
      if (is.null(ykl_de_excel_wbs[[g$gr_correction]])) {
        ykl_de_excel_wbs[[g$gr_correction]] <<- createWorkbook()
      }
      wb <- ykl_de_excel_wbs[[g$gr_correction]]
      addWorksheet(
        wb,
        g$cell_line
      )
      writeDataTable(
        wb,
        sheet = g$cell_line,
        x = df,
        tableStyle = "TableStyleLight8"
      )
      # setColWidths(
      #   ykl_gr_cor_fgsea_excel, g$gr_correction,
      #   cols = seq_len(ncol(df)),
      #   widths = "auto"
      # )
      walk(
        which(names(df) == "padj"),
        \(i) {
          data_rows <- nrow(df) + 1L  # +1 for header
          de_col_data <- df[[i]]
          significant_rows <- which(!is.na(de_col_data) & de_col_data < 0.05) + 1L  # +1 for header
          if (length(significant_rows) > 0) {
            addStyle(
              wb = wb,
              sheet = g$cell_line,
              style = significant_style,
              rows = significant_rows,
              cols = i,
              gridExpand = FALSE
            )
          }
        }
      )
      walk(
        which(str_detect(names(df), coll("log2FoldChange"))),
        \(i) {
          data_rows <- nrow(df) + 1L  # +1 for header
          conditionalFormatting(
            wb = wb,
            sheet = g$cell_line,
            cols = i,
            rows = 2:data_rows,
            type = "colourScale",
            style = c("#4575b4", "white", "#d73027"),
            rule = c(-3, 0, 3)
          )
        }
      )
      setColWidths(
        wb, g$cell_line,
        cols = seq_len(ncol(df)),
        widths = "auto"
      )
    }
  )

iwalk(
  ykl_de_excel_wbs,
  \(wb, gr_correction) {
    saveWorkbook(
      wb,
      file.path("plots_cdk7", paste0("ykl_de_results_", gr_correction, ".xlsx")),
      overwrite = TRUE
    )
  }
)


```


## FGSEA for raw YKL DGE results


```{r}
selected_des <- ykl_lfc_mat %>%
  transmute(
    gr_correction,
    data = map2(
      data, genes_to_keep,
      \(df, genes) {
        filter(
          df,
          ensembl_gene_id %in% genes
        )
      }
    )
  ) %>%
  unnest(data) %>%
  drop_na(pvalue) %>%
  group_nest(gr_correction, cell_line) %>%
  mutate(
    de = map(
      data,
      \(x) set_names(x$log2FoldChange, x$ensembl_gene_id)
    ),
    de_p = map(
      data,
      \(x) set_names(-sign(x$log2FoldChange_MLE) * log10(x$pvalue), x$ensembl_gene_id)
    )
  )

fgsea_res_raw <- selected_des %>%
  mutate(
    res_lfc = map(
      de,
      \(x) with_seed(42, quietly(fgseaMultilevel)(
        selected_gene_sets_list$gs,
        x
      ))
    ),
    res_p = map(
      de_p,
      \(x) with_seed(42, quietly(fgseaMultilevel)(
        selected_gene_sets_list$gs,
        x
      ))
    )
  ) %>%
  pivot_longer(
    c(res_lfc, res_p),
    names_to = "type",
    values_to = "res"
  ) %>%
  mutate(
    res_reduced = future_pmap(
      list(res, de, type),
      \(res, de, type) {
        collapsePathways(
          filter(res$result, padj < .05),
          selected_gene_sets_list$gs,
          de
        )
      },
      .progress = TRUE
    )
  )


fgsea_res_raw %>%
  select(gr_correction, cell_line, res, res_p) %>%
  pivot_longer(
    c(res, res_p),
    names_to = "type",
    values_to = "result"
  ) %>%
  mutate(
    warnings = map_chr(
      result,
      \(x) paste(x$warnings, collapse = "; ")
    )
  ) %>%
  pull(warnings)

fgsea_res_p_long <- fgsea_res_raw %>%
  select(cell_line, gr_correction, res_p) %>%
  unnest(res_p) %>%
  mutate(
    signed_p = -sign(NES) * log10(padj),
    p_cut = cut(padj, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("**", "*", ""))
  ) %>%
  filter(
    pathway %in% {
      group_by(., pathway) %>%
        summarize(n_sig = sum(padj < 0.05)) %>%
        filter(n_sig > 0) %>%
        pull(pathway)
    }
  ) %>%
  mutate(
    size_leading_edge = map_int(
      leadingEdge, length
    )
  ) %>%
  select(
    cell_line, gr_correction, pathway,
    pval, padj, p_cut, ES, NES, size_leading_edge, size,
    leadingEdge
  )

```

### Correlate FGSEA result with GR metrics

```{r}
ykl_fgsea_res_mats <- fgsea_res_p_long %>%
  mutate(
    signed_p = -sign(NES) * log10(pval)
  ) %>%
  select(cell_line, gr_correction, pathway, signed_p, NES) %>%
  pivot_longer(c(signed_p, NES), names_to = "metric", values_to = "value") %>%
  pivot_wider(names_from = cell_line, values_from = value) %>%
  group_nest(gr_correction, metric, .key = "mat") %>%
  mutate(
    mat = map(
      mat,
      \(x) column_to_rownames(x, "pathway") %>%
        as.matrix()
    )
  )

ykl_fgsea_res_gr_cor <- ykl_fgsea_res_mats %>%
  mutate(
    cor = map(
      mat,
      \(x) compute_correlation(
        ykl_gr_mat,
        x[, colnames(ykl_gr_mat)]
      ) %>%
        arrange(p)
    )
  )

ykl_fgsea_res_gr_cor_long <- ykl_fgsea_res_gr_cor %>%
  select(gr_correction, metric, cor) %>%
  unnest(cor)

write_csv(
  ykl_fgsea_res_gr_cor_long,
  file.path("data", "ykl_fgsea_res_gr_cor_long.csv.gz")
)
```


```{r}
cv_model <- glmnet::cv.glmnet(
  t(ykl_lfc_mat), ykl_gr_mat["neglog_GR50",],
  alpha = 0.5,
  # nfolds = ncol(ykl_lfc_mat),
  foldid = 1:ncol(ykl_lfc_mat)
)

best_model <- glmnet::glmnet(
  t(ykl_lfc_mat), ykl_gr_mat["neglog_GR50",],
  alpha = 0.5, lambda = cv_model$lambda.min
)

sparse_mat_to_df <- function(x) {
  y <- summary(as(x, "dgTMatrix"))
  y$row = rownames(x)[y$i]
  y$col = colnames(x)[y$j]
  y
}

best_model_coef <- sparse_mat_to_df(coef(best_model)) %>%
  dplyr::rename(
    ensembl_gene_id = row,
    coef = x
  ) %>%
  left_join(
    ensembl_gtf %>%
      select(ensembl_gene_id, hgnc_symbol, gene_name),
    by = "ensembl_gene_id"
  ) %>%
  as_tibble()

best_model_coef_lfc <- best_model_coef %>%
  mutate(gr_metric = "neglog_GR50") %>%
  inner_join(
    ykl_lfc_mat %>%
      as_tibble(rownames = "ensembl_gene_id") %>%
      pivot_longer(
        cols = -ensembl_gene_id,
        names_to = "cell_line",
        values_to = "log2FoldChange"
      )
  ) %>%
  inner_join(
    gr_raw %>%
      select(cell_line, GR50, GRmax, GR_AOC, neglog_GR50) %>%
      pivot_longer(-cell_line, names_to = "gr_metric", values_to = "gr_value"),
  )

p <- best_model_coef_lfc %>%
  group_nest(gr_metric) %>%
  mutate(
    p = map2(
      data, gr_metric,
      function(df, g) {
        ggplot(
          df %>%
            arrange(coef) %>%
            mutate(across(gene_name, fct_inorder)),
          aes(gr_value, log2FoldChange)
        ) +
          geom_point() +
          geom_smooth(method = "lm") +
          geom_text(
            aes(x = x, y = y, label = coef),
            hjust = 0, vjust = 1,
            nudge_x = 0.1, nudge_y = -0.1,
            data = \(df) distinct(df, gene_name, coef) %>%
              mutate(
                across(coef, \(x) signif(x, digits = 2)),
                x = if (g == "GR50") Inf else -Inf, y = Inf
              )
          ) +
          facet_wrap(~gene_name, scales = "free") +
          scale_x_continuous(
            trans = switch(
              g,
              neglog_GR50 = "identity",
              GR50 = c("reverse"),
              "identity"
            )
          ) +
          labs(
            x = g,
            y = "log2FoldChange YKL-5-124 vs DMSO",
            title = g
          )
      }
    )
  )

ps <- patchwork::wrap_plots(p$p)

ggsave(
  file.path("plots_cdk7", "ykl_gr_lasso_most_sig.pdf"),
  ps,
  width = 8,
  height = 8
)

```


```{r}
model_res_table <- list(
  baseline = baseline_lfc_cor_df %>%
    select(
      ensembl_gene_id, hgnc_symbol,
      gr_metric, r, p, p.adj
    ),
  ykl = ykl_gr_cor_df %>%
    select(
      ensembl_gene_id, hgnc_symbol,
      gr_metric, r, p, p.adj
    ),
  ykl_elastic_net = best_model_coef %>%
    transmute(
      ensembl_gene_id, hgnc_symbol,
      gr_metric = "neglog_GR50",
      coef
    )
) %>%
  purrr::reduce2(
    .y = names(.),
    .init = tibble(ensembl_gene_id = character(), hgnc_symbol = character(), gr_metric = character()),
    \(df1, df2, prefix) {
      full_join(
        df1,
        rename_with(df2, ~paste0(prefix, "_", .x), -c(ensembl_gene_id, hgnc_symbol, gr_metric)),
        by = c("ensembl_gene_id", "hgnc_symbol", "gr_metric")
      )
    }
  ) %>%
  replace_na(list(ykl_elastic_net_coef = 0)) %>%
  group_by(gr_metric) %>%
  mutate(
    rank_ykl_elastic_net = rank(-abs(ykl_elastic_net_coef)),
    rank_ykl = rank(ykl_p),
    rank_baseline = rank(baseline_p),
    rank_average = (rank_ykl_elastic_net + rank_ykl) / 3
  ) %>%
  ungroup() %>%
  mutate(
    ensembl_gene_id = factor(
      ensembl_gene_id,
      levels = filter(., gr_metric %in% c("neglog_GR50")) %>%
        group_by(ensembl_gene_id) %>%
        summarize(
          rank_average = mean(rank_average),
          .groups = "drop"
        ) %>%
        arrange(rank_average) %>%
        pull(ensembl_gene_id)
    )
  ) %>%
  arrange(ensembl_gene_id) %>%
  filter(gr_metric %in% c("neglog_GR50", "GR50"))

xlsx::write.xlsx(
  model_res_table,
  file.path("plots_cdk7", "gr_correlation_table.xlsx")
)

```


## Acquired resistance vs acute treatment


```{r}
deseq_res_acquired_data <- deseq_res_acquired %>%
  mutate(
    signed_p = -sign(log2FoldChange_MLE) * log10(padj)
  )

deseq_res_resistant_vs_treatment <- tribble(
  ~resistant, ~treatment,
  "YKLr1", "Y1uM",
  "YKLr2", "Y1uM",
  "YKLr3", "Y1uM",
  "SYr3", "S100nM",
  "YKLrTQr1", "Y1uM",
  "YKLrTQr3", "Y1uM"
) %>%
  power_inner_join(
    deseq_res_acquired_data %>%
      transmute(
        cell_line, resistant, resistant_condition = condition,
        gene_id, log2FoldChange, padj, signed_p
      ),
    by = "resistant",
    check = check_specs(
      duplicate_keys_left = "warn",
      unmatched_keys_left = "warn"
    )
  ) %>%
  power_inner_join(
    deseq_res_acquired_data %>%
      transmute(
        cell_line, treatment, treatment_condition = condition,
        gene_id, log2FoldChange, padj,signed_p
      ),
    by = c("gene_id", "treatment", "cell_line"),
    suffix = c("_resistant", "_treatment"),
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  )

```

```{r}
deseq_res_acquired_signif_any <- deseq_res_acquired_data %>%
  filter(padj < .01) %>%
  distinct(gene_id)

p <- deseq_res_resistant_vs_treatment %>%
  semi_join(
    deseq_res_acquired_signif_any,
    by = "gene_id"
  ) %>%
  mutate(
    across(
      c(starts_with("log2FoldChange"), starts_with("padj")),
      \(x) DescTools::Winsorize(x, val = quantile(x, probs = c(0.001, 0.999), na.rm = TRUE))
    )
  ) %>%
  ggplot(
    aes(
      x = log2FoldChange_resistant,
      y = log2FoldChange_treatment
    )
  ) +
  geom_point(alpha = .2) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "dashed") +
  facet_wrap(
    vars(resistant_condition, treatment_condition)
  )
p

ggsave(
  file.path("plots_cdk7", "deseq_resistant_vs_treatment_lfc.pdf"),
  p,
  width = 10,
  height = 10
)
```

### Select genes only in resistant but not treatment


```{r}
deseq_res_resistant_vs_treatment_classes <- deseq_res_resistant_vs_treatment %>%
  mutate(
    resistant_class = case_when(
      padj_resistant > .01 ~ "ns",
      log2FoldChange_resistant > 0 ~ "up",
      log2FoldChange_resistant < 0 ~ "down",
      TRUE ~ "ns"
    ),
    treatment_class = case_when(
      padj_treatment > .01 ~ "ns",
      log2FoldChange_treatment > 0 ~ "up",
      log2FoldChange_treatment < 0 ~ "down",
      TRUE ~ "ns"
    ),
    resistant_vs_treatment_class = paste("resistant", resistant_class, "treatment", treatment_class, sep = "_"),
    resistant_vs_treatment_simple = case_when(
      resistant_class == "ns" & treatment_class == "ns" ~ "ns",
      resistant_class != "ns" & treatment_class == "ns" ~ "resistant_only",
      resistant_class == "ns" & treatment_class != "ns" ~ "treatment_only",
      resistant_class == treatment_class ~ "both_same",
      resistant_class != treatment_class ~ "opposite",
      TRUE ~ NA_character_
    )
  )

deseq_res_resistant_vs_treatment_classes %>%
  count(
    resistant_condition, resistant_vs_treatment_class
  ) %>%
  View()

deseq_res_resistant_vs_treatment_classes %>%
  count(
    resistant_condition, resistant_vs_treatment_simple
  ) %>%
  View()

p <- deseq_res_resistant_vs_treatment_classes %>%
  count(
    resistant_condition, resistant_vs_treatment_simple
  ) %>%
  filter(
    resistant_vs_treatment_simple != "ns"
  ) %>%
  ggplot(
    aes(
      n, resistant_condition, fill = resistant_vs_treatment_simple
    )
  ) +
  geom_col(position = "dodge") +
  scale_fill_manual(
    values = c(
      resistant_only = "red",
      treatment_only = "blue",
      both_same = "purple",
      opposite = "green"
    )
  ) +
  scale_x_continuous(expand = expansion(mult = c(0, .05))) +
  labs(
    x = "Number of genes", y = "Resistant condition",
    fill = "Gene expression"
  )
p

ggsave(
  file.path("plots_cdk7", "deseq_resistant_vs_treatment_classes.pdf"),
  p,
  width = 8,
  height = 6
)
```


### FGSEA and topGO on resistant vs treatment classes

use resistant_only and opoosite classes. do one version with up/down separate
and one with both combined

```{r}
feasible_gene_vec <- deseq_res_resistant_vs_treatment$gene_id %>%
  unique()

resistant_vs_treatment_sets <- bind_rows(
  deseq_res_resistant_vs_treatment_classes %>%
    filter(
      resistant_vs_treatment_simple %in% c("resistant_only", "opposite")
    ) %>% {
      list(
        separate = group_by(., resistant_condition, resistant_class) %>%
          summarize(
            gene_set = list(gene_id),
            .groups = "drop"
          ),
        combined = group_by(., resistant_condition) %>%
          summarize(
            gene_set = list(gene_id),
            .groups = "drop"
          )
      )
    } %>%
    bind_rows(.id = "updown_definition") %>%
    mutate(
      resistant_vs_treatment_simple = "resistant_only_and_opposite"
    ),
  deseq_res_resistant_vs_treatment_classes %>%
    filter(
      resistant_vs_treatment_simple %in% c("resistant_only", "opposite")
    ) %>% {
      list(
        separate = group_by(., resistant_condition, resistant_class, resistant_vs_treatment_simple) %>%
          summarize(
            gene_set = list(gene_id),
            .groups = "drop"
          ),
        combined = group_by(., resistant_condition, resistant_vs_treatment_simple) %>%
          summarize(
            gene_set = list(gene_id),
            .groups = "drop"
          )
      )
    } %>%
    bind_rows(.id = "updown_definition")
) %>%
  mutate(
    gene_set_vec = map(
      gene_set,
      \(x) set_names(if_else(feasible_gene_vec %in% x, .01, 1), feasible_gene_vec)
    )
  )
```



```{r}
library(topGO)

gene_sel_fun <- \(x) x < .05

topgo_objects <- c("BP", "MF", "CC") %>%
  set_names() %>%
  map(
    \(x) new(
      "topGOdata",
      ontology = x,
      allGenes = resistant_vs_treatment_sets$gene_set_vec[[1]],
      geneSel = gene_sel_fun,
      nodeSize = 10,
      annot = annFUN.org,
      ID = "ensembl",
      mapping = "org.Hs.eg.db"
    )
  )

qsave(
  topgo_objects,
  "data/topgo_objects.qs"
)
topgo_objects <- qread("data/topgo_objects.qs")

# go_gene_map <- map(
#   topgo_objects,
#   \(x) x@graph@nodes %>%
#     set_names() %>%
#     map(\(y) genesInTerm(x, y)[[1]])
# ) %>%
#   # unlist(recursive = FALSE)
#   {do.call(c, .)}

# qsave(
#   go_gene_map,
#   "go_gene_map.qs"
# )
# go_gene_map <- qread("go_gene_map.qs")

resistant_vs_treatment_sets_topgo_raw <- resistant_vs_treatment_sets %>%
  crossing(
    ontology = names(topgo_objects)
  ) %>%
  rowwise() %>%
  mutate(
    res = {
      go <- updateGenes(
        topgo_objects[[ontology]],
        geneList = gene_set_vec,
        geneSelFun = gene_sel_fun
      )
      # browser()
      raw <- runTest(
        go,
        algorithm = "weight01",
        statistic = "fisher"
      )
      res_table <- GenTable(
        go,
        fisher = raw,
        topNodes = length(usedGO(go)),
        numChar = 300
      )
      list(
        res_raw = raw,
        res = res_table
      ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  resistant_vs_treatment_sets_topgo_raw,
  "data/resistant_vs_treatment_sets_topgo_raw.qs"
)
resistant_vs_treatment_sets_topgo_raw <- qread("data/resistant_vs_treatment_sets_topgo_raw.qs")

resistant_vs_treatment_sets_topgo_res <- resistant_vs_treatment_sets_topgo_raw %>%
  rowwise() %>%
  mutate(
    res_table = list(res[["res"]])
  ) %>%
  ungroup() %>%
  dplyr::select(-c(res, gene_set_vec, gene_set)) %>%
  unnest(res_table) %>%
  mutate(
    term_unique = paste(
      Term, GO.ID, sep = "_"
    ),
    fisher_double = as.double(str_remove(fisher, fixed("< "))),
    log_enrichment = log10(Significant / Expected),
    signed_p = -sign(log_enrichment) * log10(fisher_double)
  )

qsave(
  resistant_vs_treatment_sets_topgo_res,
  "data/resistant_vs_treatment_sets_topgo_res.qs"
)
resistant_vs_treatment_sets_topgo_res <- qread("data/resistant_vs_treatment_sets_topgo_res.qs")

go_meta <- topgo_res %>%
  distinct(
    ontology, GO.ID, Term, term_unique
  )


```


```{r}
fgsea_res_raw <- resistant_vs_treatment_sets %>%
  mutate(
    res = map(
      gene_set,
      \(x) fora(
        selected_gene_sets_list$gs,
        x, universe = feasible_gene_vec
      )
    )
  )

fgsea_res_long <- fgsea_res_raw %>%
  select(-c(gene_set, gene_set_vec)) %>%
  unnest(res) %>%
  mutate(
    signed_p = -sign(foldEnrichment) * log10(padj),
    p_cut = cut(padj, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("**", "*", ""))
  )


```

#### Heatmap

```{r}
fgsea_res_selected <- fgsea_res_long %>%
  group_nest(
    resistant_vs_treatment_simple, updown_definition, resistant_class
  ) %>%
  mutate(
    data = map(
      data,
      \(x) semi_join(
        x,
        group_by(x, pathway) %>%
          summarize(n_sig = sum(padj < 0.01)) %>%
          filter(n_sig > 1)
      )
    )
  )

fgsea_res_hms <- fgsea_res_selected %>%
  mutate(
    hm = map(
      data,
      \(x) {
        mat <- x %>%
          select(resistant_condition, pathway, signed_p) %>%
          pivot_wider(names_from = resistant_condition, values_from = signed_p, values_fill = 0) %>%
          column_to_rownames("pathway") %>%
          as.matrix()
        abs_95 <- max(abs(quantile(mat, c(0.05, 0.95))))
        Heatmap(
          mat,
          col = circlize::colorRamp2(
            c(-abs_95, 0, abs_95),
            c("blue", "white", "red")
          ),
          name = "signed -log10(padj)",
          cluster_rows = TRUE,
          cluster_columns = FALSE,
          show_row_names = TRUE,
          show_column_names = TRUE
        )
      }
    )
  )



hm <- Heatmap(
  fgsea_res_selected_mat,
  name = "signed -log10(padj)",
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  show_column_names = TRUE
)
hm

fgsea_res_selected_mat <- fgsea_res_selected %>%
  filter(updown_definition == "separate") %>%
  # Clip at 95th percentile for better color scaling
  mutate(signed_p = pmin(signed_p, quantile(signed_p, 0.95))) %>%
  select(resistant_condition, resistant_class, pathway, signed_p) %>%
  pivot_wider(names_from = c(resistant_condition, resistant_class), values_from = signed_p, values_fill = 0) %>%
  column_to_rownames("pathway") %>%
  as.matrix()

hm <- Heatmap(
  fgsea_res_selected_mat,
  name = "signed -log10(padj)",
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  show_column_names = TRUE
)
hm
```


```{r}
resistant_vs_treatment_sets_topgo_res_selected <- resistant_vs_treatment_sets_topgo_res %>%
  semi_join(
    group_by(., updown_definition, term_unique) %>%
      summarize(n_sig = sum(fisher_double < 0.01)) %>%
      filter(n_sig > 1),
    by = c("updown_definition", "term_unique")
  )

resistant_vs_treatment_sets_topgo_res_selected_mat <- resistant_vs_treatment_sets_topgo_res_selected %>%
  filter(updown_definition == "separate") %>%
  select(resistant_class, resistant_condition, term_unique, signed_p) %>%
  pivot_wider(names_from = c(resistant_class, resistant_condition), values_from = signed_p, values_fill = 0) %>%
  column_to_rownames("term_unique") %>%
  as.matrix()

hm <- Heatmap(
  resistant_vs_treatment_sets_topgo_res_selected_mat,
  name = "signed -log10(padj)",
  cluster_rows = TRUE,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  show_column_names = TRUE
)
hm

InteractiveComplexHeatmap::htShiny(hm)
```


### FGSEA on results of linear model


```{r}
selected_msigdbr_gr <- selected_msigdbr %>%
  group_nest(gs_collection, gs_subcollection, gs_name, gs_name_db, gs_db) %>%
  mutate(
    gs = map(
      data,
      \(x) unique(x$ensembl_gene)
    ) %>%
      set_names(gs_name_db)
  )

deseq_res_acquired_ykl_fgsea_input <- deseq_res_acquired_ykl %>%
  group_by(contrast, cell_line) %>%
  summarize(
    gene_vec = list(set_names(replace_na(signed_p, 0), gene_id)),
    .groups = "drop"
  )

fgsea_res_raw <- deseq_res_acquired_ykl_fgsea_input %>%
  mutate(
    res = map(
      gene_vec,
      \(x) fgseaMultilevel(
        selected_msigdbr_gr$gs,
        x
      )
    )
  )

fgsea_res_long <- fgsea_res_raw %>%
  select(-c(gene_vec)) %>%
  unnest(res) %>%
  mutate(
    signed_p = -sign(NES) * log10(padj),
    p_cut = cut(padj, breaks = c(-Inf, 0.01, 0.05, Inf), labels = c("**", "*", ""))
  )


```


## CRISPR screen


```{r}
crispr_raw <- syn("syn68586959") %>%
  read_csv()

crispr_symbol_map <- syn("syn69693337") %>%
  read_csv(skip = 1) %>%
  mutate(
    across(
      `Match type`,
      \(x) factor(x, levels = c("Approved symbol", "Alias symbol", "Previous symbol", "Entry withdrawn", "Unmatched"))
    )
  ) %>%
  group_by(Input) %>%
  slice(
    which.min(as.integer(`Match type`))
  ) %>%
  ungroup()

crispr_joined <- crispr_raw %>%
  power_inner_join(
    crispr_symbol_map %>%
      distinct(Input, approved_symbol = `Approved symbol`),
    by = c("gene" = "Input"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn",
      unmatched_keys_right = "warn"
    )
  ) %>%
  power_inner_join(
    ensembl_gtf %>%
      distinct(ensembl_gene_id, hgnc_symbol) %>%
      drop_na(),
    by = c("approved_symbol" = "hgnc_symbol"),
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  ) %>%
  mutate(
    signed_p = -sign(`average phenotype of strongest 5`) * log10(`Mann-Whitney p-value`)
  ) %>%
  # some ensembl gene ids map to multiple approved symbols, get rid of them (only 321)
  # no idea how that happens
  group_by(screen, ensembl_gene_id) %>%
  filter(n() == 1) %>%
  ungroup()
```

## Correlate CRISPR, DGE, and GR metrics

```{r}
baseline_deep_varstab_cor_df


prep_gigantic_hm <- function(
  gr_cor_agents = c("YKL-5-124", "SY-5609"),
  gr_cor_metrics = c("GR50", "GRmax"),
  crispr_screens = c("SYcrispri", "YKLcrispri", "SYcrispra", "YKLcrispra"),
  gene_expression_conditions = c("Y1uM", "S100nM", "YKLr1", "YKLr2", "YKLr3", "SYr3", "YKLrTQr1", "YKLrTQr3"),
  gene_expression_metric = c("signed_p"),
  gene_expression_cell_lines = c("OVCAR4", "SNU8", "OVCAR8", "OVCAR3"),
  gene_expression_emm_contrasts = c("YKLr - no"),
  gene_expression_emm_cell_lines = c("all_cells"),
  n_min = 2L,
  p_threshold = 0.01
) {
  ge_metric_sym <- sym(gene_expression_metric)
  cor_df <- baseline_deep_varstab_cor_df %>%
    filter(
      agent %in% gr_cor_agents,
      gr_metric %in% gr_cor_metrics
    )
  crispr_df <- crispr_joined %>%
    filter(screen %in% crispr_screens)
  expression_df <- deseq_res_acquired_data %>%
    rename(ensembl_gene_id = gene_id) %>%
    mutate(across(c(signed_p), \(x) replace_na(x, 0))) %>%
    filter(
      treatment %in% gene_expression_conditions,
      cell_line %in% gene_expression_cell_lines
    )
  emm_df <- deseq_res_acquired_ykl %>%
    rename(ensembl_gene_id = gene_id) %>%
    mutate(across(c(signed_p), \(x) replace_na(x, 0))) %>%
    filter(
      contrast %in% gene_expression_emm_contrasts,
      cell_line %in% gene_expression_emm_cell_lines
    )
  genes_to_keep <- list(
    cor_df %>%
      filter(p < p_threshold),
    crispr_df %>%
      filter(`Mann-Whitney p-value` < p_threshold),
    expression_df %>%
      filter(padj < p_threshold) %>%
      anti_join(
        deseq_res_resistant_vs_treatment_classes %>%
          filter(combined_simple == "both_same"),
        by = c("ensembl_gene_id" = "gene_id", "cell_line", "treatment")
      ),
    emm_df %>%
      filter(padj < p_threshold)
  ) %>%
    map("ensembl_gene_id") %>%
    {
      gene_counts <- table(unlist(.))
      names(gene_counts)[gene_counts >= n_min]
    }
  cor_df_formatted <- cor_df %>%
    select(agent, ensembl_gene_id, gr_metric, r) %>%
    pivot_wider(names_from = c(agent, gr_metric), values_from = r)
  crispr_df_formatted <- crispr_df %>%
    select(ensembl_gene_id, screen, signed_p) %>%
    pivot_wider(names_from = screen, values_from = signed_p)
  expression_df_formatted <- expression_df %>%
    select(ensembl_gene_id, cell_line, treatment, {{ge_metric_sym}}) %>%
    pivot_wider(names_from = c(treatment, cell_line), values_from = {{ge_metric_sym}})
  emm_df_formatted <- emm_df %>%
    select(ensembl_gene_id, cell_line, contrast, {{ge_metric_sym}}) %>%
    pivot_wider(names_from = c(contrast, cell_line), values_from = {{ge_metric_sym}})
  list(
    cor_df_formatted,
    crispr_df_formatted,
    expression_df_formatted,
    emm_df_formatted
  ) %>%
    map(
      \(x) filter(x, ensembl_gene_id %in% genes_to_keep)
    ) %>%
    reduce(full_join, by = "ensembl_gene_id")
}

scale_df <- function(df) {
  df %>%
    mutate(
      across(
        -ensembl_gene_id,
        \(x) scale(x, center = FALSE, scale = TRUE)[, 1]
      )
    )
}
```

```{r}
library(impute)

cluster_fun_eucl_imp <- function(mat, features_in_rows = TRUE) {
  if (!features_in_rows) {
    mat <- t(mat)
  }
  mat_imp <- impute.knn(
    mat, rng.seed = 42, colmax = 90
  )[["data"]]
  if (!features_in_rows) {
    mat_imp <- t(mat_imp)
    mat <- t(mat)
  }
  # browser()
  dist_mat <- dist(mat_imp)
  # dist_mat <- as.dist(mat)
  clust <- hclust(dist_mat, method = "complete")
  reorder(clust, dist_mat, method = "OLO")
}

hm_mat <- prep_gigantic_hm() %>%
  scale_df() %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

hm_row_clust <- cluster_fun_eucl_imp(hm_mat)

hm <- Heatmap(
  hm_mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  show_column_names = TRUE
)
withr::with_pdf(
  file.path("plots_cdk7", "gigantic_hm.pdf"),
  draw(hm),
  width = 15,
  height = 20
)



hm_mat <- prep_gigantic_hm(
  gr_cor_agents = c("YKL-5-124"),
  gr_cor_metrics = c("GR50", "GRmax"),
  crispr_screens = c("YKLcrispri", "YKLcrispra"),
  gene_expression_conditions = c(),
  gene_expression_metric = c("signed_p"),
  gene_expression_cell_lines = c("OVCAR4", "SNU8", "OVCAR8", "OVCAR3"),
  gene_expression_emm_contrasts = c("YKLr - no"),
  gene_expression_emm_cell_lines = c("all_cells"),
  n_min = 2L
) %>%
  scale_df() %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

hm_row_clust <- cluster_fun_eucl_imp(hm_mat)

hm <- Heatmap(
  hm_mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = FALSE,
  show_column_names = TRUE
)
withr::with_pdf(
  file.path("plots_cdk7", "gigantic_hm2.pdf"),
  draw(hm),
  width = 15,
  height = 20
)
```


```{r}
prep_gigantic_hm(
  gr_cor_agents = c("YKL-5-124"),
  gr_cor_metrics = c("GR50", "GRmax"),
  crispr_screens = c("YKLcrispri", "YKLcrispra"),
  gene_expression_conditions = c(),
  gene_expression_metric = c("signed_p"),
  gene_expression_cell_lines = c("OVCAR4", "SNU8", "OVCAR8", "OVCAR3"),
  gene_expression_emm_contrasts = c("YKLr - no"),
  gene_expression_emm_cell_lines = c("all_cells"),
  n_min = 1L
) %>%
  filter(
    abs(`YKL-5-124_GR50`) > .6
  )

```

```{r}
# Export data for ginormous heatmap Shiny app
gigantic_heatmap_data <- list(
  baseline_deep_varstab_cor_df = baseline_deep_varstab_cor_df,
  crispr_joined = crispr_joined,
  deseq_res_acquired_data = deseq_res_acquired_data,
  deseq_res_acquired_ykl = deseq_res_acquired_ykl,
  deseq_res_resistant_vs_treatment_classes = deseq_res_resistant_vs_treatment_classes
)

qsave(gigantic_heatmap_data, "data/gigantic_heatmap_data.qs")


ensembl_gtf_file <- "Homo_sapiens.GRCh38.109.gtf.gz"
if (!file.exists(ensembl_gtf_file)) {
  download.file(
    "ftp://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz",
    ensembl_gtf_file, method = "curl"
  )
}

ensembl_gtf_111 <- rtracklayer::readGFF(
  "Homo_sapiens.GRCh38.111.gtf.gz"
) %>%
  distinct(
    ensembl_gene_id = gene_id,
    hgnc_symbol = gene_name,
    gene_biotype
  ) %>%
  drop_na(ensembl_gene_id) %>%
  mutate(
    gene_name = coalesce(hgnc_symbol, ensembl_gene_id)
  )

gene_id_name_map <- with(
  ensembl_gtf_111,
  set_names(gene_name, ensembl_gene_id)
)

qsave(
  gene_id_name_map,
  "data/ensembl_111_gene_id_name_map.qs"
)
gene_id_name_map <- qread("data/ensembl_111_gene_id_name_map.qs")
```



```{r}
deseq_res_yklr_vs_treatment <- deseq_res_acquired_ykl %>%
  filter(
    contrast == "YKLr - no",
    cell_line != "all_cells"
  ) %>%
  power_inner_join(
    deseq_res_acquired_data %>%
      filter(
        treatment == "Y1uM"
      ),
    by = c("gene_id", "cell_line"),
    suffix = c("_resistant", "_treatment"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn",
      duplicate_keys_left = "warn"
    )
  )
```

```{r}
p <- deseq_res_yklr_vs_treatment %>%
  filter(
    padj_resistant < .01 | padj_treatment < .01
  ) %>%
  mutate(
    across(
      c(starts_with("log2FoldChange"), starts_with("padj")),
      \(x) DescTools::Winsorize(x, val = quantile(x, probs = c(0.001, 0.999), na.rm = TRUE))
    )
  ) %>%
  ggplot(
    aes(
      x = log2FoldChange_resistant,
      y = log2FoldChange_treatment
    )
  ) +
  geom_point(alpha = .2) +
  geom_abline(slope = 1, intercept = 0, color = "red", linetype = "dashed") +
  facet_wrap(
    vars(cell_line)
  )
p

ggsave(
  file.path("plots_cdk7", "deseq_resistant_vs_treatment_lfc.pdf"),
  p,
  width = 10,
  height = 10
)
```


Use resistant lines with Tariquidar OVCAR3-YKLr1, which is resistant primarily
due to drug efflux via ABCB1/MDR1, as a negative control to identify genes

To Do:
1. Make medium heatmap of FGSEA results from resistant single condition gene sets,
  excluding both_same
2. Same with YKLr overall signature
3. Gigantic heatmap Shiny app
4. Polish fgsea enrichment from GR correlation results

### Function to find genes by pattern


```{r}
find_genes_by_pattern <- function(
  gr_crit = NULL,
  crispr_crit = NULL,
  expression_crit = NULL,
  emm_crit = NULL,
  gr_cor_agents = c("YKL-5-124", "SY-5609"),
  gr_cor_metrics = c("GR50", "GRmax"),
  crispr_screens = c("SYcrispri", "YKLcrispri", "SYcrispra", "YKLcrispra"),
  gene_expression_conditions = c("Y1uM", "S100nM", "YKLr1", "YKLr2", "YKLr3", "SYr3", "YKLrTQr1", "YKLrTQr3"),
  gene_expression_metric = c("signed_p"),
  gene_expression_cell_lines = c("OVCAR4", "SNU8", "OVCAR8", "OVCAR3"),
  gene_expression_emm_contrasts = c("YKLr - no"),
  gene_expression_emm_cell_lines = c("all_cells"),
  combine_with = c("AND", "OR")
) {
  combine_with <- match.arg(combine_with)
  ge_metric_sym <- sym(gene_expression_metric)

  # Prepare datasets as before
  cor_df <- baseline_deep_varstab_cor_df %>%
    filter(
      agent %in% gr_cor_agents,
      gr_metric %in% gr_cor_metrics
    )
  crispr_df <- crispr_joined %>%
    filter(screen %in% crispr_screens)
  expression_df <- deseq_res_acquired_data %>%
    rename(ensembl_gene_id = gene_id) %>%
    mutate(across(c(signed_p), \(x) replace_na(x, 0))) %>%
    filter(
      treatment %in% gene_expression_conditions,
      cell_line %in% gene_expression_cell_lines
    )
  emm_df <- deseq_res_acquired_ykl %>%
    rename(ensembl_gene_id = gene_id) %>%
    mutate(across(c(signed_p), \(x) replace_na(x, 0))) %>%
    filter(
      contrast %in% gene_expression_emm_contrasts,
      cell_line %in% gene_expression_emm_cell_lines
    )

  # Format datasets
  cor_df_formatted <- cor_df %>%
    select(agent, ensembl_gene_id, gr_metric, r, p) %>%
    pivot_wider(
      names_from = c(agent, gr_metric),
      values_from = c(r, p),
      names_sep = "_"
    )

  crispr_df_formatted <- crispr_df %>%
    select(ensembl_gene_id, screen, signed_p, `Mann-Whitney p-value`) %>%
    pivot_wider(
      names_from = screen,
      values_from = c(signed_p, `Mann-Whitney p-value`),
      names_sep = "_"
    )

  expression_df_formatted <- expression_df %>%
    select(ensembl_gene_id, cell_line, treatment, {{ge_metric_sym}}, padj) %>%
    pivot_wider(
      names_from = c(treatment, cell_line),
      values_from = c({{ge_metric_sym}}, padj),
      names_sep = "_"
    )

  emm_df_formatted <- emm_df %>%
    select(ensembl_gene_id, cell_line, contrast, {{ge_metric_sym}}, padj) %>%
    pivot_wider(
      names_from = c(contrast, cell_line),
      values_from = c({{ge_metric_sym}}, padj),
      names_sep = "_"
    )

  # Join all datasets
  combined_df <- list(
    cor_df_formatted,
    crispr_df_formatted,
    expression_df_formatted,
    emm_df_formatted
  ) %>%
    reduce(full_join, by = "ensembl_gene_id")

  # Apply criteria filters
  gene_sets <- list()

  if (!is.null(gr_crit)) {
    for (i in seq_along(gr_crit)) {
      genes <- combined_df %>%
        filter(!!rlang::f_rhs(gr_crit[[i]])) %>%
        pull(ensembl_gene_id)
      gene_sets[[paste0("gr_", i)]] <- genes
    }
  }

  if (!is.null(crispr_crit)) {
    for (i in seq_along(crispr_crit)) {
      genes <- combined_df %>%
        filter(!!rlang::f_rhs(crispr_crit[[i]])) %>%
        pull(ensembl_gene_id)
      gene_sets[[paste0("crispr_", i)]] <- genes
    }
  }

  if (!is.null(expression_crit)) {
    for (i in seq_along(expression_crit)) {
      genes <- combined_df %>%
        filter(!!rlang::f_rhs(expression_crit[[i]])) %>%
        pull(ensembl_gene_id)
      gene_sets[[paste0("expression_", i)]] <- genes
    }
  }

  if (!is.null(emm_crit)) {
    for (i in seq_along(emm_crit)) {
      genes <- combined_df %>%
        filter(!!rlang::f_rhs(emm_crit[[i]])) %>%
        pull(ensembl_gene_id)
      gene_sets[[paste0("emm_", i)]] <- genes
    }
  }

  # Combine gene sets
  if (length(gene_sets) == 0) {
    return(character(0))
  } else if (length(gene_sets) == 1) {
    final_genes <- gene_sets[[1]]
  } else {
    if (combine_with == "AND") {
      final_genes <- Reduce(intersect, gene_sets)
    } else {
      final_genes <- Reduce(union, gene_sets)
    }
  }

  # Return results
  list(
    genes = final_genes,
    gene_sets = gene_sets,
    combined_data = combined_df
  )
}


hm_mat <- prep_gigantic_hm(
  gr_cor_agents = c("YKL-5-124"),
  gr_cor_metrics = c("GR50", "GRmax"),
  crispr_screens = c("YKLcrispri", "YKLcrispra"),
  gene_expression_conditions = c("Y1uM", "YKLr1", "YKLr2", "YKLr3", "YKLrTQr1", "YKLrTQr2", "YKLrTQr3"),
  gene_expression_metric = c("signed_p"),
  gene_expression_cell_lines = c("OVCAR4", "SNU8", "OVCAR8", "OVCAR3"),
  gene_expression_emm_contrasts = c("YKLr - no", "YKLrTQr - no"),
  gene_expression_emm_cell_lines = c("all_cells"),
  n_min = 1,
  p_threshold = 0.05
) %>%
  scale_df() %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

crispri_enriched_downregulated <- find_genes_by_pattern(
  crispr_crit = list(~signed_p_YKLcrispri > 1.5),
  # expression_crit = list(~`signed_p_YKLr1_OVCAR4` > -1.5 & `signed_p_YKLr1_OVCAR4` < 1.5),
  emm_crit = list(~`signed_p_YKLr - no_all_cells` < -1.5),
  combine_with = "AND"
)

clipr::write_clip(
  gene_id_name_map[crispri_enriched_downregulated$genes]
)

mat <- hm_mat[crispri_enriched_downregulated$genes, ]

hm_row_clust <- cluster_fun_eucl_imp(mat)

hm <- Heatmap(
  mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_labels = gene_id_name_map[rownames(mat)],
  show_column_names = TRUE
)
hm
withr::with_pdf(
  file.path("plots_cdk7", "gigantic_hm2.pdf"),
  draw(hm),
  width = 15,
  height = 20
)

crispri_depleted_upregulated <- find_genes_by_pattern(
  crispr_crit = list(~signed_p_YKLcrispri < -1.5),
  emm_crit = list(~`signed_p_YKLr - no_all_cells` > 1.5),
  combine_with = "AND"
)

clipr::write_clip(
  gene_id_name_map[crispri_depleted_upregulated$genes]
)

mat <- hm_mat[crispri_depleted_upregulated$genes, ]

hm_row_clust <- cluster_fun_eucl_imp(mat)

hm <- Heatmap(
  mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_labels = gene_id_name_map[rownames(mat)],
  show_column_names = TRUE
)
hm


crispra_enriched_upregulated <- find_genes_by_pattern(
  crispr_crit = list(~signed_p_YKLcrispra > 1.5),
  # expression_crit = list(~`signed_p_YKLr1_OVCAR4` > -1.5 & `signed_p_YKLr1_OVCAR4` < 1.5),
  emm_crit = list(~`signed_p_YKLr - no_all_cells` > 1.5),
  combine_with = "AND"
)

mat <- hm_mat[crispra_enriched_upregulated$genes, ]

hm_row_clust <- cluster_fun_eucl_imp(mat)

hm <- Heatmap(
  mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_labels = gene_id_name_map[rownames(mat)],
  show_column_names = TRUE
)
hm


crispra_depleted_downregulated <- find_genes_by_pattern(
  crispr_crit = list(~signed_p_YKLcrispra < -1.5),
  # expression_crit = list(~`signed_p_YKLr1_OVCAR4` > -1.5 & `signed_p_YKLr1_OVCAR4` < 1.5),
  emm_crit = list(~`signed_p_YKLr - no_all_cells` < -1.5),
  combine_with = "AND"
)

mat <- hm_mat[crispra_depleted_downregulated$genes, ]

hm_row_clust <- cluster_fun_eucl_imp(mat)

hm <- Heatmap(
  mat,
  name = "Z-score",
  na_col = "grey",
  cluster_rows = hm_row_clust,
  cluster_columns = FALSE,
  show_row_names = TRUE,
  row_labels = gene_id_name_map[rownames(mat)],
  show_column_names = TRUE
)
hm
```
