---
title: "CDK7 DESeq2"
author: "Clemens Hug"
date: "2023-07-28"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(data.table)
library(qs)
library(DESeq2)
library(synExtra)
library(IHW)
library(powerjoin)
library(conflicted)

conflicts_prefer(
  dplyr::filter,
  dplyr::select,
  dplyr::mutate,
  dplyr::rename,
  base::setdiff
)

synapser::synLogin()
syn <- synDownloader("~/data", .cache = TRUE)

syn_analysis <- "syn51221227"
```

Reading count and metadata tables generated in `qc.Rmd`.

```{r}
ensembl_gtf_file <- "Homo_sapiens.GRCh38.109.gtf.gz"
if (!file.exists(ensembl_gtf_file)) {
  download.file(
    "ftp://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz",
    ensembl_gtf_file, method = "curl"
  )
}

ensembl_gtf <- rtracklayer::readGFF(
  ensembl_gtf_file
) %>%
  distinct(
    ensembl_gene_id = gene_id,
    hgnc_symbol = gene_name,
    gene_biotype
  ) %>%
  drop_na(ensembl_gene_id) %>%
  mutate(
    gene_name = coalesce(hgnc_symbol, ensembl_gene_id)
  )

barcode_well_map <- syn("syn12979100") %>%
  readxl::read_xlsx(sheet = "barcodes_trugrade_384_set1")
  # rename(Barcode = barcode, Well = well, Plate_ID = plate_id)



prepare_barcode_counts <- function(barcode_count_file, barcode_map) {
  read_tsv(barcode_count_file, col_names = c("barcode", "count")) %>%
    mutate(
      barcode = str_replace(barcode, "^.*([ATCG]{6})$", "\\1")
    ) %>%
    left_join(
      barcode_well_map, by = "barcode"
    )
}

prepare_col_meta <- function(colmeta_file, barcode_map) {
  read_csv(colmeta_file, col_names = "barcode") %>%
    mutate(
      barcode = if (all(str_detect(barcode, fixed(":")))) str_split(barcode, fixed(":"), simplify = TRUE)[, 2] else barcode,
      col_idx = 1:n()
    ) %>%
    left_join(
      barcode_map %>%
        select(barcode, well),
      by = "barcode"
    )
}

prepare_row_meta <- function(rowmeta_file) {
  read_csv(rowmeta_file, col_names = "ensembl_gene_id") %>%
    mutate(row_idx = 1:n())
}

prepare_mtx_sparse <- function(mtx_file) {
  read_delim(
    mtx_file,
    delim = " ",
    skip = 3,
    col_names = c("row_idx", "col_idx", "count"),
    col_types = "iii"
  )
}

prepare_counts <- function(mtx, col_meta, row_meta, gene_mapping) {
  # browser()
  mtx %>%
    left_join(col_meta, by = "col_idx") %>%
    left_join(row_meta, by = "row_idx") %>%
    select(well, ensembl_gene_id, count) %>%
    left_join(
      gene_mapping, by = "ensembl_gene_id"
    )
}


```

## DGE metadata

```{r}
raw_synapse_old <- synGlob(
  "syn52226362", "*", "*mtx*"
)

meta_old <- syn("syn21544269") %>%
  read_csv() %>%
  transmute(
    experiment = "cdk_467_2019_08",
    plate_id = plate,
    batch_id = plate,
    well,
    cell_line,
    agent  = recode(agent, "DMSO" = "control"),
    agent_conc = concentration,
    time = timepoint
  ) %>%
  drop_na(cell_line)

meta_new <- syn("syn51181405") %>%
  read_csv() %>%
  filter(
    source_96 %in% c("Plate 2", "Plate 3"),
    agent != "7 Fluids"
  ) %>%
  transmute(
    experiment = "cdk_467_2023_02",
    plate_id = 1,
    batch_id = str_match(source_96, "Plate ([0-9]+)")[, 2] %>% as.numeric(),
    well = str_replace(well_384, "^([A-Z])([0-9])$", "\\10\\2"),
    cell_line,
    agent  = recode(agent, "DMSO" = "control"),
    agent_conc = concentration,
    time = timepoint
  ) %>%
  mutate(
    agent = if_else(str_detect(cell_line, "YKLr"), "YKLr", agent),
    agent_conc = if_else(str_detect(cell_line, "YKLr"), 1, agent_conc),
    cell_line = str_replace(cell_line, "-YKLr", "")
  )

meta_all <- bind_rows(meta_old, meta_new) %>%
  mutate(
    sample_id = paste(experiment, plate_id, well, sep = "_"),
    batch = paste(experiment, batch_id)
  )

write_csv(
  meta_all,
  file.path("deseq", "dge_combined_meta.csv.gz")
)
```


```{r}
meta_all %>%
  dplyr::count(agent) %>%
  View()

meta_all %>%
  dplyr::count(cell_line, agent) %>%
  View()
```



```{r}
raw_synapse_all <- raw_synapse_old %>%
  enframe("file_name", "synapse_id") %>%
  filter(!str_detect(file_name, "dupes")) %>%
  mutate(
    file_type = case_when(
      str_ends(file_name, ".mtx") ~ "mtx",
      str_ends(file_name, "colnames") ~ "colnames",
      str_ends(file_name, "rownames") ~ "rownames"
    ),
    plate_id = as.double(str_match(file_name, "S([0-9]+)")[, 2]),
    experiment = "cdk_467_2019_08"
  ) %>%
  bind_rows(
    tribble(
      ~synapse_id, ~file_type,
      "syn51181399", "mtx",
      "syn51181400", "colnames",
      "syn51181404", "rownames"
    ) %>%
      mutate(
        plate_id = 1,
        experiment = "cdk_467_2023_02"
      )
  ) %>%
  mutate(
    file_path = map_chr(synapse_id, syn)
  )

files_all <- raw_synapse_all %>%
  pivot_wider(values_from = c(file_name, synapse_id, file_path), names_from = file_type)

counts_raw <- files_all %>%
  mutate(
    col_meta = map(file_path_colnames, ~prepare_col_meta(.x, barcode_map = barcode_well_map)),
    row_meta = map(file_path_rownames, ~prepare_row_meta(.x)),
    mtx = map(file_path_mtx, ~prepare_mtx_sparse(.x)),
    counts = pmap(
      list(mtx, col_meta, row_meta),
      prepare_counts,
      gene_mapping = ensembl_gtf
    )
  )

```

```{r}
counts_all <- counts_raw %>%
  select(experiment, plate_id, counts) %>%
  unnest(counts)

well_counts <- counts_all %>%
  group_by(experiment, plate_id, well) %>%
  summarize(
    total_count = sum(count),
    .groups = "drop"
  ) %>%
  inner_join(
    meta_all,
    by = c("experiment", "plate_id", "well")
  )

samples_to_keep <- well_counts %>%
  filter(total_count > 80000)

genes_to_keep <- counts_all %>%
  group_by(
    ensembl_gene_id
  ) %>%
  summarize(
    keep = sum(count > 5) > 5,
    .groups = "drop"
  )
```

```{r}
meta_kept <- meta_all %>%
  semi_join(
    samples_to_keep,
    by = c("experiment", "plate_id", "well")
  )

conditions_kept <- meta_kept %>%
  distinct(experiment, plate_id, batch_id,cell_line, agent, time) %>%
  mutate(
    condition_id = paste(experiment, plate_id, batch_id, cell_line, agent, time, sep = "_")
  )

counts_kept <- counts_all %>%
  semi_join(
    samples_to_keep,
    by = c("experiment", "plate_id", "well")
  ) %>%
  semi_join(
    genes_to_keep %>%
      filter(keep),
    by = "ensembl_gene_id"
  ) %>%
  inner_join(
    meta_all %>%
      select(experiment, plate_id, well, sample_id),
    by = c("experiment", "plate_id", "well")
  )
```

Visualize experimental design
```{r}
meta_kept %>%
  mutate(
    batch = paste(experiment, batch_id, sep = "_")
  ) %>%
  dplyr::count(
    experiment, batch, cell_line, agent, time, agent_conc
  ) %>%
  ggplot(
    aes(
      agent,
      fct_inseq(as.character(agent_conc)),
      fill = batch
    )
  ) +
  geom_tile() +
  facet_grid(cell_line ~ time) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

```


```{r}
counts_mat <- counts_kept %>%
  select(sample_id, ensembl_gene_id, count) %>%
  pivot_wider(names_from = sample_id, values_from = count, values_fill = 0L) %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

size_factors <- DESeq2::estimateSizeFactorsForMatrix(counts_mat)

counts_mat_corrected <- t(t(counts_mat) / size_factors)

counts_corrected <- counts_mat_corrected %>%
  as_tibble(rownames = "ensembl_gene_id") %>%
  pivot_longer(
    cols = -ensembl_gene_id,
    names_to = "sample_id",
    values_to = "count"
  )
```


```{r}
meta_deseq <- meta_kept %>%
  mutate(
    across(time, ~as.factor(as.character(.x))),
    across(batch_id, ~as.factor(as.character(.x)))
  ) %>%
  arrange(match(sample_id, colnames(counts_mat))) %>%
  column_to_rownames("sample_id")
```


Variance stabilizing transformation

```{r}
dds <- DESeqDataSetFromMatrix(
  counts_mat,
  colData = meta_deseq,
  design = ~experiment + agent + cell_line + time
) %>%
  estimateSizeFactors() %>%
  estimateDispersions()

dge_all_counts_long <- list(
  raw = counts(dds, normalized = FALSE),
  normalized = counts(dds, normalized = TRUE),
  varstab = varianceStabilizingTransformation(dds) %>%
    assay()
) %>%
  map(\(x) as_tibble(x, rownames = "ensembl_gene_id")) %>%
  bind_rows(.id = "type") %>%
  left_join(
    ensembl_gtf %>%
      select(ensembl_gene_id, hgnc_symbol, gene_name),
    by = "ensembl_gene_id"
  )

write_csv(
  dge_all_counts_long,
  file.path("deseq", "dge_counts_long.csv.gz")
)

```

```{r}
pca_vsd <- vsd_mat %>%
  t() %>%
  prcomp()

pca_vsd_df <- pca_vsd$x[, 1:10] %>%
  as_tibble(rownames = "sample_id") %>%
  inner_join(
    meta_deseq %>%
      rownames_to_column("sample_id"),
    by = "sample_id"
  )

p <- ggplot(
  pca_vsd_df,
  aes(x = PC1, y = PC2, color = cell_line, shape = experiment)
) +
  geom_point()
```

Overlapping conditions between batches

```{r}
meta_batch_correction <- meta_kept %>%
  semi_join(
    meta_kept %>%
      distinct(experiment, cell_line, agent) %>%
      filter(agent == "control") %>%
      group_by(cell_line, agent) %>%
      filter(n() > 1)
  )
```

Remove batch effects using ComBat

```{r}
library(sva)

source("combat_with_reference.R")
combat_res <- combat_with_reference(
  counts_mat[, meta_batch_correction$sample_id],
  batch = with(
    meta_batch_correction,
    paste(experiment, sep = "_")
  ),
  covar_mod = model.matrix(
    ~cell_line,
    data = meta_batch_correction
  ),
  covar_mod_all = model.matrix(
    ~cell_line + agent,
    data = meta_deseq
  ),
  batch_all = with(
    meta_deseq,
    paste(experiment, sep = "_")
  ),
  counts_all = counts_mat
)
```

```{r}
dds_bc <- DESeqDataSetFromMatrix(
  combat_res,
  colData = meta_deseq,
  design = ~experiment + agent + cell_line + time
)

vsd_bc <- varianceStabilizingTransformation(dds_bc)

vsd_mat_bc <- assay(vsd_bc)
```

```{r}
pca_vsd_bc <- vsd_mat_bc %>%
  t() %>%
  prcomp()

pca_vsd_bc_df <- pca_vsd_bc$x[, 1:10] %>%
  as_tibble(rownames = "sample_id") %>%
  inner_join(
    meta_deseq %>%
      rownames_to_column("sample_id"),
    by = "sample_id"
  )

p <- ggplot(
  pca_vsd_bc_df,
  aes(x = PC1, y = PC2, color = cell_line, shape = experiment)
) +
  geom_point()
```

Much improved batch effect

```{r}
extract_deseq_result <- function(de, contrast, name, res_args = list(), shrink_args = list()) {
  if (missing(contrast)) {
    res <- rlang::exec(results, de, name = name, !!!res_args)
  } else {
    res <- rlang::exec(results, de, contrast = contrast, !!!res_args)
  }
  if (!"type" %in% names(shrink_args)) {
    shrink_args$type <- "ashr"
  }
  shrunken <- rlang::exec(lfcShrink, de, res = res, !!!shrink_args)
  shrunken %>%
    as_tibble(rownames = "gene_id") %>%
    left_join(
      res %>%
        as_tibble(rownames = "gene_id") %>%
        select(gene_id, log2FoldChange, lfcSE),
      by = "gene_id", suffix = c("", "_MLE")
    )
}
```

No batch correction necessary here. Only 4 conditions where that would even
be possible. Otherwise each condition is unique to a batch.

```{r}
deseq_input <- meta_kept %>%
  filter(agent != "control") %>%
  group_nest(experiment, agent, cell_line, time, .key = "treated") %>%
  left_join(
    meta_kept %>%
      filter(agent == "control") %>%
      group_nest(experiment, cell_line, time, .key = "control"),
    by = c("experiment", "cell_line", "time")
  ) %>%
  mutate(
    control = map2(
      control, treated,
      \(x, y) filter(x, batch %in% y$batch)
    )
  ) %>%
  filter(
    map_int(treated, nrow) > 1,
    map_int(control, nrow) > 1
  ) %>%
  mutate(
    meta_deseq = map2(
      treated, control,
      ~bind_rows(.x, .y) %>%
        mutate(
          log_agent_conc = log10(agent_conc + 1)
        )
    ),
    batch_correct = map_lgl(meta_deseq, \(x) n_distinct(x$batch) > 1)
  )

deseq_fit <- deseq_input %>%
  mutate(
    deseq = map2(
      meta_deseq, batch_correct,
      \(m, bc) {
        DESeqDataSetFromMatrix(
          counts_mat[, m$sample_id],
          colData = column_to_rownames(m, "sample_id"),
          # design = ~log_agent_conc
          design = if (bc) {~log_agent_conc + batch} else {~log_agent_conc}
        ) %>%
          DESeq(test = "Wald")
      }
    )
  )

qsave(
  deseq_fit,
  file.path("deseq", "cdk7_deseq_res_raw2.qs")
)
# deseq_fit <- qread("deseq/cdk7_deseq_res_raw2.qs")

deseq_res <- deseq_fit %>%
  rowwise() %>%
  mutate(
    res = {
      extract_deseq_result(
        deseq, name = "log_agent_conc"
      ) %>%
        dplyr::rename(
          ensembl_gene_id = gene_id
        ) %>%
        left_join(
          ensembl_gtf %>%
            select(ensembl_gene_id, hgnc_symbol),
          by = "ensembl_gene_id"
        ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  deseq_res %>%
    select(-deseq),
  file.path("deseq", "cdk7_deseq_res2.qs")
)
# deseq_res <- qread("deseq/cdk7_deseq_res2.qs")
```

```{r}
deseq_fit_lrt <- deseq_input %>%
  rowwise() %>%
  mutate(
    meta_deseq = meta_deseq %>%
      mutate(
        agent_conc_fct = fct_inseq(as.character(agent_conc))
      ) %>%
      list(),
    deseq = DESeqDataSetFromMatrix(
      counts_mat[, meta_deseq$sample_id],
      colData = column_to_rownames(meta_deseq, "sample_id"),
      design = if (batch_correct) {~agent_conc_fct + batch} else {~agent_conc_fct}
    ) %>%
      DESeq(
        test = "LRT",
        reduced = if (batch_correct) {~batch} else {~1}
      ) %>%
      list()
  ) %>%
  ungroup()

qsave(
  deseq_fit_lrt,
  file.path("deseq", "cdk7_deseq_res_lrt_raw2.qs")
)
# deseq_fit_lrt <- qread("deseq/cdk7_deseq_res_lrt_raw2.qs")

deseq_res_lrt <- deseq_fit_lrt %>%
  rowwise() %>%
  mutate(
    res = results(
      deseq, tidy = TRUE
    ) %>%
      dplyr::rename(ensembl_gene_id = row) %>%
      left_join(
        ensembl_gtf %>%
          select(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      as_tibble() %>%
      list()
  ) %>%
  ungroup()
```

```{r}
deseq_res_lrt_long <- deseq_res_lrt %>%
  select(experiment, agent, cell_line, time, res) %>%
  unnest(res)

deseq_res_long <- deseq_res %>%
  select(experiment, agent, cell_line, time, res) %>%
  unnest(res)

write_csv(
  deseq_res_long,
  file.path("deseq", "cdk7_deseq_res_long.csv")
)
# deseq_res_long <- read_csv(
#   file.path("deseq", "cdk7_deseq_res_long.csv")
# )

write_csv(
  deseq_res_lrt_long,
  file.path("deseq", "cdk7_deseq_res_lrt_long.csv")
)
# deseq_res_lrt_long <- read_csv(
#   file.path("deseq", "cdk7_deseq_res_lrt_long.csv")
# )
```


Combined

```{r}
deseq_res_combined <- deseq_res_long %>%
  powerjoin::power_inner_join(
    deseq_res_lrt_long %>%
    select(
      experiment,
      agent, cell_line, time,
      ensembl_gene_id,
      pvalue_lrt = pvalue,
      padj_lrt = padj
    ),
    by = c(
      "experiment",
      "agent", "cell_line", "time",
      "ensembl_gene_id"
    ),
    check = powerjoin::check_specs(
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn",
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn"
    )
  )

write_csv(
  deseq_res_combined,
  file.path("deseq", "cdk7_deseq_res_combined_long.csv")
)
```

Upload DGE results to Synapse

```{r}
syn_analysis <- synMkdir("syn52226362", "analysis")

synStoreMany(
  file.path(
    "deseq",
    c(
      "cdk7_deseq_res_long.csv",
      "cdk7_deseq_res_lrt_long.csv",
      "cdk7_deseq_res_combined_long.csv",
      "dge_combined_meta.csv.gz",
      "dge_counts_long.csv.gz"
    )
  ),
  syn_analysis,
  forceVersion = FALSE
)

```


Checking goodness of fit of the log-linear model using advice
from https://support.bioconductor.org/p/117448

```{r}

x <- deseq_fit$deseq[[1]]
y <- mcols(x)$deviance - 2*rowSums(dnbinom(counts(x), mu=counts(x), size=1/dispersions(x), log=TRUE))

counts_corrected %>%
  filter(
    ensembl_gene_id %in% names(head(sort(y, decreasing = TRUE))),
    sample_id %in% colnames(x)
  ) %>%
  inner_join(
    deseq_fit$meta_deseq[[1]] %>%
      rownames_to_column("sample_id")
  ) %>%
  ggplot(
    aes(log_agent_conc, count, color = agent)
  ) +
    geom_point() +
    facet_wrap(~ensembl_gene_id, scales = "free_y")
```

Apeglm leads to some weird artifacts where the log2FoldChange is not shrunk but
expanded multifold. Using ashr instead.

```{r}
p <- deseq_res_long %>%
  filter(
    experiment == "cdk_467_2019_08",
    agent == "YKL-5-124"
  ) %>%
  ggplot(
    aes(log2FoldChange, -log10(padj))
  ) +
    ggrastr::rasterize(
      geom_point(shape = 16, alpha = 0.3),
      dev = "ragg", dpi = 200
    ) +
    facet_wrap(~cell_line, scale = "free")

ggsave(
  file.path("plots_cdk7", "ykl_lfc_padj.pdf"),
  p,
  width = 10,
  height = 8
)
```


```{r}
deseq_res_long %>%
  group_by(experiment, plate_id, agent, cell_line, time) %>%
  summarize(
    n_sig = sum(padj < 0.05, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  ggplot(aes("0", n_sig, color = cell_line)) +
    ggbeeswarm::geom_quasirandom() +
    facet_wrap(~experiment + plate_id)

deseq_res_lrt_long %>%
  group_by(experiment, plate_id, agent, cell_line, time) %>%
  summarize(
    n_sig = sum(padj < 0.05, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  ggplot(aes("0", n_sig, color = cell_line)) +
    ggbeeswarm::geom_quasirandom() +
    facet_wrap(~experiment + plate_id)

deseq_n_res_lin_vs_lrt <- deseq_res_long %>%
  group_by(experiment, plate_id, agent, cell_line, time) %>%
  summarize(
    n_sig_lin = sum(padj < 0.05, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  left_join(
    deseq_res_lrt_long %>%
      group_by(experiment, plate_id, agent, cell_line, time) %>%
      summarize(
        n_sig_lrt = sum(padj < 0.05, na.rm = TRUE),
        .groups = "drop"
      )
  )

p <- deseq_n_res_lin_vs_lrt %>%
  ggplot(
    aes(n_sig_lin, n_sig_lrt, color = cell_line)
  ) +
    geom_point() +
    facet_wrap(~experiment + plate_id) +
    coord_equal() +
    geom_abline(slope = 1, intercept = 0) +
    scale_x_continuous(
      trans = scales::pseudo_log_trans(base = 10),
      breaks = function(...) {
        l <- list(...)
        l[[1]][1] <- max(l[[1]][1], 1)
        x <- rlang::exec(scales::breaks_log(), !!!l)
        c(0, x)
      }
    ) +
    scale_y_continuous(
      trans = scales::pseudo_log_trans(base = 10),
      breaks = function(...) {
        l <- list(...)
        l[[1]][1] <- max(l[[1]][1], 1)
        x <- rlang::exec(scales::breaks_log(), !!!l)
        c(0, x)
      }
    )
ggsave(
  file.path("plots_cdk7", "n_sig_lin_vs_lrt.pdf"),
  p,
  width = 10,
  height = 8
)

deseq_n_res_lin_vs_lrt %>% filter(n_sig_lrt > 10*n_sig_lin)

deseq_n_res_lin_vs_lrt %>% filter(n_sig_lrt > 10*n_sig_lin) %>%
  dplyr::slice(2) %>%
  inner_join(
    deseq_res_lrt %>%
      select(experiment, plate_id, agent, cell_line, time, res, meta_deseq)
  ) %>%
  mutate(
    across(res, map, \(x) filter(x, padj < 0.05)),
    across(meta_deseq, map, \(x) select(x, sample_id, log_agent_conc))
  ) %>%
  unnest(res) %>%
  unnest(meta_deseq) %>%
  inner_join(
    counts_corrected
  ) %>%
  ggplot(
    aes(log_agent_conc, count, color = agent)
  ) +
    geom_point() +
    facet_wrap(~ensembl_gene_id, scales = "free_y")

counts_corrected %>%
  inner_join(
    deseq_res_lrt_long %>%
      inner_join(
        deseq_n_res_lin_vs_lrt %>% filter(n_sig_lrt > 10*n_sig_lin)
      ) %>%
      filter(padj < 0.05) %>%
      inner_join(
        meta_kept
      )
  ) %>%
  ggplot(
    aes(agent_conc, count, color = agent)
  ) +
    geom_point() +
    facet_wrap(~ensembl_gene_id, scales = "free_y")
```


```{r}


deseq_res_lfc_shrunk_wide <- deseq_res_long %>%
  inner_join(
    conditions_kept
  ) %>%
  select(condition_id, ensembl_gene_id, log2FoldChange) %>%
  mutate(across(log2FoldChange, ~replace_na(.x, 0))) %>%
  pivot_wider(names_from = condition_id, values_from = log2FoldChange, values_fill = 0)

deseq_res_lfc_shrunk_wide_mat <- deseq_res_lfc_shrunk_wide %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()
```


```{r}
pca_lfc_shrunk <- deseq_res_lfc_shrunk_wide_mat[
  deseq_res_long %>%
    group_by(ensembl_gene_id) %>%
    filter(any(padj < 0.05)) %>%
    ungroup() %>%
    pull(ensembl_gene_id) %>%
    unique(),
] %>%
  t() %>%
  prcomp()

pca_lfc_shrunk_df <- pca_lfc_shrunk$x[, 1:10] %>%
  as_tibble(rownames = "condition_id") %>%
  inner_join(
    conditions_kept,
    by = "condition_id"
  )

p <- ggplot(
  pca_lfc_shrunk_df,
  aes(x = PC1, y = PC2, color = cell_line, shape = experiment)
) +
  geom_point()
```

## Baseline DGE

```{r read_gr_metrics}
gr_raw <- syn("syn52242760") %>%
  read_csv() %>%
  mutate(neglog_GR50 = -log10(GR50))

baseline_meta <- meta_kept %>%
  filter(
    agent == "control",
    experiment == "cdk_467_2019_08",
    cell_line %in% gr_raw$cell_line
  )

baseline_cell_overlap <- unique(baseline_meta$cell_line)
```


```{r}
de_baseline <- DESeqDataSetFromMatrix(
  counts_mat[, baseline_meta$sample_id],
  colData = column_to_rownames(baseline_meta, "sample_id"),
  design = ~0 + cell_line
) %>%
  DESeq()

qsave(
  de_baseline,
  file.path("deseq", "cdk7_deseq_baseline.qs")
)
```

```{r}
dge_counts_baseline_trans <- list(

)
```


```{r}
library(emmeans)
de_baseline_emm <- lm(
  count ~ 0 + cell_line,
  data = mutate(
    baseline_meta,
    count = counts_mat["ENSG00000000003", sample_id]
  )
) %>%
  emmeans(~cell_line)

de_baseline_contrasts <- contrast(
  de_baseline_emm,
  method = "eff",
  adjust = "none"
)
```
de_baseline_contrasts@linfct
      cell_line59M cell_lineCOV318 cell_lineCOV362 cell_lineCOV504 cell_lineFUOV1 cell_lineKuramochi cell_lineONCODG1 cell_lineOVCAR3 cell_lineOVCAR8 cell_lineSKOV3 cell_lineSNU8
 [1,]   0.90909091     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [2,]  -0.09090909      0.90909091     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [3,]  -0.09090909     -0.09090909      0.90909091     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [4,]  -0.09090909     -0.09090909     -0.09090909      0.90909091    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [5,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909     0.90909091        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [6,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909         0.90909091      -0.09090909     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [7,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909       0.90909091     -0.09090909     -0.09090909    -0.09090909   -0.09090909
 [8,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909      0.90909091     -0.09090909    -0.09090909   -0.09090909
 [9,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909      0.90909091    -0.09090909   -0.09090909
[10,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909     0.90909091   -0.09090909
[11,]  -0.09090909     -0.09090909     -0.09090909     -0.09090909    -0.09090909        -0.09090909      -0.09090909     -0.09090909     -0.09090909    -0.09090909    0.90909091

So this approach below is correct. From the emmeans doc:
eff.emmc gives weights (k-1)/k and -1/k respectively, as in subtracting the overall EMM from each EMM

```{r}
de_baseline_res <- resultsNames(de_baseline) %>%
  set_names() %>%
  map(
    ~{
      n <- length(resultsNames(de_baseline))
      contrast_vec <- rep(
        -1/n,
        n
      ) %>%
        set_names(resultsNames(de_baseline))
      contrast_vec[match(.x, names(contrast_vec))] <- (n - 1) / n
      r <- results(de_baseline, contrast = contrast_vec)
      left_join(
        as.data.frame(r) %>%
          as_tibble(rownames = "ensembl_gene_id"),
        lfcShrink(
          de_baseline, contrast = contrast_vec, res = r, type = "ashr"
        ) %>%
          as_tibble(rownames = "ensembl_gene_id") %>%
          select(ensembl_gene_id, log2FoldChange),
        by = "ensembl_gene_id",
        suffix = c("_MLE", "")
      ) %>%
        left_join(
          ensembl_gtf %>%
            select(ensembl_gene_id, hgnc_symbol),
          by = "ensembl_gene_id"
        )
    }
  ) %>%
  bind_rows(.id = "cell_line") %>%
  mutate(across(cell_line, \(x) str_replace(x, "cell_line", "")))

write_csv(
  de_baseline_res,
  file.path("deseq", "cdk7_deseq_baseline_res.csv")
)
# de_baseline_res <- read_csv(
#   file.path("deseq", "cdk7_deseq_baseline_res.csv")
# )
```

```{r}
synStoreMany(
  c(
    file.path("deseq", "cdk7_deseq_baseline_res.csv")
  ),
  syn_analysis,
  forceVersion = FALSE
)

```

## Deep RNA-seq baseline

Also some at /n/files/ImStor/sorger/data/rnaseq/lincs/baseline_phase2



```{r}
deep_counts <- syn("syn60529544") %>%
  read_csv()

deep_meta <- syn("syn60529546") %>%
  readxl::read_excel() %>%
  mutate(
    sample_id = paste0("S", Sample)
  ) %>%
  dplyr::rename(cell_line = `Cell line`)

# Used Ensembl gene annotation version 92
ensembl_gtf_file_old <- "Homo_sapiens.GRCh38.92.gtf.gz"
if (!file.exists(ensembl_gtf_file_old)) {
  download.file(
    "ftp://ftp.ensembl.org/pub/release-92/gtf/homo_sapiens/Homo_sapiens.GRCh38.92.gtf.gz",
    ensembl_gtf_file_old, method = "curl"
  )
}

ensembl_gtf_old <- rtracklayer::readGFF(
  ensembl_gtf_file_old
) %>%
  distinct(
    ensembl_gene_id = gene_id,
    hgnc_symbol = gene_name,
    gene_biotype
  ) %>%
  drop_na(ensembl_gene_id) %>%
  mutate(
    gene_name = coalesce(hgnc_symbol, ensembl_gene_id)
  )


ensembl_gtf_combined <- bind_rows(
  dge = ensembl_gtf,
  deep = ensembl_gtf_old,
  .id = "assay"
)

write_csv(
  ensembl_gtf_combined,
  file.path("data", "ensembl_gtf_combined.csv.gz")
)
```


### Count correction

```{r}
deep_counts_mat <- deep_counts %>%
  semi_join(
    ensembl_gtf_old %>%
      filter(gene_biotype == "protein_coding"),
    by = c("gene" = "ensembl_gene_id")
  ) %>%
  mutate(
    across(
      -gene,
      as.integer
    )
  ) %>%
  column_to_rownames("gene") %>%
  as.matrix()

deep_size_factors <- estimateSizeFactorsForMatrix(deep_counts_mat)
deep_counts_mat_corrected <- t(t(deep_counts_mat) / deep_size_factors)


deep_counts_corrected <- deep_counts_mat_corrected %>%
  as_tibble(rownames = "ensembl_gene_id") %>%
  pivot_longer(
    cols = -ensembl_gene_id,
    names_to = "sample_id",
    values_to = "count"
  )

deep_counts_varstab_mat <- rlogTransformation(deep_counts_mat)
deep_counts_varstab <- deep_counts_varstab_mat %>%
  as_tibble(rownames = "ensembl_gene_id") %>%
  pivot_longer(
    cols = -ensembl_gene_id,
    names_to = "sample_id",
    values_to = "count"
  )

deep_all_counts_long <- bind_rows(
  raw = deep_counts %>%
    dplyr::rename(ensembl_gene_id = gene) %>%
    semi_join(
      ensembl_gtf_old %>%
        filter(gene_biotype == "protein_coding"),
      by = c("ensembl_gene_id")
    ) %>%
    pivot_longer(
      cols = -ensembl_gene_id,
      names_to = "sample_id",
      values_to = "count"
    ),
  normalized = deep_counts_corrected,
  varstab = deep_counts_varstab,
  .id = "type"
) %>%
  left_join(
    ensembl_gtf_old %>%
      select(ensembl_gene_id, hgnc_symbol, gene_name),
    by = "ensembl_gene_id"
  ) %>%
  left_join(
    deep_meta %>%
      select(sample_id, cell_line),
    by = "sample_id"
  )

write_csv(
  deep_all_counts_long,
  file.path("deseq", "deep_baseline_counts_long.csv.gz")
)

# deep_deseq <- DESeqDataSetFromMatrix(
#   deep_counts_mat[, deep_meta$sample_id],
#   deep_meta %>%
#     column_to_rownames("sample_id"),
#   design = ~0 + cell_line
# ) %>%
#   DESeq()
```

### Naive log2FC

Computing naive log2FC for each cell line compared to the average of all other
using the rlog transformed counts.

```{r}
deep_counts_varstab_means <- rowMeans(deep_counts_varstab_mat)
deep_counts_varstab_differences <- apply(
  deep_counts_varstab_mat,
  2,
  \(x) x - deep_counts_varstab_means
) %>%
  as_tibble(rownames = "ensembl_gene_id")

deep_counts_varstab_zscores <- t(scale(t(deep_counts_varstab_mat))) %>%
  as_tibble(rownames = "ensembl_gene_id")
deep_counts_varstab_mad <- apply(
  deep_counts_varstab_mat,
  1,
  mad
) %>%
  enframe("ensembl_gene_id", "mad")

write_csv(
  deep_counts_varstab_zscores,
  file.path("deseq", "deep_baseline_varstab_zscores.csv.gz")
)
write_csv(
  deep_counts_varstab_mad,
  file.path("deseq", "deep_baseline_varstab_mad.csv.gz")
)
write_csv(
  deep_counts_varstab_differences,
  file.path("deseq", "deep_baseline_varstab_differences.csv.gz")
)

deep_counts_mat_corrected_means <- rowMeans(deep_counts_mat_corrected)
deep_counts_naive_log2fc <- apply(
  deep_counts_mat_corrected, 2,
  \(x) log2(x / deep_counts_mat_corrected_means)
) %>%
  as_tibble(rownames = "ensembl_gene_id")

write_csv(
  deep_counts_naive_log2fc,
  file.path("deseq", "deep_baseline_naive_log2fc.csv.gz")
)
```

### EdgeR

Can't do DESeq2 because no replicates. Use edgeR with estimated dispersion of .4
instead. .4 recommended in manual for human samples.

```{r}
library(edgeR)
deep_dge_list <- DGEList(
  counts = deep_counts_mat[, deep_meta$sample_id],
  # genes = tibble(ensembl_gene_id = rownames(deep_counts_mat)),
  samples = deep_meta %>%
    select(cell_line)
) %>%
  normLibSizes()

deep_design <- model.matrix(~0 + cell_line, data = deep_meta) %>%
  magrittr::set_rownames(deep_meta$sample_id)

deep_glm <- glmFit(
  deep_dge_list, deep_design,
  dispersion = .4
)

deep_results <- deep_meta %>%
  distinct(cell_line) %>%
  mutate(
    res = map(
      cell_line,
      \(x) {
        n <- ncol(deep_design)
        contrast_vec <- rep(
          -1 / n,
          n
        )
        contrast_vec[match(paste0("cell_line", x), colnames(deep_design))] <- (n - 1) / n
        glmLRT(
          deep_glm,
          contrast = contrast_vec
        ) %>%
          topTags(n = Inf) %>%
          as.data.frame() %>%
          as_tibble()
      }
    )
  )

deep_results_long_new <- deep_results %>%
  select(cell_line, res) %>%
  unnest(res)

write_csv(
  deep_results_long,
  file.path("deseq", "deep_baseline_edger_res.csv.gz")
)
deep_results_long <- read_csv(
  file.path("deseq", "deep_baseline_edger_res.csv.gz")
)
```



```{r}
p <- deep_results_long %>%
  ggplot(
    aes(logFC, -log10(FDR))
  ) +
    ggrastr::rasterize(geom_point(), dev = "ragg", dpi = 200) +
    facet_wrap(~cell_line)

```


```{r}
synStoreMany(
  c(
    file.path("deseq", "deep_baseline_counts_long.csv.gz"),
    file.path("deseq", "deep_baseline_varstab_zscores.csv.gz"),
    file.path("deseq", "deep_baseline_varstab_mad.csv.gz"),
    file.path("deseq", "deep_baseline_naive_log2fc.csv.gz"),
    file.path("deseq", "deep_baseline_varstab_differences.csv.gz"),
    file.path("data", "ensembl_gtf_combined.csv.gz")
  ),
  parentId = "syn68500169",
  forceVersion = FALSE
)

```


## Trying to regress out effects of cell growth differences between cell lines

YKL-5-124
JWZ, JWZ-5-21 # degrader with YKL warhead
LY3405105
LDC4297
CT7001
SY5609, SY-5609

```{r}
gr_all <- syn("syn60582220") %>%
  read_csv() %>%
  mutate(
    agent = recode(
      agent, JWZ = "JWZ-5-21", SY5609 = "SY-5609"
    ),
    doubling_per_hour = (log2(cell_count) - log2(cell_count__time0)) / timepoint,
    mult_per_hour = 2**doubling_per_hour
  )

gr_controls <- gr_all %>%
  distinct(cell_line, cell_count__time0, cell_count__ctrl) %>%
  mutate(
    doubling_per_hour = (log2(cell_count__ctrl) - log2(cell_count__time0)) / 24,
    mult_per_hour = 2**doubling_per_hour
  )

growth_data <- gr_all %>%
  bind_rows(
    gr_controls %>%
      mutate(
        agent = "control",
        concentration = 0
      )
  ) %>%
  group_by(
    cell_line, agent, concentration
  ) %>%
  summarize(
    doubling_per_hour = mean(doubling_per_hour),
    mult_per_hour = 10**mean(log10(mult_per_hour)),
    across(starts_with("cell_count"), mean),
    .groups = "drop"
  )

meta_deseq_gr <- meta_deseq %>%
  rownames_to_column("sample_id") %>%
  power_inner_join(
    growth_data %>%
      dplyr::rename(agent_conc_gr = concentration),
    by = join_by(cell_line, agent),
    check = check_specs(
      unmatched_keys_left = "warn"
      # duplicate_keys_right = "warn"
    )
  ) %>%
  group_by(
    cell_line, agent, agent_conc
  ) %>%
  filter(abs(agent_conc - agent_conc_gr) == min(abs(agent_conc - agent_conc_gr))) %>%
  ungroup() %>%
  filter((agent_conc == agent_conc_gr) | (abs(log2(agent_conc) - log2(agent_conc_gr)) < log2(1.3)))

```

```{r}
deseq_input_gr <- meta_deseq_gr %>%
  filter(agent != "control") %>%
  group_nest(experiment, time, .key = "treated") %>%
  left_join(
    meta_deseq_gr %>%
      filter(agent == "control") %>%
      group_nest(experiment, time, .key = "control"),
    by = c("experiment", "time")
  ) %>%
  # mutate(
  #   control = map2(
  #     control, treated,
  #     ~filter(.x, batch %in% .y$batch)
  #   )
  # ) %>%
  filter(
    map_int(treated, nrow) > 1,
    map_int(control, nrow) > 1
  ) %>%
  mutate(
    meta_deseq = map2(
      treated, control,
      ~bind_rows(.x, .y) %>%
        mutate(
          log_agent_conc = log10(agent_conc + 1)
        )
    ),
    meta_deseq_ready = map(
      meta_deseq,
      \(x) mutate(
          x,
          agent = if_else(
            agent == "control",
            setdiff(agent, "control")[1],
            agent
          ) %>%
            factor()
        )
    )
    # batch_correct = map_lgl(meta_deseq, \(x) n_distinct(x$batch) > 1)
  )

make_mm <- function(f, data) {
  mm <- model.matrix(f, data = data)
  # REmove all zero columns
  mm[, colSums(mm) > 0]
}

deseq_fit_gr <- deseq_input_gr %>%
  mutate(
    deseq = map(
      meta_deseq,
      \(m) {
        DESeqDataSetFromMatrix(
          counts_mat[, m$sample_id],
          colData = column_to_rownames(m, "sample_id"),
          design = ~1,
          # design = if (bc) {~log_agent_conc + batch} else {~log_agent_conc}
        )
      }
    ),
    mm = map2(
      deseq, meta_deseq_ready,
      \(d, m) {
        make_mm(
          if (n_distinct(m$agent) > 2)
            ~0 + log_agent_conc:agent:cell_line + cell_line + doubling_per_hour
          else
            ~0 + log_agent_conc:cell_line + cell_line + doubling_per_hour,
          m
        )
      }
    )
  )

deseq_fit_gr <- deseq_fit_gr %>%
  dplyr::slice(2) %>%
  mutate(
    deseq = map2(
      deseq, mm,
      \(d, mm) {
        d <- DESeq(
          d,
          test = "Wald",
          full = mm
        )
      }
    )
  )

qsave(
  deseq_fit_gr,
  file.path("deseq", "cdk7_deseq_gr_res_raw.qs")
)
# deseq_fit_gr <- qread("deseq/cdk7_deseq_gr_res_raw.qs")

deseq_res_gr <- deseq_fit_gr %>%
  rowwise() %>%
  mutate(
    res = {
      res_names <- resultsNames(deseq)
      contrast_vec <- rep_len(0, length(res_names))
      map(
        set_names(res_names),
        \(x) {
          message("Extracting ", x)
          contrast_vec_ <- contrast_vec
          contrast_vec_[match(x, res_names)] <- 1
          # browser()
          extract_deseq_result(
            deseq, contrast = contrast_vec_
          )
        }
      ) %>%
        bind_rows(.id = "covariate") %>%
        dplyr::rename(
          ensembl_gene_id = gene_id
        ) %>%
        left_join(
          ensembl_gtf %>%
            select(ensembl_gene_id, hgnc_symbol),
          by = "ensembl_gene_id"
        ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  deseq_res_gr %>%
    select(-deseq),
  file.path("deseq", "cdk7_deseq_gr_res.qs")
)
# deseq_res_gr <- qread("deseq/cdk7_deseq_gr_res.qs")
```

```{r}
deseq_res_gr_long <- deseq_res_gr %>%
  mutate(
    res = map(
      res,
      \(x) filter(
        x,
        str_starts(covariate, fixed("log_agent_conc"))
      ) %>%
        separate_wider_regex(
          covariate,
          patterns = c(
            "log_agent_conc(?:\\.agent)?", agent = ".*", "\\.cell_line", cell_line = ".+"
          ),
          cols_remove = FALSE
        )
    )
  ) %>%
  select(experiment, time, res) %>%
  unnest(res) %>%
  mutate(
    # empty agent is YKL. Not in the formula because it's the only one
    agent = if_else(agent == "", "YKL-5-124", agent) %>%
    # Deseq replaces dashes with dots, reversing
      str_replace_all(fixed("."), "-")
  )

write_csv(
  deseq_res_gr_long,
  file.path("deseq", "cdk7_deseq_gr_res_long.csv.gz")
)
# deseq_res_gr_long <- read_csv(file.path("deseq", "cdk7_deseq_gr_res_long.csv.gz"))
```


```{r}
deseq_res_gr_and_normal <- power_inner_join(
  deseq_res_long,
  deseq_res_gr_long %>%
    mutate(time = as.double(as.character(time))),
  by = c(
    "experiment", "time", "cell_line", "agent", "ensembl_gene_id", "hgnc_symbol"
  ),
  suffix = c("_normal", "_gr"),
  check = check_specs(
    duplicate_keys_left = "warn",
    duplicate_keys_right = "warn",
    # unmatched_keys_left = "warn",
    unmatched_keys_right = "warn"
  )
)

write_csv(
  deseq_res_gr_and_normal,
  file.path("deseq", "cdk7_deseq_gr_and_normal_res_long.csv.gz")
)
# deseq_res_gr_and_normal <- read_csv(
#   file.path("deseq", "cdk7_deseq_gr_and_normal_res_long.csv.gz")
# )
```

```{r}
synStoreMany(
  c(
    file.path("deseq", "cdk7_deseq_gr_and_normal_res_long.csv.gz"),
    file.path("deseq", "cdk7_deseq_gr_res_long.csv.gz")
  ),
  parentId = "syn59470136",
  forceVersion = FALSE
)

```


```{r}
p <- ggplot(
  deseq_res_gr_and_normal,
  aes(
    log2FoldChange_normal, log2FoldChange_gr,
    color = agent
  )
) +
  geom_point(alpha = .5) +
  geom_abline(slope = 1, intercept = 0) +
  facet_wrap(~experiment + cell_line)



p <- ggplot(
  deseq_res_gr_and_normal %>%
    filter(experiment == "cdk_467_2019_08"),
  aes(
    log2FoldChange_normal, log2FoldChange_gr,
    color = agent
  )
) +
  geom_point(alpha = .5) +
  geom_abline(slope = 1, intercept = 0) +
  facet_wrap(~experiment + cell_line)

ggsave(
  file.path("plots_cdk7", "deseq_log2fc_ykl_gr_corrected_vs_normal.pdf"),
  p,
  width = 10,
  height = 8
)
```

### Baseline DGE

```{r}
baseline_meta_gr <- baseline_meta %>%
  power_inner_join(
    growth_data %>%
      filter(agent == "control") %>%
      select(-agent),
    by = c("cell_line"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

de_baseline_dge_gr <- DESeqDataSetFromMatrix(
  counts_mat[, baseline_meta_gr$sample_id],
  colData = column_to_rownames(baseline_meta_gr, "sample_id"),
  design = ~1 + doubling_per_hour
) %>%
  DESeq()

de_baseline_dge_gr_res <- de_baseline_dge_gr %>%
  extract_deseq_result(name = "doubling_per_hour") %>%
  dplyr::rename(ensembl_gene_id = gene_id) %>%
  power_left_join(
    ensembl_gtf %>%
      select(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )


write_csv(
  de_baseline_dge_gr_res,
  file.path("deseq", "cdk7_dge_deseq_baseline_res_gr.csv")
)
```


Plot some genes to spot check

```{r}
dge_gr_res_spot_check <- withr::with_seed(
  42,
  de_baseline_dge_gr_res %>%
    filter(padj < 0.05) %>%
    slice_sample(n = 20)
) %>%
  power_inner_join(
    power_inner_join(
      counts_corrected,
      baseline_meta_gr,
      by = "sample_id",
      check = check_specs(
        duplicate_keys_right = "warn"
      )
    ),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  )

p <- dge_gr_res_spot_check %>%
  ggplot(
    aes(
      doubling_per_hour, count,
      color = cell_line
    )
  ) +
    geom_point() +
    # geom_text(
    #   aes(label = reason_included),
    #   x = -Inf, y = Inf, color = "black",
    #   hjust = 0, vjust = 1,
    #   data = \(x) distinct(x, hgnc_symbol, reason_included)
    # ) +
    scale_y_continuous(transform = scales::pseudo_log_trans()) +
    facet_wrap(~hgnc_symbol, scales = "free_y") +
    theme_linedraw()

ggsave(
  file.path("plots_cdk7", "dge_gr_res_spot_check.pdf"),
  p,
  width = 14,
  height = 10
)
```


### Deep RNA-seq

```{r}
deep_meta_gr <- deep_meta %>%
  mutate(
    cell_line = recode(
      cell_line,
      `SNU-8` = "SNU8",
      `OVCA-R4` = "OVCAR4",
      `JHOM-1` = "JHOM1",
      `JHOS-2` = "JHOS2",
      `OAW-28` = "OAW28"
    )
  ) %>%
  power_inner_join(
    growth_data %>%
      filter(agent == "control"),
    by = c("cell_line"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    doubling_per_hour_scaled = scale(doubling_per_hour)[, 1]
  )

deep_dge_gr <- DESeqDataSetFromMatrix(
  deep_counts_mat[, deep_meta_gr$sample_id],
  colData = column_to_rownames(deep_meta_gr, "sample_id"),
  design = ~1 + doubling_per_hour
) %>%
  DESeq()

deep_dge_gr_res <- deep_dge_gr %>%
  extract_deseq_result(name = "doubling_per_hour") %>%
  dplyr::rename(ensembl_gene_id = gene_id) %>%
  power_left_join(
    ensembl_gtf_old %>%
      select(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )


deep_dge_gr2 <- DESeqDataSetFromMatrix(
  deep_counts_mat[, deep_meta_gr$sample_id],
  colData = column_to_rownames(deep_meta_gr, "sample_id"),
  design = ~1 + doubling_per_hour_scaled
) %>%
  DESeq()

deep_dge_gr_res2 <- deep_dge_gr2 %>%
  extract_deseq_result(name = "doubling_per_hour_scaled") %>%
  dplyr::rename(ensembl_gene_id = gene_id) %>%
  power_left_join(
    ensembl_gtf_old %>%
      select(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

deep_dge_gr_res_both <- deep_dge_gr_res2 %>%
  inner_join(
    deep_dge_gr_res,
    by = c("ensembl_gene_id", "hgnc_symbol"),
    suffix = c("_scaled", "_normal"),
  )

p <- deep_dge_gr_res_both %>%
  ggplot(
    aes(log2FoldChange_normal, log2FoldChange_scaled)
  ) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0)

p <- deep_dge_gr_res_both %>%
  ggplot(
    aes(-sign(log2FoldChange_normal) * log10(padj_normal), -sign(log2FoldChange_scaled) * log10(padj_scaled))
  ) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0)
    # coord_equal()

ggsave(
  file.path("plots_cdk7", "deep_de_genes_scaled_vs_normal.pdf"),
  p,
  width = 10,
  height = 8
)

write_csv(
  deep_dge_gr_res_both,
  file.path("deseq", "cdk7_deep_baseline_res_gr.csv")
)
# deep_dge_gr_res_both <- read_csv(file.path("deseq", "cdk7_deep_baseline_res_gr.csv"))
```


```{r}
synStoreMany(
  c(
    file.path("deseq", "cdk7_deseq_baseline_res.csv"),
    file.path("deseq", "cdk7_deep_baseline_res_gr.csv")
  ),
  syn_analysis,
  forceVersion = FALSE
)

```


Plot some genes to spot check

```{r}
deep_gr_res_spot_check <- withr::with_seed(
  42,
  deep_dge_gr_res_both %>%
    mutate(
      reason_included = imap(
        list(
          top_padj_scaled = order(padj_scaled)[1:10],
          top_padj_normal = order(padj_normal)[1:10],
          random_padj_scaled = which(padj_scaled < 0.05) %>%
            sample(10),
          random_padj_normal = which(padj_normal < 0.05) %>%
            sample(10)
        ),
        \(x, y) {
          z <- rep_len("", nrow(deep_dge_gr_res_both))
          z[x] <- y
          z
        }
      ) %>%
        {
          lift_dl(paste, sep = "")(.)
        }
    )
) %>%
  filter(reason_included != "") %>%
  power_inner_join(
    power_inner_join(
      deep_counts_corrected,
      deep_meta_gr,
      by = "sample_id",
      check = check_specs(
        duplicate_keys_right = "warn"
      )
    ),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  )

p <- deep_gr_res_spot_check %>%
  ggplot(
    aes(
      doubling_per_hour, count,
      color = cell_line
    )
  ) +
    geom_point() +
    # geom_text(
    #   aes(label = reason_included),
    #   x = -Inf, y = Inf, color = "black",
    #   hjust = 0, vjust = 1,
    #   data = \(x) distinct(x, hgnc_symbol, reason_included)
    # ) +
    scale_y_log10() +
    facet_wrap(~reason_included + hgnc_symbol, scales = "free_y")

ggsave(
  file.path("plots_cdk7", "deep_gr_res_spot_check.pdf"),
  p,
  width = 14,
  height = 10
)


p <- deep_gr_res_spot_check %>%
  filter(reason_included == "random_padj_normal") %>%
  ggplot(
    aes(
      doubling_per_hour, count,
      color = cell_line
    )
  ) +
    geom_point() +
    # geom_text(
    #   aes(label = reason_included),
    #   x = -Inf, y = Inf, color = "black",
    #   hjust = 0, vjust = 1,
    #   data = \(x) distinct(x, hgnc_symbol, reason_included)
    # ) +
    scale_y_continuous(trans = scales::transform_pseudo_log()) +
    facet_wrap(~hgnc_symbol, scales = "free_y")

ggsave(
  file.path("plots_cdk7", "deep_gr_res_spot_check_random_normal.pdf"),
  p,
  width = 7,
  height = 5
)
```

unscaled version much much better than scaled. Weird artifacts in the scaled version.

Comparing DGE vs deep

```{r}
de_baseline_dge_vs_deep <- deep_dge_gr_res_both %>%
  inner_join(
    de_baseline_dge_gr_res,
    by = c("ensembl_gene_id", "hgnc_symbol"),
    suffix = c("_deep", "_baseline")
  )

p <- de_baseline_dge_vs_deep %>%
  ggplot(
    aes(log2FoldChange_normal, log2FoldChange)
  ) +
    geom_point(alpha = 0.5) +
    geom_abline(slope = 1, intercept = 0) +
    coord_equal() +
    labs(
      x = "Deep",
      y = "DGE"
    )

ggsave(
  file.path("plots_cdk7", "doubling_time_deep_vs_baseline_unscaled.pdf"),
  p,
  width = 5,
  height = 5
)

cor.test(
  de_baseline_dge_vs_deep$log2FoldChange_normal,
  de_baseline_dge_vs_deep$log2FoldChange
)
```

Poor correlation

## Correcting counts

```{r}
de_baseline_dge_gr_res %>%
  ggplot(
    aes(
      log2FoldChange_MLE, log2FoldChange
    )
  ) +
  geom_point(alpha = .5)

deseq_res_long_post_gr_meta <- deseq_res_long %>%
  distinct(experiment, agent, cell_line, time) %>%
  inner_join(
    meta_deseq_gr
  )

deseq_input_gr_just_doubling <- meta_deseq_gr %>%
  group_nest(experiment, .key = "meta") %>%
  mutate(
    counts = map(
      meta,
      \(x) counts_mat[, x$sample_id]
    )
    # batch_correct = map_lgl(meta_deseq, \(x) n_distinct(x$batch) > 1)
  )


deseq_fit_gr_just_doubling <- deseq_input_gr_just_doubling %>%
  mutate(
    deseq = map2(
      meta, counts,
      \(m, x) {
        DESeqDataSetFromMatrix(
          x,
          colData = column_to_rownames(m, "sample_id"),
          design = ~doubling_per_hour,
          # design = if (bc) {~log_agent_conc + batch} else {~log_agent_conc}
        ) %>%
          DESeq()
      }
    )
  )


deseq_res_gr_just_doubling <- deseq_fit_gr_just_doubling %>%
  rowwise() %>%
  mutate(
    res = extract_deseq_result(deseq, name = "doubling_per_hour") %>%
      dplyr::rename(ensembl_gene_id = gene_id) %>%
      left_join(
        ensembl_gtf %>%
          select(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      list(),
    res_int = extract_deseq_result(deseq, name = "Intercept") %>%
      dplyr::rename(ensembl_gene_id = gene_id) %>%
      left_join(
        ensembl_gtf %>%
          select(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      list(),
    cor_terms = list(
      res$log2FoldChange %*% t(meta$doubling_per_hour)
    ),
    counts_cor = list(
      exp(log(counts(
        deseq, normalized = TRUE
      )) - cor_terms)
    )
  ) %>%
  ungroup()

qsave(
  deseq_res_gr_just_doubling %>%
    select(-deseq),
  file.path("deseq", "cdk7_deseq_gr_res_just_doubling.qs")
)
# deseq_res_gr <- qread("deseq/cdk7_deseq_gr_res.qs")

dge_gr_res_all_spot_check <- withr::with_seed(
  42,
  deseq_res_gr_just_doubling$res[[1]] %>%
    filter(padj < 0.05) %>%
    slice_sample(n = 20)
) %>%
  power_inner_join(
    power_inner_join(
      counts_corrected,
      baseline_meta_gr,
      by = "sample_id",
      check = check_specs(
        duplicate_keys_right = "warn"
      )
    ),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn"
    )
  )

p <- dge_gr_res_spot_check %>%
  ggplot(
    aes(
      doubling_per_hour, count,
      color = cell_line
    )
  ) +
    geom_point() +
    # geom_text(
    #   aes(label = reason_included),
    #   x = -Inf, y = Inf, color = "black",
    #   hjust = 0, vjust = 1,
    #   data = \(x) distinct(x, hgnc_symbol, reason_included)
    # ) +
    scale_y_continuous(transform = scales::pseudo_log_trans()) +
    facet_wrap(~hgnc_symbol, scales = "free_y") +
    theme_linedraw()

ggsave(
  file.path("plots_cdk7", "dge_gr_res_spot_check.pdf"),
  p,
  width = 14,
  height = 10
)

```

### Running Deseq on corrected counts

```{r}
deseq_res_gr_just_doubling_count_mat_cor <- rlang::exec(
  cbind,
  !!!deseq_res_gr_just_doubling$counts_cor
) %>%
  round()


deseq_input <- meta_deseq_gr %>%
  filter(agent != "control") %>%
  group_nest(experiment, agent, cell_line, time, .key = "treated") %>%
  left_join(
    meta_deseq_gr %>%
      filter(agent == "control") %>%
      group_nest(experiment, cell_line, time, .key = "control"),
    by = c("experiment", "cell_line", "time")
  ) %>%
  mutate(
    control = map2(
      control, treated,
      \(x, y) filter(x, batch %in% y$batch)
    )
  ) %>%
  filter(
    map_int(treated, nrow) > 1,
    map_int(control, nrow) > 1
  ) %>%
  mutate(
    meta_deseq = map2(
      treated, control,
      ~bind_rows(.x, .y) %>%
        mutate(
          log_agent_conc = log10(agent_conc + 1)
        )
    ),
    batch_correct = map_lgl(meta_deseq, \(x) n_distinct(x$batch) > 1)
  )

deseq_fit <- deseq_input %>%
  mutate(
    deseq = map2(
      meta_deseq, batch_correct,
      \(m, bc) {
        DESeqDataSetFromMatrix(
          deseq_res_gr_just_doubling_count_mat_cor[, m$sample_id],
          colData = column_to_rownames(m, "sample_id"),
          # design = ~log_agent_conc
          design = if (bc) {~log_agent_conc + batch} else {~log_agent_conc}
        ) %>%
          DESeq(test = "Wald")
      }
    )
  )

qsave(
  deseq_fit,
  file.path("deseq", "cdk7_deseq_res_raw_counts_gr_corrected.qs")
)
# deseq_fit <- qread("deseq/cdk7_deseq_res_raw2.qs")

deseq_res <- deseq_fit %>%
  rowwise() %>%
  mutate(
    res = {
      extract_deseq_result(
        deseq, name = "log_agent_conc"
      ) %>%
        dplyr::rename(
          ensembl_gene_id = gene_id
        ) %>%
        left_join(
          ensembl_gtf %>%
            select(ensembl_gene_id, hgnc_symbol),
          by = "ensembl_gene_id"
        ) %>%
        list()
    }
  ) %>%
  ungroup()

qsave(
  deseq_res %>%
    select(-deseq),
  file.path("deseq", "cdk7_deseq_res_counts_gr_corrected.qs")
)
# deseq_res <- qread("deseq/cdk7_deseq_res2.qs")


deseq_res_long <- deseq_res %>%
  select(experiment, agent, cell_line, time, res) %>%
  unnest(res)

write_csv(
  deseq_res_long,
  file.path("deseq", "cdk7_deseq_res_counts_gr_corrected_long.csv")
)
```


```{r}
synStoreMany(
  c(
    file.path("deseq", "cdk7_deseq_res_counts_gr_corrected_long.csv")
  ),
  parentId = "syn59470136",
  forceVersion = FALSE
)

```

## Acquired resistance

```{r}
ensembl_gtf_file_111 <- "Homo_sapiens.GRCh38.111.gtf.gz"
if (!file.exists(ensembl_gtf_file_111)) {
  download.file(
    "ftp://ftp.ensembl.org/pub/release-111/gtf/homo_sapiens/Homo_sapiens.GRCh38.111.gtf.gz",
    ensembl_gtf_file_111, method = "curl"
  )
}

ensembl_gtf_111 <- rtracklayer::readGFF(ensembl_gtf_file_111) %>%
  filter(type == "gene") %>%
  distinct(
    ensembl_gene_id = gene_id,
    hgnc_symbol = gene_name,
    gene_biotype
  ) %>%
  drop_na(ensembl_gene_id) %>%
  as_tibble()

meta_acquired <- syn("syn69681450") %>%
  read_csv()

de_acquired <- syn("syn69681454") %>%
  qread()

counts_acquired <- syn("syn69681453") %>%
  read_csv()
```


Compare every condition vs the parental condition

```{r}
condition_meta_acquired <- meta_acquired %>%
  distinct(cell_line, treatment, condition, resistant, bridge_sample)

res_acquired <- condition_meta_acquired %>%
  filter(treatment != "parent") %>%
  power_inner_join(
    condition_meta_acquired %>%
      filter(treatment == "parent") %>%
      distinct(cell_line, control_condition = condition),
    by = c("cell_line"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  rowwise() %>%
  mutate(
    res = extract_deseq_result(
      de_acquired,
      contrast = c(
        "condition",
        condition,
        control_condition
      )
    ) %>%
      list()
  ) %>%
  ungroup()

res_acquired_long <- res_acquired %>%
  unnest(res)

write_csv(
  res_acquired_long,
  file.path("deseq", "cdk7_res_acquired_long.csv.gz")
)
```


### Acquired multifactor model

```{r}
meta_acquired_multi <- meta_acquired %>%
  mutate(
    drug_treatment = case_when(
      str_detect(treatment, fixed("YKLr")) | treatment == "Y1uM" ~ "YKL-5-124",
      str_detect(treatment, fixed("SYr")) | treatment == "S100nM" ~ "SY-5609",
      TRUE ~ "DMSO"
    ) %>%
     as.factor() %>%
      relevel("DMSO"),
    resistant = replace_na(resistant, "no") %>%
      recode(parent = "no") %>%
      as.factor() %>%
      relevel("no"),
    resistant_overall = str_remove(resistant, "[0-9]+$") %>%
      as.factor() %>%
      relevel("no")
    # resistant_specific = paste0(resistant, "_", cell_line) %>%
    #   as.factor() %>%
    #   relevel("no_OVCAR4")
  ) %>%
  group_by(cell_line, resistant_overall) %>%
  mutate(
    resistant_batch = if (cur_group()$resistant_overall == "no")
      1L
    else
      as.integer(as.factor(as.character(resistant)))
  ) %>%
  ungroup() %>%
  mutate(
    resistant_batch = paste0("batch", resistant_batch) %>%
      as.factor() %>%
      relevel("batch1")
  )

df <- ~batch + cell_line + drug_treatment + drug_treatment:cell_line + resistant_overall + resistant_overall:cell_line + resistant_overall:cell_line:resistant_batch

mm <- model.matrix(
  df,
  data = meta_acquired_multi
) %>%
  .[, colSums(.) != 0]

de_acquired_multi <- DESeqDataSetFromMatrix(
  counts(de_acquired, normalized = FALSE)[, meta_acquired_multi$sample_id],
  colData = column_to_rownames(meta_acquired_multi, "sample_id"),
  design = mm
) %>%
  DESeq()

# Fit linear model for a single gene
single_gene_data <- meta_acquired_multi %>%
  mutate(
    expression = counts(de_acquired, normalized = TRUE)["ENSG00000170312", sample_id]
  )

mdl <- lm(
  expression ~ batch + cell_line + drug_treatment + drug_treatment:cell_line + resistant_overall + resistant_overall:cell_line + resistant_overall:cell_line:resistant_batch,
  data = single_gene_data
)

library(emmeans)
# Perform pairwise comparisons between resistant_overall groups using estimated marginal means
#
# This code extracts all pairwise contrasts between levels of the resistant_overall factor
# from the fitted model. To recover the actual contrast vectors/coefficients that emmeans
# uses for these comparisons, you can:
#
# 1. Use summary() with infer=c(TRUE,TRUE) to see contrast coefficients:
#    summary(xxxx, infer=c(TRUE,TRUE))
#
# 2. Extract contrast matrix directly:
#    xxxx@linfct  # for the linear function coefficients
#
# 3. Use contrast() with explicit contrast specification:
#    contrast(emmeans(mdl, ~resistant_overall), "pairwise",
#             adjust="none", infer=c(TRUE,TRUE))
#
# 4. Access the underlying contrast matrix:
#    attr(xxxx, "grid")  # to see the reference grid
#    xxxx@V              # for variance-covariance matrix of contrasts
# To exclude the effect of drug_treatment, we need to marginalize over it
# This calculates the effect of resistant_overall averaging across all drug_treatment levels

ykl_emm <- emmeans(
  mdl, ~resistant_overall,
  at = list(drug_treatment = c("YKL-5-124"))
) %>%
  contrast("trt.vs.ctrl", ref = "no")
ykl_emm

ykl_emm_by_cell <- emmeans(
  mdl, ~resistant_overall | cell_line,
  at = list(drug_treatment = c("YKL-5-124"))
) %>%
  contrast("trt.vs.ctrl", ref = "no")
ykl_emm_by_cell


res_emm_ykl <- list(
  ykl_emm,
  ykl_emm_by_cell
) %>%
  map(
    \(x) x@grid  %>%
      as_tibble() %>%
      mutate(
        contrast_vec = asplit(x@linfct, 1)
      )
  ) %>%
  bind_rows() %>%
  mutate(
    cell_line = replace_na(as.character(cell_line), "all_cells")
  ) %>%
  filter(
    contrast %in% c(
      "YKLr - no",
      "YKLrTQr - no"
    )
  ) %>%
  mutate(
    res = map(
      contrast_vec,
      \(x) {
        contrast_de <- set_names(
          x[colnames(mm)],
          resultsNames(de_acquired_multi)
        )
        extract_deseq_result(
          de_acquired_multi,
          contrast = contrast_de
        ) %>%
          mutate(
            signed_p = -log10(pvalue) * sign(log2FoldChange_MLE)
          )
      }
    )
  )

res_emm_ykl_long <- res_emm_ykl %>%
  select(-contrast_vec) %>%
  unnest(res) %>%
  power_inner_join(
    ensembl_gtf_111 %>%
      select(ensembl_gene_id, hgnc_symbol, gene_biotype),
    by = c("gene_id" = "ensembl_gene_id"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  res_emm_ykl_long,
  file.path("deseq", "cdk7_res_acquired_emm_ykl.csv.gz")
)

sy_emm <- emmeans(mdl, ~resistant_overall,
                at = list(drug_treatment = c("SY-5609"))) %>%
  contrast("trt.vs.ctrl", ref = "no")


xxxx
x <- c( "resistant_overallSY", "resistant_overallYKL", "resistant_overallYKLrTQ") %>%
  set_names() %>%
  map(
    \(x) extract_deseq_result(de_acquired_multi, name = x)
  ) %>%
  bind_rows(.id = "contrast")




res_emm <- xxxx@grid  %>%
  as_tibble() %>%
  mutate(
    res = imap(
      contrast,
      \(x, i) {
        extract_deseq_result(
          de_acquired_multi,
          contrast = set_names(xxxx@linfct[i, colnames(mm)], colnames(mm))
        )
      }
    )
  )
```


```{r}
counts_acquired_long <- counts_acquired %>%
  pivot_longer(
    -c(type, ensembl_gene_id),
    names_to = "sample_id",
    values_to = "count"
  ) %>%
  power_inner_join(
    meta_acquired_multi,
    by = "sample_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  power_inner_join(
    ensembl_gtf_111 %>%
      select(ensembl_gene_id, hgnc_symbol, gene_biotype),
    by = "ensembl_gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

counts_acquired_long_vst <- counts_acquired_long %>%
  filter(type == "vst")

write_csv(
  counts_acquired_long,
  file.path("data", "cdk7_counts_acquired_long.csv.gz")
)
counts_acquired_long <- read_csv(
  file.path("data", "cdk7_counts_acquired_long.csv.gz")
)

library(ggbeeswarm)
p <- counts_acquired_long %>%
  filter(
    type == "vst",
    ensembl_gene_id == "ENSG00000174720"
  ) %>%
  ggplot(
    aes(
      x = count,
      y = interaction(cell_line, resistant_overall),
      color = drug_treatment,
      shape = resistant_batch
    )
  ) +
  geom_quasirandom()
p
```



### Upload Synapse

```{r}
synStoreMany(
  c(
    file.path("deseq", "cdk7_res_acquired_long.csv.gz"),
    file.path("deseq", "cdk7_res_acquired_emm_ykl.csv.gz")
  ),
  parentId = "syn69681449",
  forceVersion = FALSE
)

```

