---
title: "DESeq"
author: "Clemens Hug"
date: "2023-03-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(data.table)
library(qs)
library(DESeq2)
library(synExtra)
library(IHW)

synapser::synLogin()
syn <- synDownloader("~/data", .cache = TRUE)

syn_analysis <- "syn51221227"
```

Reading count and metadata tables generated in `qc.Rmd`.

```{r}
de_meta <- syn("syn51221610") %>%
  read_csv()
de_counts <- syn("syn51221611") %>%
  read_csv()
res_lrt <- syn("syn51221607") %>%
  read_csv()

genes_for_pca <- res_lrt %>%
  filter(padj < 0.01) %>%
  pull("ensembl_gene_id")

ensembl_gtf_file <- "Homo_sapiens.GRCh38.109.gtf.gz"
if (!file.exists(ensembl_gtf_file)) {
  download.file(
    "ftp://ftp.ensembl.org/pub/release-109/gtf/homo_sapiens/Homo_sapiens.GRCh38.109.gtf.gz",
    ensembl_gtf_file, method = "curl"
  )
}

ensembl_gtf <- rtracklayer::readGFF(
  ensembl_gtf_file
) %>%
  filter(gene_biotype == "protein_coding") %>%
  distinct(
    ensembl_gene_id = gene_id,
    hgnc_symbol = gene_name,
    gene_biotype
  ) %>%
  drop_na(ensembl_gene_id)

de_count_mat <- de_counts %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

counts_vst <- syn("syn51221606") %>%
  read_csv()

vst_mat <- counts_vst %>%
  select(-hgnc_symbol) %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

counts_normalized <- syn("syn51221599") %>%
  read_csv()

```

```{r}
de_meta %>%
  filter(source_96 != "Plate 4") %>%
  dplyr::count(
    cell_line, agent, concentration, timepoint
  ) %>%
  arrange(
    cell_line, agent, concentration
  ) %>%
  View()
```

Only consider genes that have at least 5 counts in more than 10 samples to
speed up analysis. Doesn't have impact on outcome because DESeq2 would exclude
these anyways, but take longer.

```{r}
genes_to_keep <- de_count_mat %>%
  magrittr::is_greater_than(5) %>%
  rowSums() %>% {
    names(.)[. >= 10]
  }
```

Perform differential expression analysis by plate. No treatment / control pairs
cross plate boundaries so splitting it analysis up by plate makes things easier
and faster here.

I chose to set up a single model for each plate, instead of separate models for
each treatment / control pair, because estimates of expression variance might
be better that way, see https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#if-i-have-multiple-groups-should-i-run-all-together-or-split-into-pairs-of-groups

```{r}
de_groups <- de_meta %>%
  group_by(source_96) %>%
  summarize(
    des = DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, well],
      cur_data(),
      ~condition
    ) %>%
      DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)) %>%
      list()
  )
```

Identifying controls. Anything with DMSO agent is control.

```{r}
de_controls <- de_meta %>%
  filter(agent == "DMSO") %>%
  distinct(
    source_96, cell_line, timepoint, control_condition = condition
  )
```

Matching treatments with appropriate controls by checking whether for any given
treatment we have DMSO control with matching plate, cell_line, and time point.

```{r}
de_comps <- de_meta %>%
  filter(agent != "DMSO") %>%
  distinct(
    source_96, cell_line, agent, concentration, timepoint, condition
  ) %>%
  inner_join(
    de_controls
  )

results_deseq_with_shrunk <- function(de, contrast, results_args = list(), shrink_args = list()) {
  if (!missing(contrast))
    results_args <- c(
      results_args,
      list(contrast = contrast)
    )
  res <- rlang::exec(results, de, !!!results_args)
  res_shrunk <- rlang::exec(lfcShrink, de, res = res, !!!shrink_args)
  res %>%
    as.data.frame() %>%
    rownames_to_column("ensembl_gene_id") %>%
    mutate(
      padj_ihw = ihw(
        pvalue ~ baseMean,
        data = .,
        alpha = 0.1
      ) %>%
        adj_pvalues()
    ) %>%
    left_join(
      res_shrunk %>%
        as.data.frame() %>%
        rownames_to_column("ensembl_gene_id") %>%
        select(ensembl_gene_id, log2FoldChange, lfcSE),
      by = "ensembl_gene_id",
      suffix = c("_mle", "")
    )
}
results_args <- list(
  parallel = TRUE,
  BPPARAM = BiocParallel::MulticoreParam(workers = 6)
)

de_results <- de_groups %>%
  inner_join(
    de_comps
  ) %>%
  rowwise() %>%
  mutate(
    de_res = results_deseq_with_shrunk(
      des,
      contrast = c("condition", condition, control_condition),
      results_args = results_args,
      shrink_args = c(results_args, list(type = "ashr"))
    )  %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      list()
  ) %>%
  ungroup()

qsave(
  de_results,
  "deseq/deseq_res_vs_dmso.qs"
)
# de_results <- qread("deseq/deseq_res_vs_dmso.qs")

de_results_long <- de_results %>%
  select(-des) %>%
  unnest(de_res)

fwrite(
  de_results_long,
  "deseq/deseq_res_vs_dmso.csv.gz"
)

synStoreMany(
  c(
    "deseq/deseq_res_vs_dmso.csv.gz"
  ),
  parentId = "syn51221227",
  forceVersion = FALSE
)
```

Compute number of differentially expressed genes per treatment condition and
average number of gene counts in the contributing wells.

```{r}
well_counts <- de_count_mat %>%
  colSums() %>%
  enframe("well", "count") %>%
  inner_join(
    de_meta %>%
      distinct(well, condition),
    by = "well"
  )

de_stats <- de_results %>%
  select(-des) %>%
  mutate(
    n_de = map_int(de_res, ~nrow(filter(.x, padj < 0.05)))
  ) %>%
  select(-de_res) %>%
  rowwise() %>%
  mutate(
    count_avg = mean(well_counts[well_counts$condition %in% c(condition, control_condition), ]$count)
  ) %>%
  ungroup()

```

Visualize number of DE genes using beeswarm plots.

```{r}
library(ggbeeswarm)
library(ggrepel)

p <- de_stats %>%
  ggplot(aes(cell_line, n_de, color = source_96)) +
    geom_quasirandom() +
    geom_text_repel(
      aes(label = agent),
      show.legend = FALSE,
      data = ~.x %>%
        group_by(cell_line) %>%
        arrange(desc(n_de)) %>%
        mutate(agent = if_else(seq_len(n()) <= 4, agent, "")) %>%
        ungroup()
    ) +
    labs(color = "Plate", y = "N differentially expressed", x = "Cell line")

dir.create("plots/deseq")
ggsave(
  "plots/deseq/n_de_beeswarm.pdf", p, width = 7, height = 6
)

p <- de_stats %>%
  ggplot(aes(source_96, n_de, color = cell_line)) +
    geom_quasirandom() +
    geom_text_repel(
      aes(label = agent, color = NULL),
      show.legend = FALSE,
      box.padding = 0.5,
      max.overlaps = Inf,
      data = ~.x %>%
        group_by(source_96) %>%
        arrange(desc(n_de)) %>%
        mutate(agent = if_else(seq_len(n()) <= 4, agent, "")) %>%
        ungroup()
    ) +
    labs(x = "Plate", y = "N differentially expressed", color = "Cell line")

ggsave(
  "plots/deseq/n_de_beeswarm_by_plate.pdf", p, width = 7, height = 6
)
```


```{r}
p <- de_stats %>%
  ggplot(aes(count_avg, n_de, color = cell_line, shape = source_96)) +
    geom_point() +
    labs(x = "Mean well count", y = "N differentially expressed", color = "Cell line", shape = "Plate")

ggsave(
  "plots/deseq/n_de_vs_well_count.pdf", p, width = 7, height = 6
)

```

Perform differential expression for matching samples from old DGE experiments

```{r old_datasets}
old_samples_de_meta <- fread("deseq/old_samples_de_meta.csv.gz")
old_samples_de_counts <- fread("deseq/old_samples_raw_counts_long.csv.gz")

old_samples_de_count_mat <- old_samples_de_counts %>%


de_old_groups <- old_samples_de_meta %>%
  group_by(experiment) %>%
  summarize(
    des = DESeqDataSetFromMatrix(
      old_samples_de_counts[old_samples_de_meta$well %in% well & old_samples_de_meta$experiment == experiment[1]]
      de_count_mat[genes_to_keep, well],
      cur_data(),
      ~condition
    ) %>%
      DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)) %>%
      list()
  )

de_controls <- de_meta %>%
  filter(agent == "DMSO") %>%
  distinct(
    source_96, cell_line, timepoint, control_condition = condition
  )

de_comps <- de_meta %>%
  filter(agent != "DMSO") %>%
  distinct(
    source_96, cell_line, agent, concentration, timepoint, condition
  ) %>%
  inner_join(
    de_controls
  )

de_results <- de_groups %>%
  inner_join(
    de_comps
  ) %>%
  rowwise() %>%
  mutate(
    de_res = results(
      des,
      contrast = c("condition", condition, control_condition),
      tidy = TRUE,
      parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)
    ) %>%
      dplyr::rename(ensembl_gene_id = row) %>%
      list()
  ) %>%
  ungroup()

```

## Vemurafenib + FBS analysis

FBS rescues vemurafenib to some extent. Explore how they interact.

First, identifying relevant samples.

```{r}
de_meta_vemu <- de_meta %>%
  filter(source_96 == "Plate 4") %>%
  extract(
    agent, c("agent", "concentration_fbs"),
    regex = "(Vemurafenib)_([0-9]+)%FBS"
  ) %>%
  mutate(across(concentration_fbs, as.double))

write_csv(
  de_meta_vemu,
  "deseq/de_meta_vemu.csv"
)
```


Function for plotting expression of a single gene for sanity checking vemurafenib
results.


```{r}
library(ggbeeswarm)
counts_normalized_nzmin <- counts_normalized %>%
  select(-hgnc_symbol) %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix() %>% {
    .5 * min(.[. > 0])
  }
plot_expression_single_gene_vemu <- function(gene_symbol, counts = c("normalized", "vst")) {
  counts <- match.arg(counts)
  counts_table <- switch(
    counts,
    vst = counts_vst,
    normalized = counts_normalized
  )
  df <- counts_table %>%
    filter(hgnc_symbol == gene_symbol)
  if (nrow(df) == 0)
    # Try gene id instead
    df <- counts_table %>%
      filter(ensembl_gene_id == gene_symbol)
   p <- df %>%
    pivot_longer(
      -c(ensembl_gene_id, hgnc_symbol), names_to = "well", values_to = "normalized_count"
    ) %>%
    inner_join(
      de_meta_vemu,
      by = "well"
    ) %>%
    #  {
    #    if (counts == "normalized")
    #      mutate(
    #        .,
    #        zero_count = normalized_count == 0,
    #        normalized_count = if_else(zero_count, counts_normalized_nzmin, normalized_count)
    #       )
    #    else mutate(., zero_count = FALSE)
    # } %>%
    # group_by(cell_line, agent, concentration, concentration_fbs, timepoint) %>%
    # summarize(across(normalized_count, mean), .groups = "drop") %>%
    mutate(across(starts_with("concentration"), ~fct_inseq(as.character(.x), ordered = TRUE))) %>%
    ggplot(
      aes(concentration, y = normalized_count, fill = concentration_fbs, group = concentration_fbs)
    ) +
      stat_summary(geom = "col", position = "dodge") +
      geom_quasirandom(dodge.width = 0.9, varwidth = TRUE, position = "dodge") +
      facet_wrap(~cell_line) +
      # scale_shape_manual(values = c(`TRUE` = 25, `FALSE` = 16), guide = "none") +
      labs(
        x = "Vemurafenib treatment",
        fill = "FBS concentration",
        y = paste(switch(counts, vst = "Variance stabilized", normalized = "Normalized"), "gene count")
      )
   if (counts == "normalized")
     p <- p + scale_y_continuous(
       trans = scales::pseudo_log_trans(base = 10), breaks = function(...) {
         l <- list(...)
         l[[1]][1] <- max(l[[1]][1], 1)
         x <- rlang::exec(scales::breaks_log(), !!!l)
         c(0, x)
       }
      )
     # p <- p + scale_y_log10()
   p
}

```


PCA how FBS and vemurafenib interact qualitatively.

```{r}
genes_for_pca <- res_lrt %>%
  filter(padj < 0.01) %>%
  pull("ensembl_gene_id")

pca_raw_vemu <- prcomp(vst_mat[genes_for_pca, de_meta_vemu$well], scale = FALSE, center = TRUE)
pca_tidy_vemu <- broom::tidy(pca_raw_vemu, matrix = "rotation")
pca_var_tidy_vemu <- broom::tidy(pca_raw_vemu, matrix = "eigenvalues")

ps <- combn(1:4, 2, simplify = FALSE) %>% {
  tibble(component_1 = map_int(., 1), component_2 = map_int(., 2))
} %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
          pca_tidy_vemu %>%
            filter(PC %in% c(component_1, component_2)) %>%
            pivot_wider(names_from = PC, values_from = value, names_prefix = "PC_") %>%
            left_join(de_meta_vemu, by = c("column" = "well")),
          aes_string(paste0("PC_", component_1), paste0("PC_", component_2), color = "cell_line", shape = "as.factor(concentration)", size = "fct_inseq(as.character(concentration_fbs), ordered = TRUE)")
        ) +
          geom_point(alpha = 0.5) +
          labs(
            x = paste0("PC ", component_1, " (", signif(filter(pca_var_tidy_vemu, PC == component_1)$percent * 100, digits = 2), "%)"),
            y = paste0("PC ", component_2, " (", signif(filter(pca_var_tidy_vemu, PC == component_2)$percent * 100, digits = 2), "%)"),
            shape = "Vemurafenib",
            color = "Cell line",
            size = "FBS concentration"
          )
    )
  )

library(patchwork)
ps_combined <- wrap_plots(
  ps$p, guides = "collect"
)

ggsave(
  "plots/qc/vsn_counts_pca_vemu.pdf", ps_combined, width = 14, height = 10
)
```

## DESeq2

Conducting differential expression analysis using different strategies.

### Compare vs baseline sample

Compare each condition to a baseline condition. In this case use the non-vemu
treated 1% FBS condition as baseline. Compare each condition to the matching
baseline

```{r}
de_vemu <- DESeqDataSetFromMatrix(
  de_count_mat[genes_to_keep, de_meta_vemu$well],
    de_meta_vemu,
    ~condition
) %>%
  DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6))
```

Use ashr log2FoldChange shrinkage here for moderating extreme fold changes.

```{r }
de_vs_baseline_vemu <- de_meta_vemu %>%
  filter(!(concentration == 0 & concentration_fbs == 1)) %>%
  distinct(
    cell_line, agent, concentration_fbs, concentration, timepoint, condition
  ) %>%
  inner_join(
    de_meta_vemu %>%
      filter(concentration == 0, concentration_fbs == 1) %>%
      distinct(cell_line, agent, timepoint, condition_baseline = condition)
  ) %>%
  rowwise() %>%
  mutate(
    res = results_deseq_with_shrunk(
      de_vemu,
      contrast = c("condition", condition, condition_baseline),
      results_args = results_args,
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      list()
  ) %>%
  ungroup()

qsave(
  de_vs_baseline_vemu,
  "deseq/deseq_res_vs_baseline.qs"
)
# de_vs_baseline_vemu <- qread("deseq/deseq_res_vs_baseline.qs")
```

#### PCA 1-4

Plot all combinations of principal components 1-4 against each other

```{r}
de_vs_baseline_vemu_res <- de_vs_baseline_vemu %>%
  unnest(res)

de_vs_baseline_vemu_res %>%
  filter(hgnc_symbol == "ATP5MF") %>%
  mutate(across(starts_with("concentration"), ~fct_inseq(as.character(.x), ordered = TRUE))) %>%
  ggplot(aes(concentration, log2FoldChange, fill = concentration_fbs)) +
    geom_col(position = position_dodge2(preserve = "single")) +
    facet_wrap(~cell_line)

de_vs_baseline_vemu_res_mat <- de_vs_baseline_vemu_res %>%
  select(condition, ensembl_gene_id, log2FoldChange) %>%
  pivot_wider(names_from = condition, values_from = log2FoldChange) %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

de_vs_baseline_vemu_res_pca <- prcomp(
  t(de_vs_baseline_vemu_res_mat)
)

de_vs_baseline_vemu_res_pca_neat <- de_vs_baseline_vemu_res_pca %>%
  broom::tidy() %>%
  dplyr::rename(condition = row) %>%
  pivot_wider(names_from = PC, values_from = value, names_prefix = "PC_") %>%
  inner_join(
    de_meta_vemu %>%
      filter(!(concentration == 0 & concentration_fbs == 1)) %>%
      distinct(
        cell_line, agent, concentration_fbs, concentration, timepoint, condition
      )
  )

de_vs_baseline_vemu_res_pca_neat_var <- broom::tidy(de_vs_baseline_vemu_res_pca, matrix = "eigenvalues")

library(ggrepel)
ps <- combn(1:4, 2, simplify = FALSE) %>% {
  tibble(component_1 = map_int(., 1), component_2 = map_int(., 2))
} %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
          de_vs_baseline_vemu_res_pca_neat,
          aes_string(paste0("PC_", component_1), paste0("PC_", component_2), color = "cell_line", shape = "as.factor(concentration)", label = "concentration_fbs")
        ) +
          geom_point(alpha = 0.5, size = 3) +
          geom_text_repel(alpha = 0.7, color = "black", max.overlaps = Inf) +
          scale_shape_manual(values = c(`1` = 8, `0` = 16)) +
          labs(
            x = paste0("PC ", component_1, " (", signif(filter(de_vs_baseline_vemu_res_pca_neat_var, PC == component_1)$percent * 100, digits = 2), "%)"),
            y = paste0("PC ", component_2, " (", signif(filter(de_vs_baseline_vemu_res_pca_neat_var, PC == component_2)$percent * 100, digits = 2), "%)"),
            shape = "Vemurafenib",
            color = "Cell line",
            size = "FBS concentration"
          )
    )
  )

library(patchwork)
ps_combined <- wrap_plots(
  ps$p, guides = "collect"
)

ggsave(
  "vemu/plots/pca_1-4_vemu_baseline.pdf", width = 12, height = 8
)
```

#### PCA 1-10

Plot PC1 against PC2-10

```{r}
ps <- 2:10 %>% {
  tibble(component_1 = 1, component_2 = .)
} %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
          de_vs_baseline_vemu_res_pca_neat,
          aes_string(paste0("PC_", component_1), paste0("PC_", component_2), color = "cell_line", shape = "as.factor(concentration)", label = "concentration_fbs")
        ) +
          geom_point(alpha = 0.5, size = 3) +
          geom_text_repel(alpha = 0.7, color = "black", max.overlaps = Inf) +
          scale_shape_manual(values = c(`1` = 8, `0` = 16)) +
          labs(
            x = paste0("PC ", component_1, " (", signif(filter(de_vs_baseline_vemu_res_pca_neat_var, PC == component_1)$percent * 100, digits = 2), "%)"),
            y = paste0("PC ", component_2, " (", signif(filter(de_vs_baseline_vemu_res_pca_neat_var, PC == component_2)$percent * 100, digits = 2), "%)"),
            shape = "Vemurafenib",
            color = "Cell line",
            size = "FBS concentration"
          )
    )
  )

library(patchwork)
ps_combined <- wrap_plots(
  ps$p, guides = "collect"
)

ggsave(
  "vemu/plots/pca_1-10_vemu_baseline.pdf", width = 12, height = 8
)
```

#### PCA by cell line

Same PCA except perform it separately on each cell line. Should make them
a bit cleaner by removing cell type specific effect.

```{r}
de_vs_baseline_vemu_res_pca_by_cell_line <- de_vs_baseline_vemu %>%
  group_by(cell_line) %>%
  summarize(
    pca_raw = prcomp(
      t(de_vs_baseline_vemu_res_mat[, cur_data()$condition])
    ) %>%
      list(),
    .groups = "drop"
  ) %>%
  rowwise() %>%
  mutate(
    pca_neat = pca_raw %>%
      broom::tidy() %>%
      dplyr::rename(condition = row) %>%
      pivot_wider(names_from = PC, values_from = value, names_prefix = "PC_") %>%
      inner_join(
        de_meta_vemu %>%
          filter(!(concentration == 0 & concentration_fbs == 1)) %>%
          distinct(
            cell_line, agent, concentration_fbs, concentration, timepoint, condition
          )
      ) %>%
      list(),
    pca_var = broom::tidy(pca_raw, matrix = "eigenvalues") %>%
      list()
  )

ps <- de_vs_baseline_vemu_res_pca_by_cell_line %>%
  rowwise() %>%
  mutate(
    comps = tibble(component_1 = 1, component_2 = 2:min(8, nrow(pca_raw$x))) %>%
      list()
  ) %>%
  unnest(comps) %>%
  rowwise() %>%
  mutate(
    p = list(
      ggplot(
          pca_neat,
          aes_string(paste0("PC_", component_1), paste0("PC_", component_2), color = "cell_line", shape = "as.factor(concentration)", label = "concentration_fbs")
        ) +
          geom_point(alpha = 0.5, size = 3) +
          geom_text_repel(alpha = 0.7, color = "black", max.overlaps = Inf) +
          scale_shape_manual(values = c(`1` = 8, `0` = 16)) +
          labs(
            x = paste0("PC ", component_1, " (", signif(filter(pca_var, PC == component_1)$percent * 100, digits = 2), "%)"),
            y = paste0("PC ", component_2, " (", signif(filter(pca_var, PC == component_2)$percent * 100, digits = 2), "%)"),
            shape = "Vemurafenib",
            color = "Cell line",
            size = "FBS concentration",
            title = cell_line
          )
    )
  )

library(patchwork)
ps_combined <- ps %>%
  group_by(cell_line) %>%
  summarize(
    p_wrapped = wrap_plots(p, guides = "collect") %>%
      list(),
    .groups = "drop"
  )

pwalk(
  ps_combined,
  function(cell_line, p_wrapped, ...)
    ggsave(
      paste0("vemu/plots/pca_1-10_by_cell_line_", cell_line, ".pdf"),
      p_wrapped, width = 12, height = 8
    )
)

```

### Compare vemurafenib treated vs untreated pairwise

```{r}
de_comps_vemu <- de_meta_vemu %>%
  filter(concentration == 1) %>%
  distinct(
    cell_line, agent, concentration_fbs, timepoint, condition
  ) %>%
  inner_join(
    de_meta_vemu %>%
      filter(concentration == 0) %>%
      distinct(
        cell_line, agent, concentration_fbs, timepoint, condition_control = condition
      )
  )

de_results_vemu_treated_vs_untreated <- de_comps_vemu %>%
  rowwise() %>%
  mutate(
    de_res = results_deseq_with_shrunk(
      de_vemu,
      contrast = c("condition", condition, condition_control),
      results_args = results_args,
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      ) %>%
      list()
  ) %>%
  ungroup() %>%
  mutate(
    n_de = map_int(de_res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(de_res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )

qsave(
  de_results_vemu_treated_vs_untreated,
  "vemu/deseq_res_vemu_treated_vs_untreated.qs"
)


de_results_vemu_treated_vs_untreated_long <- de_results_vemu_treated_vs_untreated %>%
  select(-starts_with("n_de")) %>%
  unnest(de_res)

fwrite(
  de_results_vemu_treated_vs_untreated_long,
  "deseq/deseq_res_vemurafenib_vs_untreated.csv.gz"
)

syn_vemu_deseq <- synMkdir(syn_analysis, "deseq_vemu")

synStoreMany(
  c(
    "deseq/deseq_res_vemurafenib_vs_untreated.csv.gz"
  ),
  parentId = syn_vemu_deseq,
  forceVersion = FALSE
)
```

Make heatmap

```{r}
de_results_vemu_treated_vs_untreated_res_meta <- de_results_vemu_treated_vs_untreated %>%
  select(cell_line, agent, concentration_fbs, timepoint, condition) %>%
  arrange(cell_line, concentration_fbs) %>%
  mutate(
    condition = fct_inorder(condition)
  )

de_results_vemu_treated_vs_untreated_mat <- de_results_vemu_treated_vs_untreated %>%
  select(condition, de_res) %>%
  unnest(de_res) %>%
  select(condition, ensembl_gene_id, log2FoldChange) %>%
  pivot_wider(names_from = condition, values_from = log2FoldChange, values_fill = 0) %>%
  column_to_rownames("ensembl_gene_id") %>%
  as.matrix()

de_results_vemu_treated_vs_untreated %>%
  select(condition, de_res) %>%
  unnest(de_res) %>%
  group_by(ensembl_gene_id) %>%
  summarise(n_sig = sum(padj < 0.05, na.rm = TRUE), .groups = "drop")


library(ComplexHeatmap)
library(seriation)

ps <- de_results_vemu_treated_vs_untreated_res_meta %>%
  filter(cell_line != "COLO858") %>%
  group_by(cell_line) %>%
  summarize(
    p = de_results_vemu_treated_vs_untreated_mat[
      de_results_vemu_treated_vs_untreated %>%
        select(condition, de_res) %>%
        unnest(de_res) %>%
        group_by(ensembl_gene_id) %>%
        filter(any(padj < 0.05, na.rm = TRUE)) %>%
        pull(ensembl_gene_id) %>%
        unique(),
      as.character(cur_data()$condition)
    ] %>%
    Heatmap(
      cluster_columns = FALSE,
      cluster_rows = function(mat) {
        d <- dist(mat)
        hclust(d, method = "average") %>%
          reorder(dist = d, method = "olo")
      }
    ) %>%
      list(),
    .groups = "drop"
  )

p <- de_results_vemu_treated_vs_untreated_mat[
    de_results_vemu_treated_vs_untreated %>%
      select(condition, de_res) %>%
      unnest(de_res) %>%
      group_by(ensembl_gene_id) %>%
      filter(any(padj < 0.05, na.rm = TRUE)) %>%
      pull(ensembl_gene_id) %>%
      unique(),
    as.character(de_results_vemu_treated_vs_untreated_res_meta$condition)
  ] %>%
  Heatmap(
    cluster_columns = FALSE,
    top_annotation = HeatmapAnnotation(
      df = de_results_vemu_treated_vs_untreated_res_meta %>%
        select(condition, cell_line, concentration_fbs, timepoint) %>%
        column_to_rownames("condition")
    ),
    cluster_rows = function(mat) {
      d <- dist(mat)
      set.seed(42)
      (d, method = "average") %>%
        reorder(dist = d, method = "olo")
    },
    show_row_names = FALSE
  )


```

### Fit linear model

Scaling vemurafenib and FBS concentration to unit variance and zero mean for
better model convergence. Shouldn't affect outcome except that the magnitude of
log2FC values are a bit harder to interpret.

```{r}
de_meta_vemu_nz <- de_meta_vemu %>%
 filter(concentration_fbs > 0)

de_vemu_lin <- DESeqDataSetFromMatrix(
  de_count_mat[genes_to_keep, de_meta_vemu_nz$well],
    de_meta_vemu_nz %>%
      mutate(across(c(concentration, concentration_fbs), scale)),
    ~cell_line + concentration_fbs*concentration
) %>%
  DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6))

de_vemu_lin_res <- resultsNames(de_vemu_lin) %>%
  setdiff("Intercept") %>%
  set_names() %>%
  map(
    ~results_deseq_with_shrunk(
      de_vemu_lin,
      results_args = c(results_args, list(name = .x)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble()
  ) %>%
  enframe("coefficient", "res") %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05)))
  )
```


```{r}
de_vemu_lin_res_long <- de_vemu_lin_res %>%
  filter(coefficient != "Intercept") %>%
  select(coefficient, res) %>%
  unnest(res) %>%
  mutate(
    padj_directed = -log10(if_else(padj == 0, 0.5 * min(padj[padj > 0]), padj)) * sign(log2FoldChange)
  )

de_vemu_lin_res_long_plot_data <- de_vemu_lin_res_long %>%
  arrange(pvalue) %>%
  group_by(coefficient) %>%
  slice_head(n = 100) %>%
  ungroup() %>%
  distinct(ensembl_gene_id) %>%
  left_join(
    de_vemu_lin_res_long
  )

de_vemu_lin_res_long_plot_mat <- de_vemu_lin_res_long_plot_data %>%
  drop_na(hgnc_symbol) %>%
  select(hgnc_symbol, coefficient, log2FoldChange) %>%
  pivot_wider(values_from = log2FoldChange, names_from = coefficient, values_fill = 0) %>%
  column_to_rownames("hgnc_symbol") %>%
  as.matrix()

library(ComplexHeatmap)
library(seriation)

p <- Heatmap(
  de_vemu_lin_res_long_plot_mat[
    ,
    setdiff(colnames(de_vemu_lin_res_long_plot_mat), c("cell_line_COLO858_vs_A375", "cell_line_WM989_vs_A375"))
  ],
  cluster_columns = FALSE,
  cluster_rows = function(mat) {
    d <- dist(mat)
    (d, method = "average") %>%
      reorder(dist = d, method = "olo")
  }
)

de_vemu_lin_log2_wide <- de_vemu_lin_res_long %>%
  select(ensembl_gene_id, coefficient, log2FoldChange, padj) %>%
  pivot_wider(names_from = coefficient, values_from = c(padj, log2FoldChange))

cor.test(
  ~ log2FoldChange_concentration_fbs + log2FoldChange_concentration, data = de_vemu_lin_log2_wide
)

mod <- lm(
  concentration_fbs ~ concentration, data = de_vemu_lin_log2_wide
)

p <- de_vemu_lin_log2_wide %>%
  mutate(
    signif = (padj_concentration_fbs < 0.05 | padj_concentration < 0.05) %>%
      replace_na(FALSE)
  ) %>%
  arrange(signif) %>%
  ggplot(
    aes(
      log2FoldChange_concentration_fbs, log2FoldChange_concentration,
      alpha = signif,
      color = signif
    )
  ) +
    geom_point(shape = 16) +
    geom_smooth(aes(alpha = NULL, color = NULL), method = "lm", show.legend = FALSE) +
    scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.05)) +
    scale_color_manual(values = c(`TRUE` = "black", `FALSE` = "gray50")) +
    labs(x = "log2 fold change vemurafenib", y = "log2 fold change FBS", color = "significant", alpha = "significant")

dir.create("vemu/plots", recursive = TRUE)
ggsave(
  "vemu/plots/vemu_vs_fbs_log2fc_scatter.pdf", width = 7, height = 5
)

```



```{r}

de_vemu_lin_by_cell_line <- de_meta_vemu %>%
  filter(cell_line != "COLO858") %>%
  group_by(cell_line) %>%
  summarize(
    de = DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, cur_data()$well],
        cur_data() %>%
          mutate(across(c(concentration, concentration_fbs), ~scale(as.double(as.integer(fct_inseq(as.character(.x))))))),
        ~concentration_fbs*concentration
    ) %>%
      DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)) %>%
      list(),
    .groups = "drop"
  )

de_vemu_lin_res_by_cell_line <- de_vemu_lin_by_cell_line %>%
  mutate(
    coefs = map(de, ~tibble(coef = setdiff(resultsNames(.x), "Intercept")))
  ) %>%
  unnest(coefs) %>%
  rowwise() %>%
  mutate(
    res = results_deseq_with_shrunk(
      de,
      results_args = c(results_args, list(name = coef)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble() %>%
    list()
  ) %>%
  ungroup() %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )

qsave(
  de_vemu_lin_res_by_cell_line %>%
    select(-de),
  "deseq/deseq_res_linear_ordinal_by_cell_line.qs"
)
# de_vemu_lin_res_by_cell_line <- qread("deseq/deseq_res_linear_ordinal_by_cell_line.qs")

de_vemu_lin_res_by_cell_line_long <- de_vemu_lin_res_by_cell_line %>%
  select(-starts_with("n_de"), -any_of("de")) %>%
  unnest(res)
fwrite(
  de_vemu_lin_res_by_cell_line_long,
  "deseq/deseq_res_linear_ordinal_by_cell_line_long.csv.gz"
)
de_vemu_lin_res_by_cell_line_long <- fread(
  "deseq/deseq_res_linear_ordinal_by_cell_line_long.csv.gz"
)

de_vemu_lin_res_by_cell_line_wide <- de_vemu_lin_res_by_cell_line_long %>%
  select(cell_line, coef, ensembl_gene_id, hgnc_symbol, log2FoldChange, padj, padj_ihw) %>%
  pivot_wider(
    names_from = c(coef, cell_line),
    values_from = c(log2FoldChange, padj, padj_ihw)
  )
fwrite(
  de_vemu_lin_res_by_cell_line_wide,
  "deseq/deseq_res_linear_ordinal_by_cell_line_wide.csv.gz"
)

syn_vemu_deseq <- synMkdir(syn_analysis, "deseq_vemu")

synStoreMany(
  c(
    "deseq/deseq_res_linear_ordinal_by_cell_line_long.csv.gz",
    "deseq/deseq_res_linear_ordinal_by_cell_line_wide.csv.gz"
  ),
  parentId = syn_vemu_deseq,
  forceVersion = FALSE
)
```
How cell-line-specific is DE? Plotting log2FC against each other

```{r}
ps <- de_vemu_lin_res_by_cell_line %>%
  select(cell_line, coef, res) %>%
  unnest(res) %>%
  select(ensembl_gene_id, hgnc_symbol, coef, cell_line, log2FoldChange, padj) %>%
  pivot_wider(names_from = cell_line, values_from = c(log2FoldChange, padj)) %>%
  mutate(
    signif = replace_na(padj_A375 < 0.05, FALSE) | replace_na(padj_WM989 < 0.05, FALSE)
  ) %>%
  arrange(signif) %>%
  ggplot(
    aes(
      log2FoldChange_A375, log2FoldChange_WM989,
      alpha = signif,
      color = signif,
      # for plotly
      text = paste0(
        "<b>", fcoalesce(hgnc_symbol, ensembl_gene_id), "</b><br>",
        "p A375 ", signif(padj_A375, 3), " p WM989 ", signif(padj_WM989, 3), "<br>",
        "log2FC A375 ", signif(log2FoldChange_A375, 3), " log2FC WM989 ", signif(log2FoldChange_WM989, 3)
      )
    )
  ) +
    geom_point(shape = 16) +
    # geom_smooth(aes(alpha = NULL, color = NULL), method = "lm", show.legend = FALSE) +
    scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.05)) +
    scale_color_manual(values = c(`TRUE` = "black", `FALSE` = "gray50")) +
    facet_wrap(~coef, scales = "free")

ggsave(
  "vemu/plots/cell_line_effect_log2fc_scatter.pdf", ps, width = 13, height = 5
)


library(plotly)
pl <- ggplotly(ps, tooltip = "text") %>%
  toWebGL()
htmlwidgets::saveWidget(
  pl,
  "vemu/plots/cell_line_effect_log2fc_scatter.html",
  selfcontained = TRUE
)
```

By cell line in single model with cell-line specific effects

```{r}

de_vemu_lin_by_cell_line_single <- de_meta_vemu %>%
  filter(cell_line != "COLO858") %>%
  summarize(
    de = DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, cur_data()$well],
        cur_data() %>%
          mutate(across(c(concentration, concentration_fbs), ~scale(as.integer(fct_inseq(as.character(.x)))))),
        ~cell_line*concentration_fbs*concentration
    ) %>%
      DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)) %>%
      list(),
    .groups = "drop"
  )

de_vemu_lin_res_by_cell_line_single <- de_vemu_lin_by_cell_line_single %>%
  mutate(
    coefs = map(de, ~tibble(coef = setdiff(resultsNames(.x), "Intercept")))
  ) %>%
  unnest(coefs) %>%
  rowwise() %>%
  mutate(
    res = results_deseq_with_shrunk(
      de,
      results_args = c(results_args, list(name = coef)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble() %>%
    list()
  ) %>%
  ungroup() %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )

de_vemu_lin_res_by_cell_line_single_pairwise <- de_vemu_lin_by_cell_line_single %>%
  mutate(
    contr = tribble(
      ~coef, ~cell_line, ~contrast,
      "concentration", "A375", list(c("concentration")),
      "concentration", "WM989", list(c("concentration", "cell_lineWM989.concentration")),
      "concentration_fbs", "A375", list(c("concentration_fbs")),
      "concentration_fbs", "WM989", list(c("concentration_fbs", "cell_lineWM989.concentration_fbs"))
    ) %>%
      list()
  ) %>%
  unnest(contr) %>%
  rowwise() %>%
  mutate(
    res = results_deseq_with_shrunk(
      de,
      results_args = c(results_args, list(contrast = contrast)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble() %>%
    list()
  ) %>%
  ungroup() %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )



ps <- de_vemu_lin_res_by_cell_line_single_pairwise %>%
  select(cell_line, coef, res) %>%
  unnest(res) %>%
  select(ensembl_gene_id, hgnc_symbol, coef, cell_line, log2FoldChange, padj) %>%
  pivot_wider(names_from = cell_line, values_from = c(log2FoldChange, padj)) %>%
  left_join(
    de_vemu_lin_res_by_cell_line_single %>%
      filter(coef %in% c("cell_lineWM989.concentration_fbs", "cell_lineWM989.concentration")) %>%
      separate(coef, into = c("_", "coef"), sep = "\\.") %>%
      select(coef, res) %>%
      unnest(res) %>%
      select(ensembl_gene_id, coef, padj_interaction = padj)
  ) %>%
  mutate(
    signif = replace_na(padj_interaction < 0.05, FALSE),
    gene_id = fcoalesce(hgnc_symbol, ensembl_gene_id)
  ) %>%
  arrange(signif) %>%
  {crosstalk::SharedData$new(., key = ~gene_id)} %>%
  ggplot(
    aes(
      log2FoldChange_A375, log2FoldChange_WM989,
      alpha = signif,
      color = signif,
      # for plotly
      text = paste0(
        "<b>", gene_id, "</b><br>",
        "p A375 ", signif(padj_A375, 3), " p WM989 ", signif(padj_WM989, 3), "<br>",
        "p interaction ", signif(padj_interaction, 3), "<br>",
        "log2FC A375 ", signif(log2FoldChange_A375, 3), " log2FC WM989 ", signif(log2FoldChange_WM989, 3)
      )
    )
  ) +
    geom_point(shape = 16) +
    # geom_smooth(aes(alpha = NULL, color = NULL), method = "lm", show.legend = FALSE) +
    scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.1)) +
    scale_color_manual(values = c(`TRUE` = "black", `FALSE` = "gray50")) +
    facet_wrap(~coef, scales = "free")

ggsave(
  "vemu/plots/cell_line_effect_log2fc_scatter.pdf", ps, width = 13, height = 5
)

ggplotly(ps, tooltip = "text") %>%
  plotly::highlight(on = "plotly_hover", color = "red", opacityDim = 1) %>%
  toWebGL()
```

Get only baseline cell line effect across FBS concentrations. Get log2FoldChange
for each cell line compared to average of all cell lines

Use custom contrast in `results()` to get log2FoldChange for each cell line compared to
the average of all cell lines.

```{r}
# Do comparision separately for each FBS concentration

de_vemu_cell_line_at_baseline <- de_meta_vemu %>%
  filter(concentration == 0) %>%
  group_by(concentration_fbs) %>%
  summarize(
    de = DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, cur_data()$well],
      # cur_data(),
      # ~cell_line
       cur_data(),
      ~0 + cell_line
    ) %>%
      DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)) %>%
      list(),
    .groups = "drop"
  )

de_vemu_cell_line_at_baseline_results_names <- de_vemu_cell_line_at_baseline %>%
  chuck("de", 1) %>%
  resultsNames()

de_vemu_cell_line_at_baseline_results_contrasts <- map(
    de_vemu_cell_line_at_baseline_results_names %>%
      set_names(
        paste(str_replace(., "cell_line", ""), "vs_mean", sep = "_")
      ),
    \(x) {
      n <- length(de_vemu_cell_line_at_baseline_results_names)
      contrast_vec <- rep(
        -1/n,
        n
      ) %>%
        set_names(de_vemu_cell_line_at_baseline_results_names)
      contrast_vec[match(x, names(contrast_vec))] <- (n - 1) / n
      contrast_vec
    }
) %>%
  enframe("contrast_name", "contrast")

de_vemu_res_cell_line_at_baseline <- de_vemu_cell_line_at_baseline %>%
  cross_join(
    de_vemu_cell_line_at_baseline_results_contrasts
  ) %>%
  rowwise() %>%
  mutate(
    res = results_deseq_with_shrunk(
      de,
      results_args = c(results_args, list(contrast = contrast)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble() %>%
    list()
  ) %>%
  ungroup() %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )

qsave(
  de_vemu_res_cell_line_at_baseline %>%
    select(-de),
  "deseq/deseq_res_cell_line_effect_vs_mean.qs"
)
# de_vemu_res_cell_line_at_baseline <- qread("deseq/deseq_res_cell_line_effect_vs_mean.qs")
# Old was using cell_line*concentration_fbs
# de_vemu_res_cell_line_at_baseline_old <- qread("deseq/deseq_res_cell_line_effect_vs_mean_old.qs")

 de_vemu_res_cell_line_at_baseline_long <- de_vemu_res_cell_line_at_baseline %>%
  select(-any_of("de"), -starts_with("n_de"), -contrast) %>%
  unnest(res)

x <- de_vemu_res_cell_line_at_baseline_old %>%
  select(-c(contrast)) %>%
  unnest(res) %>%
  mutate(across(contrast_name, \(x) str_replace(x, "cell_line_", ""))) %>%
  inner_join(
    de_vemu_res_cell_line_at_baseline_long,
    by = join_by(contrast_name, ensembl_gene_id, hgnc_symbol),
    suffix = c("_old", "_new")
  )

p <- x %>%
  ggplot(
    aes(
      log2FoldChange_old,
      log2FoldChange_new
    )
  ) +
    geom_point(shape = 16, alpha = 0.5) +
    facet_grid(concentration_fbs~contrast_name) +
    coord_equal()
```

Plot cell line effects vs each other

```{r}
de_vemu_res_cell_line_at_baseline_comp <- de_vemu_res_cell_line_at_baseline %>%
  pull(contrast_name) %>%
  combn(2) %>%
  t() %>%
  magrittr::set_colnames(c("contrast_1", "contrast_2")) %>%
  as_tibble() %>%
  crossing(
    concentration_fbs = de_vemu_res_cell_line_at_baseline %>%
      pull(concentration_fbs) %>%
      unique()
  ) %>%
  inner_join(
    de_vemu_res_cell_line_at_baseline_long %>%
      select(concentration_fbs, contrast_name, ensembl_gene_id, hgnc_symbol, log2FoldChange, padj, padj_ihw),
    by = c("contrast_1" = "contrast_name", "concentration_fbs"),
    relationship = "many-to-many"
  ) %>%
  inner_join(
    de_vemu_res_cell_line_at_baseline_long %>%
      select(concentration_fbs, contrast_name, ensembl_gene_id, log2FoldChange, padj, padj_ihw),
    by = c("contrast_2" = "contrast_name", "ensembl_gene_id", "concentration_fbs"),
    suffix = c("_1", "_2")
  )

p <- de_vemu_res_cell_line_at_baseline_comp %>%
  filter(concentration_fbs == 0) %>%
  ggplot(
    aes(
      log2FoldChange_1,
      log2FoldChange_2
    )
  ) +
    geom_point(shape = 16, alpha = 0.5) +
    facet_wrap(~contrast_1 + contrast_2)

ggsave(
  "vemu/plots/deseq_cell_line_vs_mean_comp.pdf",
  p, width = 10, height = 10
)
```

```{r}

de_vemu_lin <- DESeqDataSetFromMatrix(
  de_count_mat[genes_to_keep, de_meta_vemu$well],
    de_meta_vemu %>%
      mutate(across(c(concentration, concentration_fbs), ~as.double(as.integer(fct_inseq(as.character(.x)))))),
    ~cell_line + concentration_fbs*concentration
) %>%
  DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6))

de_vemu_lin_res <- resultsNames(de_vemu_lin) %>%
  set_names() %>%
  map(
    ~results(de_vemu_lin, name = .x, tidy = TRUE) %>%
      dplyr::rename(ensembl_gene_id = row) %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      )
  ) %>%
  enframe("coefficient", "res") %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05)))
  )

de_meta_vemu_nz <- de_meta_vemu %>%
  filter(concentration_fbs != 0)

de_vemu_lin <- DESeqDataSetFromMatrix(
  de_count_mat[genes_to_keep, de_meta_vemu_nz$well],
    de_meta_vemu_nz,
    ~cell_line + concentration_fbs*concentration
) %>%
  DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6))

de_vemu_lin_res <- resultsNames(de_vemu_lin) %>%
  set_names() %>%
  map(
    ~results(de_vemu_lin, name = .x, tidy = TRUE) %>%
      dplyr::rename(ensembl_gene_id = row) %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      )
  ) %>%
  enframe("coefficient", "res") %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05)))
  )

de_meta_vemu_no_colo <- de_meta_vemu_nz %>%
  filter(cell_line != "COLO858")

de_vemu_lin <- DESeqDataSetFromMatrix(
  de_count_mat[genes_to_keep, de_meta_vemu_no_colo$well],
    de_meta_vemu_no_colo %>%
      mutate(across(c(concentration, concentration_fbs), scale)),
    ~cell_line*concentration_fbs*concentration
) %>%
  DESeq(parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6))

de_vemu_lin_res <- resultsNames(de_vemu_lin) %>%
  set_names() %>%
  map(
    ~results(de_vemu_lin, name = .x, tidy = TRUE) %>%
      dplyr::rename(ensembl_gene_id = row) %>%
      left_join(
        ensembl_gtf %>%
          distinct(ensembl_gene_id, hgnc_symbol),
        by = "ensembl_gene_id"
      )
  ) %>%
  enframe("coefficient", "res") %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05)))
  )



```

Run LRT test to check for cell-line-specific vemu effect at *any*
FBS concentration

```{r}
de_vemu_lrt_full <- de_meta_vemu %>%
  filter(cell_line != "COLO858") %>% {
    model.matrix(
      #~cell_line + concentration_fbs + concentration + cell_line:concentration_fbs:concentration,
      ~cell_line + concentration + concentration_fbs + cell_line:concentration_fbs,
      data = mutate(., across(c(concentration, concentration_fbs), ~fct_inseq(as.character(.x))))
    )
  }

de_vemu_lrt_full <- de_meta_vemu %>%
  filter(cell_line != "COLO858") %>% {
    DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, .$well],
        mutate(., across(c(concentration, concentration_fbs), ~fct_inseq(as.character(.x)))),
        ~cell_line*concentration*concentration_fbs
    ) %>%
      DESeq(
        test = "LRT",
        reduced = ~cell_line + concentration + concentration_fbs + cell_line:concentration_fbs,
        parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)
      )
  }

de_vemu_lrt_full_res <- results(de_vemu_lrt_full) %>%
  as_tibble(rownames = "ensembl_gene_id") %>%
  inner_join(
    ensembl_gtf %>%
      distinct(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id"
  )

de_vemu_lrt_full_rank <- de_meta_vemu %>%
  filter(cell_line != "COLO858") %>% {
    DESeqDataSetFromMatrix(
      de_count_mat[genes_to_keep, .$well],
        mutate(., across(c(concentration, concentration_fbs), ~scale(as.double(as.integer(fct_inseq(as.character(.x))))))),
        ~cell_line* + concentration + concentration_fbs + cell_line:concentration_fbs + cell_line:concentration + cell_line:concentration_fbs:concentration
    ) %>%
      DESeq(
        test = "LRT",
        reduced = ~cell_line + concentration_fbs + concentration + cell_line:concentration_fbs + cell_line:concentration_fbs:concentration,
        parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)
      )
  }

de_vemu_lrt_full_rank_res <- results(de_vemu_lrt_full_rank) %>%
  as_tibble(rownames = "ensembl_gene_id") %>%
  inner_join(
    ensembl_gtf %>%
      distinct(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id"
  )

ps <- de_vemu_lin_res_by_cell_line %>%
  select(cell_line, coef, res) %>%
  unnest(res) %>%
  select(ensembl_gene_id, hgnc_symbol, coef, cell_line, log2FoldChange, padj) %>%
  pivot_wider(names_from = cell_line, values_from = c(log2FoldChange, padj)) %>%
  inner_join(
    de_vemu_lrt_full_rank_res %>%
      transmute(ensembl_gene_id, signif = padj < 0.05),
    by = "ensembl_gene_id"
  ) %>%
  mutate(
    signif = replace_na(signif, FALSE)
  ) %>%
  # mutate(
  #   signif = replace_na(padj_A375 < 0.05, FALSE) | replace_na(padj_WM989 < 0.05, FALSE)
  # ) %>%
  arrange(signif) %>%
  ggplot(
    aes(
      log2FoldChange_A375, log2FoldChange_WM989,
      alpha = signif,
      color = signif,
      # for plotly
      text = paste0(
        "<b>", fcoalesce(hgnc_symbol, ensembl_gene_id), "</b><br>",
        "p A375 ", signif(padj_A375, 3), " p WM989 ", signif(padj_WM989, 3), "<br>",
        "log2FC A375 ", signif(log2FoldChange_A375, 3), " log2FC WM989 ", signif(log2FoldChange_WM989, 3)
      )
    )
  ) +
    geom_point(shape = 16) +
    # geom_smooth(aes(alpha = NULL, color = NULL), method = "lm", show.legend = FALSE) +
    scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.05)) +
    scale_color_manual(values = c(`TRUE` = "red", `FALSE` = "gray50")) +
    facet_wrap(~coef, scales = "free")


```

I suspect that this approach isn't working as well as comparing the cell-line
specific effects directly. The reason is that the LRT test is not picking up
genes that are completely not expressed in one cell line and differentially
expressed in the other. This is because the LRT test is comparing the full
model to the reduced model, and the reduced model is not able to fit the
expression of genes that are not expressed in one cell line

## Phenotypic data

The phenotypic data contains information about cell proliferation, cell migration,
nuclear shuttling, apoptosis, cell elongation, and cell area for each cell line,
with and without vemurafenib, and at low (1%) and high (5%) FBS concentrations.

Performing differential expression analysis on the phenotypic data.

### Checking multicolinearity

First, checking if any of the predictors are highly correlated with each other.

If nuclear shuttling correlates with cell elongation, exclude shuttling for now

If I need to reduce to three possible states, group them like this
(0 0.25) (0.5 0.75) (1)

```{r}
phenotypic_data <- syn("syn51438632") %>%
  readxl::read_excel(skip = 2, col_names = FALSE) %>%
  column_to_rownames("...1") %>%
  t() %>%
  as_tibble() %>%
  # Excels merged cells are filled in with NA,
  # so fill them in with the previous values
  fill(
    everything(),
    .direction = "down"
  ) %>%
  mutate(
    across(Vemurafenib, ~str_replace(.x, "uM", ""))
  ) %>%
  type_convert()

phenotypic_cor_mat <- phenotypic_data %>%
  filter(`Cell lines` != "COLO858") %>%
  select(-c(`Cell lines`, `FBS level`, `Vemurafenib`)) %>%
  scale() %>%
  cor()

library(corrplot)
corrplot(phenotypic_cor_mat)
```


```{r}
phenotypic_traits_map <- phenotypic_data %>%
  colnames() %>%
  setdiff(c("Cell lines", "FBS level", "Vemurafenib")) %>% {
    set_names(
      .,
      janitor::make_clean_names(.)
    )
  }

phenotypic_meta <- phenotypic_data %>%
  inner_join(
    tibble(`FBS level` = c("Low", "Low", "High", "High"), concentration_fbs = c(0, 1, 5, 10)),
    relationship = "many-to-many"
  ) %>%
  dplyr::rename(!!!phenotypic_traits_map) %>%
  select(
    cell_line = `Cell lines`,
    concentration_fbs,
    concentration = `Vemurafenib`,
    all_of(names(phenotypic_traits_map))
  ) %>%
  inner_join(
    de_meta_vemu
  )

phenotypic_meta_ordinal <- phenotypic_meta %>%
  transmute(
    across(
      c(cell_line),
      ~as.factor(as.character(.x))
    ),
    across(
      c(
        everything(),
        -c(cell_line, well, count, source_well_96, source_96, plate_id_96, agent, timepoint, experiment_type, condition)
      ),
      ~fct_inseq(as.character(.x), ordered = TRUE)
    )
  )

p <- GGally::ggpairs(
  phenotypic_meta_ordinal,
  lower = NULL
) +
  theme_bw()

ggsave(
  "vemu/plots/phenotypic_meta_ggpairs.pdf", p, width = 10, height = 10
)

universal_cor <- function(x, y) {
  if (is.numeric(x) && is.numeric(y)) {
    cor(x, y)
  } else if (is.ordered(x) && is.ordered(y)) {
    polycor::polychor(x, y, ML = TRUE)
  } else if (is.numeric(x) && is.ordered(y)) {
    polycor::polyserial(x, y, ML = TRUE)
  } else if (is.ordered(x) && is.numeric(y)) {
    -polycor::polyserial(y, x, ML = TRUE)
  } else if (is.factor(x) && is.ordered(y)) {
    -rcompanion::freemanTheta(y, x)
  } else if (is.factor(y) && is.ordered(x)) {
    rcompanion::freemanTheta(x, y)
  } else {
    stop("Can't compute correlation between ", class(x), " and ", class(y))
  }
}

phenotypic_meta_cor_pairs <- phenotypic_meta_ordinal %>%
  colnames() %>%
  combn(2) %>%
  t() %>%
  magrittr::set_colnames(c("var_1", "var_2")) %>%
  as_tibble() %>%
  filter(var_1 != var_2) %>%
  rowwise() %>%
  mutate(
    cor_res = universal_cor(
      phenotypic_meta_ordinal[[var_1]],
      phenotypic_meta_ordinal[[var_2]]
    )
  ) %>%
  ungroup()
```

The correlations between Vemurafenib concentration and pretty much all phenotypic
traits, except *cell_elongation* and *apoptosis_events*, are high. This
means that we might not be able to distinguish between the raw effects of
vemurafenib and what genes cause the different phenotypes.

There is almost no correlation between FBS concentration and phenotypic effects.

With regards to the *cell_line* effect, only *cell_elongation* is confounded.

```{r}

phenotypic_meta_scaled <- phenotypic_meta %>%
  mutate(
    across(
      all_of(names(phenotypic_traits_map)),
      ~scale(.x)[, 1]
    )
  )

de_vemu_phenotypic <- phenotypic_meta_scaled %>%
  filter(cell_line != "COLO858") %>% {
    DESeqDataSetFromMatrix(
      countData = de_count_mat[genes_to_keep, .$well],
      colData = .,
      design = reformulate(setdiff(names(phenotypic_traits_map), c("nuclear_shuttling")))
    ) %>%
      DESeq(
        parallel = TRUE, BPPARAM = BiocParallel::MulticoreParam(workers = 6)
      )
  }

de_vemu_phenotypic_res <- resultsNames(de_vemu_phenotypic) %>%
  setdiff("Intercept") %>%
  set_names() %>%
  map(
    ~results_deseq_with_shrunk(
      de_vemu_phenotypic,
      results_args = c(results_args, list(name = .x)),
      shrink_args = c(results_args, list(type = "ashr"))
    ) %>%
    left_join(
      ensembl_gtf %>%
        distinct(ensembl_gene_id, hgnc_symbol),
      by = "ensembl_gene_id"
    ) %>%
    as_tibble()
  ) %>%
  enframe("coefficient", "res") %>%
  mutate(
    n_de = map_int(res, ~nrow(filter(.x, padj < 0.05))),
    n_de_ihw = map_int(res, ~nrow(filter(.x, padj_ihw < 0.05)))
  )

qsave(
  de_vemu_phenotypic_res,
  "deseq/deseq_res_phenotypic.qs"
)
# de_vemu_phenotypic_res <- qread("deseq/deseq_res_phenotypic.qs")

de_vemu_phenotypic_res_long <- de_vemu_phenotypic_res %>%
  select(-starts_with("n_de")) %>%
  unnest(res)
```

Check how the differential gene expression results from the phenotypic
covariates relate to the differential expression results from the vemurafenib +
fbs covariates.

```{r}
de_vemu_lin_res_by_cell_line_with_pheno <- de_vemu_lin_res_by_cell_line_long %>%
  select(cell_line, coef, ensembl_gene_id, hgnc_symbol, log2FoldChange, padj, padj_ihw) %>%
  pivot_wider(
    names_from = c(cell_line),
    values_from = c(log2FoldChange, padj, padj_ihw)
  ) %>%
  mutate(
    linear_signif = replace_na(padj_A375 < 0.05, FALSE) | replace_na(padj_WM989 < 0.05, FALSE)
  ) %>%
  inner_join(
    de_vemu_phenotypic_res_long %>%
      transmute(
        pheno_coefficent = coefficient,
        ensembl_gene_id,
        expression = case_when(
          padj > 0.05 | is.na(padj) ~ "Not DE",
          log2FoldChange > 0 ~ "Up",
          log2FoldChange < 0 ~ "Down",
          TRUE ~ NA_character_
        ) %>%
          factor(levels = c("Up", "Down", "Not DE"))
      ),
    by = "ensembl_gene_id"
  )

de_vemu_lin_res_by_cell_line_with_pheno_scatter <- de_vemu_lin_res_by_cell_line_with_pheno %>%
  ggplot(
    aes(
      log2FoldChange_A375, log2FoldChange_WM989,
      alpha = linear_signif,
      color = expression,
      shape = linear_signif
    )
  ) +
    geom_point(shape = 16) +
    scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.2)) +
    scale_shape_manual(values = c(`TRUE` = 16, `FALSE` = 4)) +
    scale_color_manual(
      values = c(
        `Not DE` = "gray50",
        `Up` = "red",
        `Down` = "blue"
      )
    ) +
    facet_grid(
      pheno_coefficent ~ coef,
      scales = "free"
    )

ggsave(
  "vemu/plots/cell_line_effect_log2fc_scatter_pheno.pdf",
  de_vemu_lin_res_by_cell_line_with_pheno_scatter, width = 10, height = 10
)

de_vemu_lin_res_by_cell_line_with_pheno_scatter_plots <- de_vemu_lin_res_by_cell_line_with_pheno %>%
  group_by(pheno_coefficent) %>%
  summarize(
    p = list(ggplot(
      cur_data() %>%
        arrange(linear_signif),
      aes(
        log2FoldChange_A375, log2FoldChange_WM989,
        alpha = linear_signif,
        color = linear_signif
        # shape = linear_signif
      )
    ) +
      geom_point(shape = 16) +
      scale_alpha_manual(values = c(`TRUE` = 0.5, `FALSE` = 0.2)) +
      scale_color_manual(values = c(`TRUE` = "black", `FALSE` = "gray50")) +
      # scale_shape_manual(values = c(`TRUE` = 16, `FALSE` = 4)) +
      # scale_color_manual(
      #   values = c(
      #     `Not DE` = "gray50",
      #     `Up` = "red",
      #     `Down` = "blue"
      #   )
      # ) +
      ggh4x::facet_grid2(expression~coef, scales = "free", independent = "none")),
    .groups = "drop"
  )

pwalk(
  de_vemu_lin_res_by_cell_line_with_pheno_scatter_plots,
  function(pheno_coefficent, p) {
    ggsave(
      paste0("vemu/plots/cell_line_effect_log2fc_scatter_pheno_", pheno_coefficent, ".pdf"),
      p + labs(title = pheno_coefficent), width = 10, height = 6
    )
  }
)


```

Plot gene expression changes associated with phenotypic effects against gene expression
changes due to vemurafenib and FBS

```{r}
de_vemu_lin_res_by_cell_line_with_vs_pheno <- crossing(
  experimental_coefficient = unique(de_vemu_lin_res_by_cell_line_long$coef),
  phenotypic_coefficient = unique(de_vemu_phenotypic_res_long$coefficient)
) %>%
  inner_join(
    de_vemu_phenotypic_res_long %>%
      select(
        phenotypic_coefficient = coefficient,
        ensembl_gene_id, hgnc_symbol,
        log2FoldChange, padj, padj_ihw
      ),
    by = c("phenotypic_coefficient"),
    relationship = "many-to-many"
  ) %>%
  inner_join(
    de_vemu_lin_res_by_cell_line_long %>%
      select(
        coefficient = coef,
        cell_line, ensembl_gene_id, log2FoldChange, padj, padj_ihw
      ),
    suffix = c("_phenotypic", "_experimental"),
    by = c("experimental_coefficient" = "coefficient", "ensembl_gene_id"),
    relationship = "many-to-many"
  )

p <- de_vemu_lin_res_by_cell_line_with_vs_pheno %>%
  unite("experimental_coefficient", c(experimental_coefficient, cell_line)) %>%
  ggplot(
    aes(
      log2FoldChange_experimental,
      log2FoldChange_phenotypic
    )
  ) +
    geom_point(shape = 16, alpha = 0.5) +
    coord_equal() +
    ggh4x::facet_grid2(
      phenotypic_coefficient ~ experimental_coefficient,
      scales = "free"
    )

ggsave(
  "vemu/plots/cell_line_effect_log2fc_scatter_pheno_vs_experimental.pdf",
  p, width = 12, height = 10
)
```

Compute GSEA for phenotypic and experimental effects

```{r}
library(msigdbr)
library(fgsea)
all_msigdbr <- msigdbr()

msigdbr_of_interest <- all_msigdbr %>%
  filter(
    gs_cat == "H" |
    gs_subcat %in% c(
      "CP:PID",
      "CP:KEGG"
      # "CP:REACTOME"
      # "GO:BP",
      # "GO:MF"
    )
  )

fwrite(
  msigdbr_of_interest,
  "vemu/msigdbr_of_interest.csv.gz"
)

msigdbr_of_interest_gs <- msigdbr_of_interest %>%
  group_by(gs_name) %>%
  summarize(
    gs = list(ensembl_gene),
    .groups = "drop"
  ) %>%
  with(
    set_names(
      gs,
      gs_name
    )
  )

fgsea_gene_stats <- bind_rows(
  de_vemu_res_cell_line_at_baseline %>%
    dplyr::select(-any_of("de"), -starts_with("n_de")) %>%
    unnest(res) %>%
    transmute(
      concentration_fbs,
      comparison = contrast_name,
      comparison_type = "cell_line",
      ensembl_gene_id,
      log2FoldChange,
      pvalue = replace_na(pvalue, .5),
      padj_ihw,
      padj,
      signed_p = -sign(log2FoldChange) * log10(pvalue)
    ),
  de_vemu_lin_res_by_cell_line_long %>%
    transmute(
      comparison = paste(coef, cell_line, sep = "_"),
      comparison_type = "experimental",
      ensembl_gene_id,
      log2FoldChange,
      pvalue = replace_na(pvalue, .5),
      padj_ihw,
      padj,
      signed_p = -sign(log2FoldChange) * log10(pvalue)
    ),
  de_vemu_phenotypic_res_long %>%
    transmute(
      comparison = coefficient,
      comparison_type = "phenotypic",
      ensembl_gene_id,
      log2FoldChange,
      pvalue = replace_na(pvalue, .5),
      padj_ihw,
      padj,
      signed_p = -sign(log2FoldChange) * log10(pvalue)
    )
) %>%
  mutate(
    comparison_unique = if_else(is.na(concentration_fbs), comparison, paste0(comparison, "_", concentration_fbs, "%"))
  )

library(powerjoin)
fgsea_gene_stats_w_symbol <- fgsea_gene_stats %>%
  power_left_join(
    ensembl_gtf %>%
      select(ensembl_gene_id, hgnc_symbol),
    by = "ensembl_gene_id",
    check = check_specs(
      duplicate_keys_right = "warn",
      unmatched_keys_left = "warn",
      na_keys = "warn"
    )
  )

fwrite(
  fgsea_gene_stats_w_symbol,
  "vemu/deseq_res_all.csv.gz",
)

synStoreMany(
  "vemu/deseq_res_all.csv.gz",
  parentId = "syn51221227",
  forceVersion = FALSE
)
```



```{r}

fgsea_gene_stats_grouped <- fgsea_gene_stats %>%
  group_by(concentration_fbs, comparison_type, comparison, comparison_unique) %>%
  summarize(
    gene_stats = set_names(
      signed_p, ensembl_gene_id
    ) %>%
      list(),
    .groups = "drop"
  )

fgsea_res_raw <- fgsea_gene_stats_grouped %>%
  rowwise() %>%
  mutate(
    res = fgseaMultilevel(
      msigdbr_of_interest_gs,
      gene_stats,
      BPPARAM = BiocParallel::MulticoreParam(workers = 6)
    ) %>%
      list()
  ) %>%
  ungroup()

fgsea_res <- fgsea_res_raw %>%
  dplyr::select(-gene_stats) %>%
  unnest(res)

qsave(
  fgsea_res,
  "vemu/fgsea_res.qs"
)

cluster_mat <- function(mat) {
  d <- dist(mat)
  set.seed(42)
  hclust(d, method = "average") %>%
    seriation:::reorder.hclust(dist = d, method = "olo")
}

cluster_df <- function(df, x, y, z) {
  # browser()
  mat <- df %>%
    dplyr::select({{x}}, {{y}}, {{z}}) %>%
    pivot_wider(
      names_from = {{x}},
      values_from = {{z}}
    ) %>%
    column_to_rownames(rlang::as_name(rlang::ensym(y))) %>%
    as.matrix()
  # browser()
  clustering <- cluster_mat(t(mat))
  df %>%
    mutate(
      across(
        c({{x}}),
        ~factor(.x, levels = clustering$labels[clustering$order])
      )
    )
}

fgsea_res_mat <- fgsea_res %>%
  transmute(comparison_unique, pathway, signed_padj = -sign(NES)*log10(padj)) %>%
  pivot_wider(
    names_from = comparison_unique,
    values_from = signed_padj
  ) %>%
  column_to_rownames("pathway") %>%
  as.matrix() %>% {
    clust_rows <- cluster_mat(.)
    clust_cols <- cluster_mat(t(.))
    .[clust_rows$order, clust_cols$order]
  }

library(cmapR)
fgsea_res_gct <- new(
  "GCT",
  mat = fgsea_res_mat,
  rdesc = msigdbr_of_interest %>%
    distinct(
      id = gs_name,
      gs_cat, gs_subcat
    ) %>%
    dplyr::slice(match(rownames(fgsea_res_mat), id)) %>%
    as.data.frame(),
  cdesc = fgsea_gene_stats %>%
    distinct(
      id = comparison_unique,
      comparison_type
    ) %>%
    dplyr::slice(match(colnames(fgsea_res_mat), id)) %>%
    as.data.frame()
)
write_gct(
  fgsea_res_gct,
  "vemu/fgsea_res_h_pid_kegg.gct",
  appenddim = FALSE
)
```

Use topGO


```{r}
library(topGO)

topgo_gene_stats <- fgsea_gene_stats %>%
  mutate(
    signed_p = -sign(log2FoldChange) * log10(pvalue),
    across(signed_p, ~replace_na(.x, 0))
  ) %>%
  group_by(concentration_fbs, comparison, comparison_type, comparison_unique) %>%
  summarize(
    gene_stats = set_names(
      signed_p, ensembl_gene_id
    ) %>%
      list(),
    .groups = "drop"
  )

select_significant <- function(x) replace_na(x < 0.05, FALSE)

topgo_bp_object <- new(
  "topGOdata",
  ontology = "BP",
  allGenes = topgo_gene_stats$gene_stats[[1]],
  geneSelectionFun = select_significant,
  nodeSize = 10,
  annot = annFUN.org,
  ID = "ensembl",
  mapping = "org.Hs.eg.db"
)
qsave(
  topgo_bp_object,
  "vemu/topgo_bp_object.qs"
)

topgo_bp_genes <- topgo_bp_object@graph@nodes %>%
  set_names() %>%
  map(\(x) genesInTerm(topgo_bp_object, x))
qsave(
  topgo_bp_genes,
  "vemu/topgo_bp_genes.qs"
)

topgo_direction_funs <- tribble(
  ~direction, ~fun,
  "up", \(x) x > -log10(0.05),
  "down", \(x) x < log10(0.05),
  "all", \(x) abs(x) > -log10(0.05)
)

topgo_res_raw <- topgo_gene_stats %>%
  crossing(topgo_direction_funs) %>%
  rowwise() %>%
  mutate(
    res = runTest(
      updateGenes(
        topgo_bp_object,
        geneList = gene_stats,
        geneSelFun = fun
      ),
      algorithm = "weight01",
      statistic = "fisher",
      scoreOrder = "increasing"
    ) %>%
      list()
  ) %>%
  ungroup()

qsave(
  topgo_res_raw,
  "vemu/topgo_res_raw.qs"
)


topgo_res <- topgo_res_raw %>%
  rowwise() %>%
  mutate(
    res_table = GenTable(
      topgo_bp_object,
      fisher = res,
      topNodes = length(usedGO(topgo_bp_object)),
      numChar = 300
    ) %>%
      list()
  ) %>%
  ungroup() %>%
  dplyr::select(-c(res, fun, gene_stats)) %>%
  unnest(res_table) %>%
  mutate(
    term_unique = paste(
      Term, GO.ID, sep = "_"
    )
  )

qsave(
  topgo_res,
  "vemu/topgo_res.qs"
)

```


```{r}
fgsea_res_top_pathways <- fgsea_res %>%
  filter(
    pathway %in% {
      msigdbr_of_interest %>%
        # filter(gs_cat == "H") %>%
        pull(gs_name)
    }
  ) %>%
  group_by(comparison, comparison_unique) %>%
  filter(padj < 0.05) %>%
  arrange(pval) %>%
  slice_head(n = 10) %>%
  ungroup() %>%
  pull(pathway) %>%
  unique()

fgsea_res_top_clustered <- fgsea_res %>%
  filter(pathway %in% fgsea_res_top_pathways) %>%
  cluster_df(comparison_unique, pathway, NES) %>%
  cluster_df(pathway, comparison_unique, NES)

p <- fgsea_res_top_clustered %>%
  ggplot(
    aes(
      comparison_unique,
      pathway,
      fill = NES
    )
  ) +
    geom_raster() +
    geom_text(
      aes(label = padj_text),
      data = ~.x %>%
        mutate(
          padj_text = cut(
            padj,
            breaks = c(0, 0.001, 0.01, 0.05, 1),
            labels = c("***", "**", "*", "")
          )
        )
    ) +
    scale_fill_distiller(palette = "RdBu") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
    coord_equal() +
    labs(x = NULL, y = NULL)

ggsave(
  "vemu/plots/fgsea_res_h_pid_kegg_top10_clustered.pdf", p, width = 14, height = 16
)


p <- fgsea_res_top_clustered %>%
  ggplot(
    aes(
      comparison_unique,
      pathway,
      fill = NES
    )
  ) +
    geom_raster() +
    geom_text(
      aes(label = padj_text),
      data = ~.x %>%
        mutate(
          padj_text = cut(
            padj,
            breaks = c(0, 0.001, 0.01, 0.05, 1),
            labels = c("***", "**", "*", "")
          )
        )
    ) +
    scale_fill_distiller(palette = "RdBu") +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      # aspect.ratio = 10
    ) +
    # coord_equal() +
    labs(x = NULL, y = NULL) +
    ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE)

ggsave(
  "vemu/plots/fgsea_res_h_pid_kegg_top10_clustered_grouped.pdf", p, width = 12, height = 15
)

# Only cell line comparisons
fgsea_res_top_only_cell_clustered <- fgsea_res %>%
  filter(
    comparison_type == "cell_line",
    pathway %in% {
      fgsea_res %>%
        filter(
          comparison_type == "cell_line",
          pathway %in% {
            msigdbr_of_interest %>%
              # filter(gs_cat == "H") %>%
              pull(gs_name)
          }
        ) %>%
        group_by(comparison_unique) %>%
        filter(padj < 0.05) %>%
        arrange(pval) %>%
        slice_head(n = 10) %>%
        ungroup() %>%
        pull(pathway) %>%
        unique()
    }
  ) %>%
  cluster_df(comparison_unique, pathway, NES) %>%
  cluster_df(pathway, comparison_unique, NES)


p <- fgsea_res_top_only_cell_clustered %>%
  ggplot(
    aes(
      comparison_unique,
      pathway,
      fill = NES
    )
  ) +
    geom_raster() +
    geom_text(
      aes(label = padj_text),
      data = ~.x %>%
        mutate(
          padj_text = cut(
            padj,
            breaks = c(0, 0.001, 0.01, 0.05, 1),
            labels = c("***", "**", "*", "")
          )
        )
    ) +
    scale_fill_distiller(palette = "RdBu") +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
      # aspect.ratio = 10
    ) +
    # coord_equal() +
    labs(x = NULL, y = NULL) +
    ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE)

ggsave(
  "vemu/plots/fgsea_res_h_pid_kegg_top10_clustered_only_cell_lines_grouped.pdf", p, width = 5, height = 6
)

p <- Heatmap(
  fgsea_res_gct@mat[fgsea_res_top_pathways, ],
  right_annotation = rowAnnotation(
    df = fgsea_res_gct@rdesc %>%
      dplyr::slice(match(fgsea_res_top_pathways, id)) %>%
      select(gs_cat, gs_subcat)
  ),
  top_annotation = HeatmapAnnotation(
    df = fgsea_res_gct@cdesc %>%
      select(comparison_type)
  ),
  cluster_rows = cluster_mat,
  cluster_columns = cluster_mat
)
```

Topgo Heatmap


```{r}
top_go_top_clustered <- topgo_res %>%
  transmute(
    enrichment_method = "topgo",
    comparison_type, comparison, comparison_unique, direction,
    term = term_unique,
    signed_p = -log10(as.numeric(str_replace(fisher, "< ", "")))
  ) %>%
  group_by(direction) %>%
  summarize(
    data = filter(
      cur_data(),
      term %in% {
        mutate(cur_data(), significant = abs(signed_p) > -log10(0.05)) %>%
          # Remove terms that are significant for >50% of comparisons
          # group_by(term) %>%
          # filter(!((sum(significant) / n()) > 0.5)) %>%
          # ungroup() %>%
          group_by(comparison_unique) %>%
          filter(significant) %>%
          arrange(desc(abs(signed_p))) %>%
          slice_head(n = 10) %>%
          ungroup() %>%
          pull(term) %>%
          unique()
      }
    ) %>%
    cluster_df(comparison_unique, term, signed_p) %>%
    cluster_df(term, comparison_unique, signed_p) %>%
    list()
  )

top_go_top_clustered_ps <- top_go_top_clustered %>%
  rowwise() %>%
  mutate(
    p = list(ggplot(
      data,
      aes(
        comparison_unique,
        term,
        fill = signed_p
      )
    ) +
      geom_raster() +
      geom_text(
        aes(label = padj_text),
        data = ~.x %>%
          mutate(
            padj_text = cut(
              10**(-abs(signed_p)),
              breaks = c(0, 0.001, 0.01, 0.05, 1),
              labels = c("***", "**", "*", "")
            )
          ),
        color = "white"
      ) +
      # scale_fill_distiller(palette = "RdBu") +
      scale_fill_viridis_c(option = "plasma", limits = c(0, 8), oob = scales::squish) +
      theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
      # coord_equal() +
      labs(x = NULL, y = NULL) +
      ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE))
  ) %>%
  ungroup()

top_go_top_res_directional <- topgo_res %>%
  transmute(
    enrichment_method = "topgo",
    comparison_type, comparison, comparison_unique, direction,
    term = term_unique,
    p = as.numeric(str_replace(fisher, "< ", "")),
    significant = p < 0.05
  ) %>%
  pivot_wider(
    names_from = direction,
    values_from = c(p, significant)
  ) %>%
  mutate(
    signed_p = case_when(
      significant_up & significant_down ~ case_when(
        p_up < .1 * p_down ~ -log10(p_up),
        p_down < .1 * p_up ~ log10(p_down),
        TRUE ~ NA_real_
      ),
      significant_up ~ -log10(p_up),
      significant_down ~ log10(p_down),
      TRUE ~ .5 * (-log10(p_up) -log10(p_down))
    )
  )

top_go_top_clustered_directional <- top_go_top_res_directional %>%
  filter(
    term %in% {
      mutate(., significant = abs(signed_p) > -log10(0.05)) %>%
        # Remove terms that are significant for >50% of comparisons
        # group_by(term) %>%
        # filter(!((sum(significant, na.rm = TRUE) / n()) > 0.5)) %>%
        # ungroup() %>%
        group_by(comparison_unique) %>%
        filter(significant) %>%
        arrange(desc(abs(signed_p))) %>%
        slice_head(n = 10) %>%
        ungroup() %>%
        pull(term) %>%
        unique()
      }
    ) %>%
    cluster_df(comparison_unique, term, signed_p) %>%
    cluster_df(term, comparison_unique, signed_p)

library(gt)
top_go_top_clustered_directional_gt <- top_go_top_clustered_directional %>%
  dplyr::select(term, comparison_type, comparison, signed_p) %>%
  unite(comparison, comparison_type, comparison, sep = "|") %>%
  arrange(comparison) %>%
  pivot_wider(names_from = comparison, values_from = signed_p) %>%
  arrange(desc(term)) %>%
  gt(
    auto_align = TRUE,
    rowname_col = "term"
  ) %>%
  tab_spanner_delim(delim = "|") %>%
  fmt_number(
    columns = everything(),
    n_sigfig = 2
  ) %>%
  data_color(
    columns = everything(),
    method = "numeric",
    # palette = "RdBu",
    # domain = c(-4, 4),
    fn = function(x) scales::col_numeric(
      palette = "RdBu",
      reverse = TRUE,
      domain = c(-4, 4)
    )(scales::oob_squish(x, range = c(-4, 4)))
  )

gtsave(
  top_go_top_clustered_directional_gt,
  filename = "vemu/top_go_top_clustered_directional.html",
)

openxlsx::write.xlsx(
  top_go_top_clustered_directional %>%
    dplyr::select(term, comparison, signed_p) %>%
    pivot_wider(names_from = comparison, values_from = signed_p),
  "vemu/top_go_top_clustered_directional.xlsx"
)


p <- ggplot(
  top_go_top_clustered_directional,
  aes(
    comparison_unique,
    term,
    fill = signed_p
  )
) +
  geom_raster() +
  geom_text(
    aes(label = padj_text),
    data = ~.x %>%
      mutate(
        padj_text = cut(
          10**(-abs(signed_p)),
          breaks = c(0, 0.001, 0.01, 0.05, 1),
          labels = c("***", "**", "*", "")
        )
      ),
    color = "white"
  ) +
  scale_fill_distiller(palette = "RdBu", limits = c(-4, 4), oob = scales::squish) +
  # scale_fill_viridis_c(option = "plasma", limits = c(0, 8), oob = scales::squish) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  # coord_equal() +
  labs(x = NULL, y = NULL) +
  ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE)

ggsave(
  "vemu/plots/topgo_top10_clustered_directional.pdf",
  p,
  width = 10,
  height = 18
)


top_go_top_clustered_directional_cells_only <- top_go_top_res_directional %>%
  filter(
    comparison_type == "cell_line"
  ) %>%
  filter(
    term %in% {
      mutate(., significant = abs(signed_p) > -log10(0.05)) %>%
        # # Remove terms that are significant for >50% of comparisons
        # group_by(term) %>%
        # filter(!((sum(significant, na.rm = TRUE) / n()) > 0.5)) %>%
        # ungroup() %>%
        group_by(comparison_unique) %>%
        filter(significant) %>%
        arrange(desc(abs(signed_p))) %>%
        slice_head(n = 10) %>%
        ungroup() %>%
        pull(term) %>%
        unique()
      }
    ) %>%
    cluster_df(comparison_unique, term, signed_p) %>%
    cluster_df(term, comparison_unique, signed_p)

p <- ggplot(
  top_go_top_clustered_directional_cells_only,
  aes(
    comparison_unique,
    term,
    fill = signed_p
  )
) +
  geom_raster() +
  geom_text(
    aes(label = padj_text),
    data = ~.x %>%
      mutate(
        padj_text = cut(
          10**(-abs(signed_p)),
          breaks = c(0, 0.001, 0.01, 0.05, 1),
          labels = c("***", "**", "*", "")
        )
      ),
    color = "white"
  ) +
  scale_fill_distiller(palette = "RdBu", limits = c(-4, 4), oob = scales::squish) +
  # scale_fill_viridis_c(option = "plasma", limits = c(0, 8), oob = scales::squish) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  # coord_equal() +
  labs(x = NULL, y = NULL) +
  ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE)


```

Combined FGSEA and topGO results

```{r}
combined_enrichment_results <- bind_rows(
  topgo_res %>%
    transmute(
      enrichment_method = "topgo",
      comparison_type, comparison,
      term = term_unique,
      signed_p = -log10(as.numeric(fisher))
    ),
  fgsea_res %>%
    transmute(
      enrichment_method = "fgsea",
      comparison_type, comparison,
      term = pathway,
      signed_p = -sign(NES) * log10(padj)
    )
)

combined_top_pathways <- combined_enrichment_results %>%
  group_by(comparison) %>%
  filter(abs(signed_p) > -log10(0.05)) %>%
  arrange(desc(abs(signed_p))) %>%
  slice_head(n = 10) %>%
  ungroup() %>%
  pull(term) %>%
  unique()

combined_top_clustered <- combined_enrichment_results %>%
  filter(term %in% combined_top_pathways) %>%
  cluster_df(comparison, term, signed_p) %>%
  cluster_df(term, comparison, signed_p)

p <- combined_top_clustered %>%
  ggplot(
    aes(
      comparison,
      term,
      fill = signed_p
    )
  ) +
    geom_raster() +
    geom_text(
      aes(label = padj_text),
      data = ~.x %>%
        mutate(
          padj_text = cut(
            10**(-abs(signed_p)),
            breaks = c(0, 0.001, 0.01, 0.05, 1),
            labels = c("***", "**", "*", "")
          )
        )
    ) +
    scale_fill_distiller(palette = "RdBu") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
    # coord_equal() +
    labs(x = NULL, y = NULL) +
    ggforce::facet_row(~comparison_type, scales = "free_x", space = "free", drop = TRUE)


ggsave(
  "vemu/plots/fgsea_res_h_pid_kegg_topgo_top10_clustered.pdf", p, width = 10, height = 16
)


```


* Compute GSEA for vemurafenib cell line specific effects
* Copmute GSEA for interaction term
* Compute GSEA for genes rescued by high FBS from pairwise comparison


* Show heatmap of gene expression for all genes that are correlated with each
phenotypic trait - one heatmap per trait - separate by cell line

* Compute difference between cell lines at baseline (5% FBS) ignoring Vemu effect
* Do GSEA on this difference and include it in the existing heatmaps
* Make big heatmap using Morpheus


* Use not Reactome, but H, PID, and KEGG


